<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • bigsnpr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.11.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Computing polygenic scores using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="1.11.4" id="bigsnpr-1114">bigsnpr 1.11.4<a class="anchor" aria-label="anchor" href="#bigsnpr-1114"></a></h2>
<ul><li>Extend and improve LDpred2-auto to allow for estimating <span class="math inline"><em>h</em><sup>2</sup></span>, <span class="math inline"><em>p</em></span>, and <span class="math inline"><em>α</em></span>, a new third parameter that controls how expected effect sizes relate to minor allele frequencies.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.11.3" id="bigsnpr-1113">bigsnpr 1.11.3<a class="anchor" aria-label="anchor" href="#bigsnpr-1113"></a></h2>
<ul><li>Can now run <code><a href="../reference/snp_ldsc.html">snp_ldsc2()</a></code> with <code>corr</code> as an SFBM.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.11.2" id="bigsnpr-1112">bigsnpr 1.11.2<a class="anchor" aria-label="anchor" href="#bigsnpr-1112"></a></h2>
<ul><li>Now use a sparse format for sampling betas returned in LDpred2-auto, instead of a dense matrix that could require quite some memory to store.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.11.0" id="bigsnpr-1110">bigsnpr 1.11.0<a class="anchor" aria-label="anchor" href="#bigsnpr-1110"></a></h2>
<ul><li>Add two new parameters to <code><a href="../reference/snp_simuPheno.html">snp_simuPheno()</a></code>: <code>alpha</code> and <code>prob</code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.10.7" id="bigsnpr-1107">bigsnpr 1.10.7<a class="anchor" aria-label="anchor" href="#bigsnpr-1107"></a></h2>
<ul><li>Fix a liftOver error in <code><a href="../reference/snp_modifyBuild.html">snp_modifyBuild()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.10.6" id="bigsnpr-1106">bigsnpr 1.10.6<a class="anchor" aria-label="anchor" href="#bigsnpr-1106"></a></h2>
<p>Better <code><a href="../reference/snp_ldsplit.html">snp_ldsplit()</a></code>: - also return <code>$cost2</code>, the sum of squared sizes of the blocks, - for equivalent splits (with the same cost), now return the one that also minimizes cost2, - now return unique splits only (e.g. could get equivalent splits with different <code>max_size</code>).</p>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.10.5" id="bigsnpr-1105">bigsnpr 1.10.5<a class="anchor" aria-label="anchor" href="#bigsnpr-1105"></a></h2>
<ul><li>Slightly change the default parameters of lassosum2:<ul><li>
<code>delta</code> from <code>c(0.001, 0.005, 0.02, 0.1, 0.6, 3)</code> to <code>c(0.001, 0.01, 0.1, 1)</code>,</li>
<li>
<code>nlambda</code> from 20 to 30,</li>
<li>
<code>maxiter</code> from 500 to 1000.</li>
</ul></li>
<li>Add a penalty multiplicative factor for delta and lambda to regularize variants with smaller GWAS sample sizes more (when they are different, as in meta-analyses with different sets of variants).</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.10.4" id="bigsnpr-1104">bigsnpr 1.10.4<a class="anchor" aria-label="anchor" href="#bigsnpr-1104"></a></h2>
<ul><li>Now use the same updating strategy for residuals in LDpred2 as in lassosum2. This can make LDpred2-grid and LDpred2-auto an order of magnitude faster, especially for small p.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.10.2" id="bigsnpr-1102">bigsnpr 1.10.2<a class="anchor" aria-label="anchor" href="#bigsnpr-1102"></a></h2>
<ul><li>Better <code><a href="../reference/snp_modifyBuild.html">snp_modifyBuild()</a></code>: more variants should be mapped + add some QC on the mapping (a position is not mapped to more than one, the chromosome is the same, and possibly check whether we can go back to the initial position -&gt; cf. <a href="https://doi.org/10.1093/nargab/lqaa054" class="external-link uri">https://doi.org/10.1093/nargab/lqaa054</a>).</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.10.1" id="bigsnpr-1101">bigsnpr 1.10.1<a class="anchor" aria-label="anchor" href="#bigsnpr-1101"></a></h2>
<ul><li>Add two new parameters to <code><a href="../reference/snp_ldsplit.html">snp_ldsplit()</a></code>: <code>max_r2</code>, the maximum squared correlation allowed outside blocks, and <code>max_cost</code>, the maximum cost of reported solutions (i.e. the sum of all squared correlations outside blocks). Using <code>max_r2</code> offers an extra guarantee that the splitting is very good, and makes the function much faster by discarding lots of possible splits.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.10.0" id="bigsnpr-1100">bigsnpr 1.10.0<a class="anchor" aria-label="anchor" href="#bigsnpr-1100"></a></h2>
<ul><li><p>LDpred2-grid does not use OpenMP for parallelism anymore, it now simply uses multiple R processes.</p></li>
<li><p>LDpred2-grid and LDpred2-auto can now make use of <code><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed()</a></code> to get reproducible results. Note that LDpred2-inf and lassosum2 do not use any sampling.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.13" id="bigsnpr-1913">bigsnpr 1.9.13<a class="anchor" aria-label="anchor" href="#bigsnpr-1913"></a></h2>
<ul><li>Enforce <code>scipen = 50</code> when writing files to turn off scientific format (e.g. for physical positions stored as <code>double</code>).</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.12" id="bigsnpr-1912--bigsparser-06">bigsnpr 1.9.12 &amp; bigsparser 0.6<a class="anchor" aria-label="anchor" href="#bigsnpr-1912--bigsparser-06"></a></h2>
<ul><li>Use a better strategy for appending to an SFBM (<code>$add_columns()</code>).</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.8" id="bigsnpr-198">bigsnpr 1.9.8<a class="anchor" aria-label="anchor" href="#bigsnpr-198"></a></h2>
<ul><li>Fix an issue in <code><a href="../reference/snp_readBGI.html">snp_readBGI()</a></code> when using an outdated version of package {bit64}.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.7" id="bigsnpr-197">bigsnpr 1.9.7<a class="anchor" aria-label="anchor" href="#bigsnpr-197"></a></h2>
<ul><li>
<code><a href="../reference/snp_cor.html">snp_cor()</a></code> and <code><a href="../reference/snp_cor.html">bed_cor()</a></code> now use less memory.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.5" id="bigsnpr-195">bigsnpr 1.9.5<a class="anchor" aria-label="anchor" href="#bigsnpr-195"></a></h2>
<ul><li><p>Remove parameter <code>info</code> from <code><a href="../reference/snp_cor.html">snp_cor()</a></code> and <code><a href="../reference/snp_cor.html">bed_cor()</a></code> because this correction is not useful after all.</p></li>
<li><p><code><a href="../reference/snp_cor.html">snp_cor()</a></code> and <code><a href="../reference/snp_cor.html">bed_cor()</a></code> now return NaNs when e.g. the standard deviation is 0 (and warn about it). Before, these values were not reported (i.e. treated as 0).</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.4" id="bigsnpr-194">bigsnpr 1.9.4<a class="anchor" aria-label="anchor" href="#bigsnpr-194"></a></h2>
<ul><li>You can now return information on all variants with <code><a href="../reference/snp_readBGI.html">snp_readBGI()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.3" id="bigsnpr-193">bigsnpr 1.9.3<a class="anchor" aria-label="anchor" href="#bigsnpr-193"></a></h2>
<ul><li>Fix <code><a href="../reference/snp_manhattan.html">snp_manhattan()</a></code> when non-ordered (chr, pos) are provided.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.2" id="bigsnpr-192">bigsnpr 1.9.2<a class="anchor" aria-label="anchor" href="#bigsnpr-192"></a></h2>
<ul><li>Enhance function <code><a href="../reference/snp_ancestry_summary.html">snp_ancestry_summary()</a></code> by allowing to estimate ancestry proportions after PCA projection (instead of directly using the allele frequencies).</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.1" id="bigsnpr-191">bigsnpr 1.9.1<a class="anchor" aria-label="anchor" href="#bigsnpr-191"></a></h2>
<ul><li><p>Add function <code><a href="../reference/snp_cor.html">bed_cor()</a></code> (similar to <code><a href="../reference/snp_cor.html">snp_cor()</a></code> but with bed files/objects directly).</p></li>
<li><p>Add functions <code><a href="../reference/snp_ld_scores.html">snp_ld_scores()</a></code> and <code><a href="../reference/snp_ld_scores.html">bed_ld_scores()</a></code>.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.0" id="bigsnpr-190">bigsnpr 1.9.0<a class="anchor" aria-label="anchor" href="#bigsnpr-190"></a></h2>
<ul><li>Add function <code><a href="../reference/snp_ancestry_summary.html">snp_ancestry_summary()</a></code> to estimate ancestry proportions from a cohort using only its summary allele frequencies.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.11" id="bigsnpr-1811">bigsnpr 1.8.11<a class="anchor" aria-label="anchor" href="#bigsnpr-1811"></a></h2>
<ul><li>Add function <code><a href="../reference/snp_scaleBinom.html">snp_scaleAlpha()</a></code>, which is similar to <code><a href="../reference/snp_scaleBinom.html">snp_scaleBinom()</a></code>, but has a parameter <code>alpha</code> that controls the relation between the scaling and the allele frequencies.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.10" id="bigsnpr-1810">bigsnpr 1.8.10<a class="anchor" aria-label="anchor" href="#bigsnpr-1810"></a></h2>
<ul><li>Function <code><a href="../reference/snp_cor.html">snp_cor()</a></code> now also uses the upper triangle (<code>@uplo = "U"</code>) when the sparse correlation matrix is diagonal, so that it is easier to use with e.g. <code><a href="https://rdrr.io/pkg/bigsparser/man/SFBM-class.html" class="external-link">as_SFBM()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.9" id="bigsnpr-189">bigsnpr 1.8.9<a class="anchor" aria-label="anchor" href="#bigsnpr-189"></a></h2>
<ul><li>Add parameter <code>type</code> in <code><a href="../reference/snp_asGeneticPos.html">snp_asGeneticPos()</a></code> to also be able to use interpolated genetic maps from <a href="https://github.com/joepickrell/1000-genomes-genetic-maps/tree/master/interpolated_from_hapmap" class="external-link">here</a>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.8" id="bigsnpr-188">bigsnpr 1.8.8<a class="anchor" aria-label="anchor" href="#bigsnpr-188"></a></h2>
<ul><li>Add parameter <code>return_flip_and_rev</code> to <code><a href="../reference/snp_match.html">snp_match()</a></code> for whether to return internal boolean variables <code>"_FLIP_"</code> and <code>"_REV_"</code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.7" id="bigsnpr-187">bigsnpr 1.8.7<a class="anchor" aria-label="anchor" href="#bigsnpr-187"></a></h2>
<ul><li>Add <code>$perc_kept</code> in the output of <code><a href="../reference/snp_ldsplit.html">snp_ldsplit()</a></code>, the percentage of initial non-zero values kept within the blocks defined.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.6" id="bigsnpr-186">bigsnpr 1.8.6<a class="anchor" aria-label="anchor" href="#bigsnpr-186"></a></h2>
<ul><li>Faster <code><a href="../reference/snp_prodBGEN.html">snp_prodBGEN()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.5" id="bigsnpr-185">bigsnpr 1.8.5<a class="anchor" aria-label="anchor" href="#bigsnpr-185"></a></h2>
<ul><li>Add function <code><a href="../reference/snp_prodBGEN.html">snp_prodBGEN()</a></code> to compute a matrix product between BGEN files and a matrix (or a vector). This removes the need to read an intermediate FBM object with <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> to compute the product. Moreover, when using dosages, they are not rounded to two decimal places anymore.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.4" id="bigsnpr-184">bigsnpr 1.8.4<a class="anchor" aria-label="anchor" href="#bigsnpr-184"></a></h2>
<ul><li><p>Trade new parameter <code>num_iter_change</code> for a simpler <code>allow_jump_sign</code>.</p></li>
<li><p>Change defaults in LDpred2-auto to use 500 burn-in iterations (was 1000 before) followed by 200 iterations (500 before). Such a large number of iterations is usually not really needed.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.3" id="bigsnpr-183--bigsparser-05">bigsnpr 1.8.3 &amp; bigsparser 0.5<a class="anchor" aria-label="anchor" href="#bigsnpr-183--bigsparser-05"></a></h2>
<ul><li>New compact format for SFBMs which should be really useful for LDpred2 (should require about half of memory and be twice as fast). The only thing that you need to change is <code>as_SFBM(corr0, compact = TRUE)</code>. Make sure to reinstall {bigsnpr} after updating to {bigsparser} v0.5.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.2" id="bigsnpr-182">bigsnpr 1.8.2<a class="anchor" aria-label="anchor" href="#bigsnpr-182"></a></h2>
<ul><li>
<p>Prepare for incoming paper on (among other things) improved robustness of LDpred2-auto:</p>
<ul><li><p>add parameter <code>shrink_corr</code> to shrink off-diagonal elements of the LD matrix,</p></li>
<li><p>add parameter <code>num_iter_change</code> to control when starting to shrink the variants that change sign too much,</p></li>
<li><p>also return <code>corr_est</code>, the “imputed” correlations between variants and phenotypes, which can be used for post-QCing variants by comparing those to <code>beta / sqrt(n_eff * beta_se^2 + beta^2)</code>.</p></li>
</ul></li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.0" id="bigsnpr-180">bigsnpr 1.8.0<a class="anchor" aria-label="anchor" href="#bigsnpr-180"></a></h2>
<ul><li><p>Replace parameter <code>s</code> by <code>delta</code> in <code><a href="../reference/snp_lassosum2.html">snp_lassosum2()</a></code>. This new parameter <code>delta</code> better reflects that the lassosum model also uses L2-regularization (therefore, elastic-net regularization).</p></li>
<li><p>Now detect strong divergence in lassosum2 and LDpred2-grid, and return missing values for the corresponding effect sizes.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.4" id="bigsnpr-174">bigsnpr 1.7.4<a class="anchor" aria-label="anchor" href="#bigsnpr-174"></a></h2>
<ul><li>Now use a better formula for computing standard errors in <code><a href="../reference/snp_ldsc.html">snp_ldsc()</a></code> when using blocks with different sizes.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.3" id="bigsnpr-173">bigsnpr 1.7.3<a class="anchor" aria-label="anchor" href="#bigsnpr-173"></a></h2>
<ul><li>Add parameter <code>info</code> to <code><a href="../reference/snp_cor.html">snp_cor()</a></code> to correct correlations when they are computed from imputed dosage data.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.2" id="bigsnpr-172">bigsnpr 1.7.2<a class="anchor" aria-label="anchor" href="#bigsnpr-172"></a></h2>
<ul><li>Function <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> now also returns frequencies and imputation INFO scores.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.1" id="bigsnpr-171">bigsnpr 1.7.1<a class="anchor" aria-label="anchor" href="#bigsnpr-171"></a></h2>
<ul><li>Add parameter <code>rsid</code> to <code><a href="../reference/snp_asGeneticPos.html">snp_asGeneticPos()</a></code> to also allow matching with rsIDs.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.0" id="bigsnpr-170">bigsnpr 1.7.0<a class="anchor" aria-label="anchor" href="#bigsnpr-170"></a></h2>
<ul><li>Add function <code><a href="../reference/snp_lassosum2.html">snp_lassosum2()</a></code> to train the lassosum models using the exact same input data as LDpred2.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.7" id="bigsnpr-167">bigsnpr 1.6.7<a class="anchor" aria-label="anchor" href="#bigsnpr-167"></a></h2>
<ul><li>Add parameter <code>report_step</code> in <code><a href="../reference/LDpred2.html">snp_ldpred2_auto()</a></code> to report some of the internal sampling betas.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.6" id="bigsnpr-166">bigsnpr 1.6.6<a class="anchor" aria-label="anchor" href="#bigsnpr-166"></a></h2>
<ul><li>Fix crash in <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> when using BGEN files containing <code>~</code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.5" id="bigsnpr-165">bigsnpr 1.6.5<a class="anchor" aria-label="anchor" href="#bigsnpr-165"></a></h2>
<ul><li>Add parameter <code>thr_r2</code> in <code><a href="../reference/snp_cor.html">snp_cor()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.4" id="bigsnpr-164">bigsnpr 1.6.4<a class="anchor" aria-label="anchor" href="#bigsnpr-164"></a></h2>
<ul><li>Remove penalization in <code><a href="../reference/snp_ldsplit.html">snp_ldsplit()</a></code>. Instead, report the best splits for a range of numbers of blocks desired.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.3" id="bigsnpr-163">bigsnpr 1.6.3<a class="anchor" aria-label="anchor" href="#bigsnpr-163"></a></h2>
<ul><li>Penalization in <code><a href="../reference/snp_ldsplit.html">snp_ldsplit()</a></code> now makes more sense. Also fix a small bug that prevented splitting the last block in some cases.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.2" id="bigsnpr-162">bigsnpr 1.6.2<a class="anchor" aria-label="anchor" href="#bigsnpr-162"></a></h2>
<ul><li>Add function <code><a href="../reference/snp_ldsplit.html">snp_ldsplit()</a></code> for optimally splitting variants in nearly independent blocks of LD.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.1" id="bigsnpr-161">bigsnpr 1.6.1<small>2021-01-11</small><a class="anchor" aria-label="anchor" href="#bigsnpr-161"></a></h2>
<ul><li>Add option <code>file.type = "--gzvcf"</code> for using gzipped VCF in <code><a href="../reference/snp_plinkQC.html">snp_plinkQC()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.0" id="bigsnpr-160">bigsnpr 1.6.0<a class="anchor" aria-label="anchor" href="#bigsnpr-160"></a></h2>
<ul><li>Finally remove function <code>snp_assocBGEN()</code>; prefer reading small parts with <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> as a temporary <code>bigSNP</code> object and do the association test with e.g. <code>big_univLinReg()</code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.7" id="bigsnpr-157">bigsnpr 1.5.7<a class="anchor" aria-label="anchor" href="#bigsnpr-157"></a></h2>
<ul><li>Add function <code><a href="../reference/snp_thr_correct.html">snp_thr_correct()</a></code> for correcting for winner’s curse in summary statistics when using p-value thresholding.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.6" id="bigsnpr-156">bigsnpr 1.5.6<a class="anchor" aria-label="anchor" href="#bigsnpr-156"></a></h2>
<ul><li><p>Use a better formula for the scale in LDpred2, useful when there are some variants with very large effects (e.g. explaining more than 10% phenotypic variance).</p></li>
<li><p>Simplify LDpred2; there was not really any need for initialization and ordering of the Gibbs sampler.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.5" id="bigsnpr-155">bigsnpr 1.5.5<a class="anchor" aria-label="anchor" href="#bigsnpr-155"></a></h2>
<ul><li>Add option <code>return_sampling_betas</code> in <code><a href="../reference/LDpred2.html">snp_ldpred2_grid()</a></code> to return all sampling betas (after burn-in), which is useful for assessing the uncertainty of the PRS at the individual level (see <a href="https://doi.org/10.1101/2020.11.30.403188" class="external-link uri">https://doi.org/10.1101/2020.11.30.403188</a>).</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.4" id="bigsnpr-154--bigsparser-041">bigsnpr 1.5.4 &amp; bigsparser 0.4.1<a class="anchor" aria-label="anchor" href="#bigsnpr-154--bigsparser-041"></a></h2>
<ul><li><p>Faster cross-product with an SFBM, which should make all LDpred2 models faster.</p></li>
<li><p>Also return <code>$postp_est</code>, <code>$h2_init</code> and <code>$p_init</code> in LDpred2-auto.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.1" id="bigsnpr-151">bigsnpr 1.5.1<a class="anchor" aria-label="anchor" href="#bigsnpr-151"></a></h2>
<ul><li>Add multiple checks in <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> to make sure of the expected format.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.0" id="bigsnpr-150">bigsnpr 1.5.0<a class="anchor" aria-label="anchor" href="#bigsnpr-150"></a></h2>
<ul><li>Add function <code><a href="../reference/snp_fst.html">snp_fst()</a></code> for computing Fst.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.11" id="bigsnpr-1411">bigsnpr 1.4.11<a class="anchor" aria-label="anchor" href="#bigsnpr-1411"></a></h2>
<ul><li>Workaround for error <code>could not find function "ldpred2_gibbs_auto"</code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.9" id="bigsnpr-149--bigsparser-040">bigsnpr 1.4.9 &amp; bigsparser 0.4.0<a class="anchor" aria-label="anchor" href="#bigsnpr-149--bigsparser-040"></a></h2>
<ul><li>Can now directly do <code>as_SFBM(corr0)</code> instead of <code>bigsparser::as_SFBM(as(corr0, "dgCMatrix"))</code>. This should also use less memory and be faster.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.8" id="bigsnpr-148">bigsnpr 1.4.8<a class="anchor" aria-label="anchor" href="#bigsnpr-148"></a></h2>
<ul><li>Add option <code>sparse</code> to enable getting also a sparse solution in LDpred2-auto.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.7" id="bigsnpr-147--bigsparser-030">bigsnpr 1.4.7 &amp; bigsparser 0.3.0<a class="anchor" aria-label="anchor" href="#bigsnpr-147--bigsparser-030"></a></h2>
<ul><li><p>Faster <code><a href="https://rdrr.io/pkg/bigsparser/man/SFBM-class.html" class="external-link">as_SFBM()</a></code>.</p></li>
<li><p>Allow for format <code>01</code> or <code>1</code> for chromosomes in BGI files.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.6" id="bigsnpr-146">bigsnpr 1.4.6<a class="anchor" aria-label="anchor" href="#bigsnpr-146"></a></h2>
<ul><li>Fasten <code><a href="../reference/snp_match.html">snp_match()</a></code>. Also now remove duplicates by default.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.3" id="bigsnpr-143">bigsnpr 1.4.3<a class="anchor" aria-label="anchor" href="#bigsnpr-143"></a></h2>
<ul><li>Fix a bug when using very large correlation matrices in LDpred2 (although we do not recommend to do so).</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.2" id="bigsnpr-142">bigsnpr 1.4.2<a class="anchor" aria-label="anchor" href="#bigsnpr-142"></a></h2>
<ul><li><p>All 3 LDpred2 functions now use an SFBM as input format for the correlation matrix.</p></li>
<li><p>Allow for multiple initial values for p in <code><a href="../reference/LDpred2.html">snp_ldpred2_auto()</a></code>.</p></li>
<li><p>Add function <code><a href="../reference/coef_to_liab.html">coef_to_liab()</a></code> for e.g. converting heritability to the liability scale.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.1" id="bigsnpr-141">bigsnpr 1.4.1<a class="anchor" aria-label="anchor" href="#bigsnpr-141"></a></h2>
<ul><li>Change default of parameter <code>alpha</code> of function <code><a href="../reference/snp_cor.html">snp_cor()</a></code> to <code>1</code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.0" id="bigsnpr-140">bigsnpr 1.4.0<a class="anchor" aria-label="anchor" href="#bigsnpr-140"></a></h2>
<ul><li><p>Add functions <code><a href="../reference/LDpred2.html">snp_ldpred2_inf()</a></code>, <code><a href="../reference/LDpred2.html">snp_ldpred2_grid()</a></code> and <code><a href="../reference/LDpred2.html">snp_ldpred2_auto()</a></code> for running the new LDpred2-inf, LDpred2-grid and LDpred2-auto.</p></li>
<li><p>Add functions <code><a href="../reference/snp_ldsc.html">snp_ldsc()</a></code> and <code><a href="../reference/snp_ldsc.html">snp_ldsc2()</a></code> for performing LD score regression.</p></li>
<li><p>Add function <code><a href="../reference/snp_asGeneticPos.html">snp_asGeneticPos()</a></code> for transforming physical positions to genetic positions.</p></li>
<li><p>Add function <code><a href="../reference/snp_simuPheno.html">snp_simuPheno()</a></code> for simulating phenotypes.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.1" id="bigsnpr-131">bigsnpr 1.3.1<a class="anchor" aria-label="anchor" href="#bigsnpr-131"></a></h2>
<ul><li>Also use OpenMP for the parallelization of <code><a href="../reference/snp_pcadapt.html">snp_pcadapt()</a></code>, <code><a href="../reference/snp_pcadapt.html">bed_pcadapt()</a></code>, <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> and <code><a href="../reference/snp_fastImputeSimple.html">snp_fastImputeSimple()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.0" id="bigsnpr-130">bigsnpr 1.3.0<small>2020-03-09</small><a class="anchor" aria-label="anchor" href="#bigsnpr-130"></a></h2>
<ul><li><p>Parallelization of clumping algorithms has been modified. Before, chromosomes were imputed in parallel. Now, chromosomes are processed sequentially, but computations within each chromosome are performed in parallel thanks to OpenMP. This should prevent major slowdowns for very large samples sizes (due to swapping).</p></li>
<li><p>Use OpenMP to parallelize other functions as well (possibly only sequential until now).</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.6" id="bigsnpr-126">bigsnpr 1.2.6<a class="anchor" aria-label="anchor" href="#bigsnpr-126"></a></h2>
<ul><li><p>Can now run <code><a href="../reference/snp_cor.html">snp_cor()</a></code> in parallel.</p></li>
<li><p>Parallelization of <code><a href="../reference/snp_fastImpute.html">snp_fastImpute()</a></code> has been modified. Before this version, chromosomes were imputed in parallel. Now, chromosomes are processed sequentially, but computation of correlation between variants and XGBoost models are performed using parallelization.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.5" id="bigsnpr-125">bigsnpr 1.2.5<a class="anchor" aria-label="anchor" href="#bigsnpr-125"></a></h2>
<ul><li>Add function <code><a href="../reference/snp_subset.html">snp_subset()</a></code> as alias of method <code><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset()</a></code> for subsetting <code>bigSNP</code> objects.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.4" id="bigsnpr-124">bigsnpr 1.2.4<a class="anchor" aria-label="anchor" href="#bigsnpr-124"></a></h2>
<ul><li>Use new class <code>bed_light</code> internally to make parallel algorithms faster because they have to transfer less data to clusters. Also define differently functions used in <code>big_parallelize()</code> for the same reason.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.0" id="bigsnpr-120">bigsnpr 1.2.0<a class="anchor" aria-label="anchor" href="#bigsnpr-120"></a></h2>
<ul><li>Use the new implementation of robust OGK Mahalanobis distance in {bigutilsr}.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.1" id="bigsnpr-111">bigsnpr 1.1.1<a class="anchor" aria-label="anchor" href="#bigsnpr-111"></a></h2>
<ul><li>Fix error <code>object 'obj.bed' not found</code> in <code><a href="../reference/snp_readBed.html">snp_readBed2()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.0" id="bigsnpr-110">bigsnpr 1.1.0<a class="anchor" aria-label="anchor" href="#bigsnpr-110"></a></h2>
<ul><li>Cope with new read-only option in {bigstatsr} version &gt;= 1.1.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.0.2" id="bigsnpr-102">bigsnpr 1.0.2<a class="anchor" aria-label="anchor" href="#bigsnpr-102"></a></h2>
<ul><li>Add option <code>backingfile</code> to <code><a href="../reference/snp_subset.html">subset.bigSNP()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.0.1" id="bigsnpr-101">bigsnpr 1.0.1<a class="anchor" aria-label="anchor" href="#bigsnpr-101"></a></h2>
<ul><li>Add option <code>byrow</code> to <code><a href="../reference/bed_counts.html">bed_counts()</a></code>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.0.0" id="bigsnpr-100">bigsnpr 1.0.0<a class="anchor" aria-label="anchor" href="#bigsnpr-100"></a></h2>
<ul><li>
<p>Add memory-mapping on PLINK (.bed) files with missing values + new functions:</p>
<ul><li><code><a href="../reference/bed-class.html">bed()</a></code></li>
<li><code><a href="../reference/bed_MAF.html">bed_MAF()</a></code></li>
<li><code><a href="../reference/snp_autoSVD.html">bed_autoSVD()</a></code></li>
<li><code><a href="../reference/snp_clumping.html">bed_clumping()</a></code></li>
<li><code><a href="../reference/bed_counts.html">bed_counts()</a></code></li>
<li><code><a href="../reference/bed_cprodVec.html">bed_cprodVec()</a></code></li>
<li><code><a href="../reference/snp_pcadapt.html">bed_pcadapt()</a></code></li>
<li><code><a href="../reference/bed_prodVec.html">bed_prodVec()</a></code></li>
<li><code><a href="../reference/bed_projectPCA.html">bed_projectPCA()</a></code></li>
<li><code><a href="../reference/bed_projectSelfPCA.html">bed_projectSelfPCA()</a></code></li>
<li><code><a href="../reference/bed_randomSVD.html">bed_randomSVD()</a></code></li>
<li><code><a href="../reference/bed_scaleBinom.html">bed_scaleBinom()</a></code></li>
<li><code><a href="../reference/bed_tcrossprodSelf.html">bed_tcrossprodSelf()</a></code></li>
<li><code><a href="../reference/download_1000G.html">download_1000G()</a></code></li>
<li><code><a href="../reference/snp_modifyBuild.html">snp_modifyBuild()</a></code></li>
<li><code><a href="../reference/snp_plinkKINGQC.html">snp_plinkKINGQC()</a></code></li>
<li><code><a href="../reference/snp_readBed.html">snp_readBed2()</a></code></li>
<li><code><a href="../reference/sub_bed.html">sub_bed()</a></code></li>
</ul></li>
<li><p>Add 3 parameters to <code>autoSVD()</code>: <code>alpha.tukey</code>, <code>min.mac</code> and <code>max.iter</code>.</p></li>
<li><p>Remove option for changing ploidy (that was only partially supported).</p></li>
<li><p>Automatically apply <code><a href="../reference/snp_gc.html">snp_gc()</a></code> to <code>pcadapt</code>.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.12.0" id="bigsnpr-0120">bigsnpr 0.12.0<a class="anchor" aria-label="anchor" href="#bigsnpr-0120"></a></h2>
<ul><li>Add <code><a href="../reference/snp_fastImputeSimple.html">snp_fastImputeSimple()</a></code>: fast imputation via mode, mean or sampling according to allele frequencies.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.11.3" id="bigsnpr-0113">bigsnpr 0.11.3<a class="anchor" aria-label="anchor" href="#bigsnpr-0113"></a></h2>
<ul><li>Fix a bug in <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> that could not handle duplicated variants or individuals.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.11.1" id="bigsnpr-0111">bigsnpr 0.11.1<a class="anchor" aria-label="anchor" href="#bigsnpr-0111"></a></h2>
<ul><li>When using <code><a href="../reference/SCT.html">snp_grid_PRS()</a></code>, it now stores not only the FBM, but also the input parameters as attributes (the whole result basically).</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.11.0" id="bigsnpr-0110">bigsnpr 0.11.0<a class="anchor" aria-label="anchor" href="#bigsnpr-0110"></a></h2>
<ul><li><p>Add 3 SCT functions <code>snp_grid_*()</code> to improve from Clumping and Thresholding (preprint coming soon).</p></li>
<li><p>Add <code><a href="../reference/snp_match.html">snp_match()</a></code> function to match between summary statistics and some SNP information.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.10.2" id="bigsnpr-0102">bigsnpr 0.10.2<a class="anchor" aria-label="anchor" href="#bigsnpr-0102"></a></h2>
<ul><li>Parameter <code>is.size.in.bp</code> is deprecated.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.10.1" id="bigsnpr-0101">bigsnpr 0.10.1<a class="anchor" aria-label="anchor" href="#bigsnpr-0101"></a></h2>
<ul><li>Add parameter <code>read_as</code> for <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code>. It is now possible to sample BGEN probabilities as random hard calls using <code>read_as = "random"</code>. Default remains reading probabilities as dosages.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.10.0" id="bigsnpr-0100">bigsnpr 0.10.0<a class="anchor" aria-label="anchor" href="#bigsnpr-0100"></a></h2>
<ul><li><p>For memory-mapping, now use <em>mio</em> instead of <em>boost</em>.</p></li>
<li><p><code><a href="../reference/snp_clumping.html">snp_clumping()</a></code> (and <code><a href="../reference/snp_autoSVD.html">snp_autoSVD()</a></code>) now has a <code>size</code> that is inversely proportional to <code>thr.r2</code>.</p></li>
<li><p><code><a href="../reference/snp_clumping.html">snp_pruning()</a></code> is deprecated (and will be removed someday); now always use <code><a href="../reference/snp_clumping.html">snp_clumping()</a></code>.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.9.0" id="bigsnpr-090">bigsnpr 0.9.0<a class="anchor" aria-label="anchor" href="#bigsnpr-090"></a></h2>
<ul><li>When reading bed files, switch reading of Os and 2s to be consistent with other software.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.8.2" id="bigsnpr-082">bigsnpr 0.8.2<a class="anchor" aria-label="anchor" href="#bigsnpr-082"></a></h2>
<ul><li>Add function <code>snp_assocBGEN()</code> for computing quick association tests from BGEN files. Could be useful for quick screening of useful SNPs to read in bigSNP format. This function might be improved in the future.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.8.1" id="bigsnpr-081">bigsnpr 0.8.1<a class="anchor" aria-label="anchor" href="#bigsnpr-081"></a></h2>
<ul><li>Change url to download PLINK 1.9.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.8.0" id="bigsnpr-080">bigsnpr 0.8.0<a class="anchor" aria-label="anchor" href="#bigsnpr-080"></a></h2>
<ul><li>Add function <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> to read UK Biobank BGEN files in <code>bigSNP</code> format.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.7.0" id="bigsnpr-070">bigsnpr 0.7.0<a class="anchor" aria-label="anchor" href="#bigsnpr-070"></a></h2>
<ul><li><p>Add parameter <code>is.size.in.bp</code> to <code><a href="../reference/snp_autoSVD.html">snp_autoSVD()</a></code> for the clumping part.</p></li>
<li><p>Change the threshold of outlier detection in <code><a href="../reference/snp_autoSVD.html">snp_autoSVD()</a></code> (it now detects less outliers). See the documentation details if you don’t have any information about SNPs.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.6" id="bigsnpr-06">bigsnpr 0.6<a class="anchor" aria-label="anchor" href="#bigsnpr-06"></a></h2>
<ul><li>Keep up with {bigstatsr}.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.3.1" id="bigsnpr-031">bigsnpr 0.3.1<a class="anchor" aria-label="anchor" href="#bigsnpr-031"></a></h2>
<ul><li>Provide function <code>snp_gene</code> (as a gist) to get genes corresponding to ‘rs’ SNP IDs thanks to package {rsnps} from rOpenSci. See README.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.3.0" id="bigsnpr-030">bigsnpr 0.3.0<a class="anchor" aria-label="anchor" href="#bigsnpr-030"></a></h2>
<ul><li>
<strong>Package {bigsnpr} is published in <a href="http://dx.doi.org/10.1093/bioinformatics/bty185" class="external-link">Bioinformatics</a></strong>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.7" id="bigsnpr-027">bigsnpr 0.2.7<a class="anchor" aria-label="anchor" href="#bigsnpr-027"></a></h2>
<ul><li>Faster defaults + possibility to estimate correlations based on a subset of individuals for <code>snp_fastImpute</code>. Also store information in an FBM (instead of a data frame) so that imputation can be done by parts (you can stop the imputation by killing the R processes and come back to it later). Note that the defaults used in the <em>Bioinformatics</em> paper were <code>alpha = 0.02</code> and <code>size = 500</code> (instead of <code>1e-4</code> and <code>200</code> now, respectively). These new defaults are more stringent on the SNPs that are used, which makes the imputation faster (30 min instead of 42-48 min), without impacting accuracy (still 4.7-4.8% of errors).</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.5" id="bigsnpr-025">bigsnpr 0.2.5<a class="anchor" aria-label="anchor" href="#bigsnpr-025"></a></h2>
<ul><li>
<strong>This package won’t be on CRAN</strong>.</li></ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.4" id="bigsnpr-024">bigsnpr 0.2.4<small>2018-02-01</small><a class="anchor" aria-label="anchor" href="#bigsnpr-024"></a></h2>
<ul><li>No longer download PLINK automatically (because it is a CRAN policy violation).</li></ul></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

