<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • bigsnpr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Changelog" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small></small></h1>
      
    </div>

    <div id="bigsnpr-1-0-0" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-1-0-0" class="anchor"></a>bigsnpr 1.0.0</h2>
<ul>
<li>Add memory-mapping on PLINK (.bed) files with missing values + new functions:
<ul>
<li><code><a href="../reference/bed-class.html">bed()</a></code></li>
<li><code><a href="../reference/bed_MAF.html">bed_MAF()</a></code></li>
<li><code><a href="../reference/snp_autoSVD.html">bed_autoSVD()</a></code></li>
<li><code><a href="../reference/snp_clumping.html">bed_clumping()</a></code></li>
<li><code><a href="../reference/bed_counts.html">bed_counts()</a></code></li>
<li><code><a href="../reference/bed_cprodVec.html">bed_cprodVec()</a></code></li>
<li><code><a href="../reference/snp_pcadapt.html">bed_pcadapt()</a></code></li>
<li><code><a href="../reference/bed_prodVec.html">bed_prodVec()</a></code></li>
<li><code><a href="../reference/bed_projectPCA.html">bed_projectPCA()</a></code></li>
<li><code><a href="../reference/bed_projectSelfPCA.html">bed_projectSelfPCA()</a></code></li>
<li><code><a href="../reference/bed_randomSVD.html">bed_randomSVD()</a></code></li>
<li><code><a href="../reference/bed_scaleBinom.html">bed_scaleBinom()</a></code></li>
<li><code><a href="../reference/bed_tcrossprodSelf.html">bed_tcrossprodSelf()</a></code></li>
<li><code><a href="../reference/download_1000G.html">download_1000G()</a></code></li>
<li><code><a href="../reference/snp_modifyBuild.html">snp_modifyBuild()</a></code></li>
<li><code><a href="../reference/snp_plinkKINGQC.html">snp_plinkKINGQC()</a></code></li>
<li><code><a href="../reference/snp_readBed.html">snp_readBed2()</a></code></li>
<li><code><a href="../reference/sub_bed.html">sub_bed()</a></code></li>
</ul>
</li>
<li><p>Add 3 parameters to <code>autoSVD()</code>: <code>alpha.tukey</code>, <code>min.mac</code> and <code>max.iter</code>.</p></li>
<li><p>Remove option for changing ploidy (that was only partially supported).</p></li>
<li><p>Automatically apply <code><a href="../reference/snp_gc.html">snp_gc()</a></code> to <code>pcadapt</code>.</p></li>
</ul>
</div>
    <div id="bigsnpr-0-12-0" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-12-0" class="anchor"></a>bigsnpr 0.12.0</h2>
<ul>
<li>Add <code><a href="../reference/snp_fastImputeSimple.html">snp_fastImputeSimple()</a></code>: fast imputation via mode, mean or sampling according to allele frequencies.</li>
</ul>
</div>
    <div id="bigsnpr-0-11-3" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-11-3" class="anchor"></a>bigsnpr 0.11.3</h2>
<ul>
<li>Fix a bug in <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> that could not handle duplicated variants or individuals.</li>
</ul>
</div>
    <div id="bigsnpr-0-11-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-11-1" class="anchor"></a>bigsnpr 0.11.1</h2>
<ul>
<li>When using <code><a href="../reference/SCT.html">snp_grid_PRS()</a></code>, it now stores not only the FBM, but also the input parameters as attributes (the whole result basically).</li>
</ul>
</div>
    <div id="bigsnpr-0-11-0" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-11-0" class="anchor"></a>bigsnpr 0.11.0</h2>
<ul>
<li><p>Add 3 SCT functions <code>snp_grid_*()</code> to improve from Clumping and Thresholding (preprint coming soon).</p></li>
<li><p>Add <code><a href="../reference/snp_match.html">snp_match()</a></code> function to match between summary statistics and some SNP information.</p></li>
</ul>
</div>
    <div id="bigsnpr-0-10-2" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-10-2" class="anchor"></a>bigsnpr 0.10.2</h2>
<ul>
<li>Parameter <code>is.size.in.bp</code> is deprecated.</li>
</ul>
</div>
    <div id="bigsnpr-0-10-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-10-1" class="anchor"></a>bigsnpr 0.10.1</h2>
<ul>
<li>Add parameter <code>read_as</code> for <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code>. It is now possible to sample BGEN probabilities as random hard calls using <code>read_as = "random"</code>. Default remains reading probabilities as dosages.</li>
</ul>
</div>
    <div id="bigsnpr-0-10-0" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-10-0" class="anchor"></a>bigsnpr 0.10.0</h2>
<ul>
<li><p>For memory-mapping, now use <em>mio</em> instead of <em>boost</em>.</p></li>
<li><p><code><a href="../reference/snp_clumping.html">snp_clumping()</a></code> (and <code><a href="../reference/snp_autoSVD.html">snp_autoSVD()</a></code>) now has a <code>size</code> that is inversely proportional to <code>thr.r2</code>.</p></li>
<li><p><code><a href="../reference/snp_clumping.html">snp_pruning()</a></code> is deprecated (and will be removed someday); now always use <code><a href="../reference/snp_clumping.html">snp_clumping()</a></code>.</p></li>
</ul>
</div>
    <div id="bigsnpr-0-9-0" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-9-0" class="anchor"></a>bigsnpr 0.9.0</h2>
<ul>
<li>When reading bed files, switch reading of Os and 2s to be consistent with other software.</li>
</ul>
</div>
    <div id="bigsnpr-0-8-2" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-8-2" class="anchor"></a>bigsnpr 0.8.2</h2>
<ul>
<li>Add function <code><a href="../reference/snp_assocBGEN.html">snp_assocBGEN()</a></code> for computing quick association tests from BGEN files. Could be useful for quick screening of useful SNPs to read in bigSNP format. This function might be improved in the future.</li>
</ul>
</div>
    <div id="bigsnpr-0-8-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-8-1" class="anchor"></a>bigsnpr 0.8.1</h2>
<ul>
<li>Change url to download PLINK 1.9.</li>
</ul>
</div>
    <div id="bigsnpr-0-8-0" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-8-0" class="anchor"></a>bigsnpr 0.8.0</h2>
<ul>
<li>Add function <code><a href="../reference/snp_readBGEN.html">snp_readBGEN()</a></code> to read UK Biobank BGEN files in <code>bigSNP</code> format.</li>
</ul>
</div>
    <div id="bigsnpr-0-7-0" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-7-0" class="anchor"></a>bigsnpr 0.7.0</h2>
<ul>
<li><p>Add parameter <code>is.size.in.bp</code> to <code><a href="../reference/snp_autoSVD.html">snp_autoSVD()</a></code> for the clumping part.</p></li>
<li><p>Change the threshold of outlier detection in <code><a href="../reference/snp_autoSVD.html">snp_autoSVD()</a></code> (it now detects less outliers). See the documentation details if you don’t have any information about SNPs.</p></li>
</ul>
</div>
    <div id="bigsnpr-0-6" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-6" class="anchor"></a>bigsnpr 0.6</h2>
<ul>
<li>Keep up with {bigstatsr}.</li>
</ul>
</div>
    <div id="bigsnpr-0-3-1" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-3-1" class="anchor"></a>bigsnpr 0.3.1</h2>
<ul>
<li>Provide function <code>snp_gene</code> (as a gist) to get genes corresponding to ‘rs’ SNP IDs thanks to package {rsnps} from rOpenSci. See README.</li>
</ul>
</div>
    <div id="bigsnpr-0-3-0" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-3-0" class="anchor"></a>bigsnpr 0.3.0</h2>
<ul>
<li>
<strong>Package {bigsnpr} is published in <a href="http://dx.doi.org/10.1093/bioinformatics/bty185">Bioinformatics</a></strong>.</li>
</ul>
</div>
    <div id="bigsnpr-0-2-7" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-2-7" class="anchor"></a>bigsnpr 0.2.7</h2>
<ul>
<li>Faster defaults + possibility to estimate correlations based on a subset of individuals for <code>snp_fastImpute</code>. Also store information in an FBM (instead of a data frame) so that imputation can be done by parts (you can stop the imputation by killing the R processes and come back to it later). Note that the defaults used in the <em>Bioinformatics</em> paper were <code>alpha = 0.02</code> and <code>size = 500</code> (instead of <code>1e-4</code> and <code>200</code> now, respectively). These new defaults are more stringent on the SNPs that are used, which makes the imputation faster (30 min instead of 42-48 min), without impacting accuracy (still 4.7-4.8% of errors).</li>
</ul>
</div>
    <div id="bigsnpr-0-2-5" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-2-5" class="anchor"></a>bigsnpr 0.2.5</h2>
<ul>
<li>
<strong>This package won’t be on CRAN</strong>.</li>
</ul>
</div>
    <div id="bigsnpr-0-2-4" class="section level2">
<h2 class="hasAnchor">
<a href="#bigsnpr-0-2-4" class="anchor"></a>bigsnpr 0.2.4</h2>
<ul>
<li>No longer download PLINK automatically (because it is a CRAN policy violation).</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bigsnpr-1-0-0">1.0.0</a></li>
        <li><a href="#bigsnpr-0-12-0">0.12.0</a></li>
        <li><a href="#bigsnpr-0-11-3">0.11.3</a></li>
        <li><a href="#bigsnpr-0-11-1">0.11.1</a></li>
        <li><a href="#bigsnpr-0-11-0">0.11.0</a></li>
        <li><a href="#bigsnpr-0-10-2">0.10.2</a></li>
        <li><a href="#bigsnpr-0-10-1">0.10.1</a></li>
        <li><a href="#bigsnpr-0-10-0">0.10.0</a></li>
        <li><a href="#bigsnpr-0-9-0">0.9.0</a></li>
        <li><a href="#bigsnpr-0-8-2">0.8.2</a></li>
        <li><a href="#bigsnpr-0-8-1">0.8.1</a></li>
        <li><a href="#bigsnpr-0-8-0">0.8.0</a></li>
        <li><a href="#bigsnpr-0-7-0">0.7.0</a></li>
        <li><a href="#bigsnpr-0-6">0.6</a></li>
        <li><a href="#bigsnpr-0-3-1">0.3.1</a></li>
        <li><a href="#bigsnpr-0-3-0">0.3.0</a></li>
        <li><a href="#bigsnpr-0-2-7">0.2.7</a></li>
        <li><a href="#bigsnpr-0-2-5">0.2.5</a></li>
        <li><a href="#bigsnpr-0-2-4">0.2.4</a></li>
      </ul>
    </div>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


