<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of Massive SNP Arrays • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Analysis of Massive SNP Arrays">
<meta property="og:description" content="Easy-to-use, efficient, flexible and scalable tools for analyzing 
    massive SNP arrays. Privé et al. (2018) &lt;doi:10.1093/bioinformatics/bty185&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div class="section level1">
<div class="page-header"><h1 id="bigsnpr">bigsnpr<a class="anchor" aria-label="anchor" href="#bigsnpr"></a>
</h1></div>
<p>{bigsnpr} is an R package for the analysis of massive SNP arrays, primarily designed for human genetics. It enhances the features of <a href="https://privefl.github.io/bigstatsr/" class="external-link">package {bigstatsr}</a> for the purpose of analyzing genotype data.</p>
<p>To get you started:</p>
<ul>
<li><p><a href="https://privefl.github.io/bigsnpr/articles/demo.html" class="external-link">Quick demo</a></p></li>
<li><p>List of functions <a href="https://privefl.github.io/bigsnpr/reference/index.html" class="external-link">from bigsnpr</a> and <a href="https://privefl.github.io/bigstatsr/reference/index.html" class="external-link">from bigstatsr</a></p></li>
<li><p><a href="https://privefl.github.io/bigsnpr-extdoc/" class="external-link">Extended documentation with more examples</a> + <a href="https://youtu.be/7VxBT5A_AcA" class="external-link">course recording</a></p></li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>In R, run</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"privefl/bigsnpr"</span><span class="op">)</span></span></code></pre></div>
<p>or for the CRAN version</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"bigsnpr"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="input-formats">Input formats<a class="anchor" aria-label="anchor" href="#input-formats"></a>
</h2>
<p>This package reads <em>bed</em>/<em>bim</em>/<em>fam</em> files (PLINK preferred format) using functions <code><a href="reference/snp_readBed.html">snp_readBed()</a></code> and <code><a href="reference/snp_readBed.html">snp_readBed2()</a></code>. Before reading into this package’s special format, quality control and conversion can be done using PLINK, which can be called directly from R using <code><a href="reference/snp_plinkQC.html">snp_plinkQC()</a></code> and <code><a href="reference/snp_plinkKINGQC.html">snp_plinkKINGQC()</a></code>.</p>
<p>This package can also read <strong>UK Biobank BGEN files</strong> using function <code><a href="reference/snp_readBGEN.html">snp_readBGEN()</a></code>. This function takes around 40 minutes to read 1M variants for 400K individuals using 15 cores.</p>
<p>This package uses a class called <code>bigSNP</code> for representing SNP data. A <code>bigSNP</code> object is a list with some elements:</p>
<ul>
<li>
<code>$genotypes</code>: A <a href="https://privefl.github.io/bigstatsr/reference/FBM.code256-class.html" class="external-link"><code>FBM.code256</code></a>. Rows are samples and columns are variants. This stores genotype calls or <strong>dosages</strong> (rounded to 2 decimal places).</li>
<li>
<code>$fam</code>: A <code>data.frame</code> with some information on the individuals.</li>
<li>
<code>$map</code>: A <code>data.frame</code> with some information on the variants.</li>
</ul>
<p><strong>Note that most of the algorithms of this package don’t handle missing values.</strong> You can use <code><a href="reference/snp_fastImpute.html">snp_fastImpute()</a></code> (taking a few hours for a chip of 15K x 300K) and <code><a href="reference/snp_fastImputeSimple.html">snp_fastImputeSimple()</a></code> (taking a few minutes only) to impute missing values of <em>genotyped</em> variants.</p>
<p>Package {bigsnpr} also provides functions that directly work on bed files with a few missing values (the <code>bed_*()</code> functions). See paper <a href="https://doi.org/10.1093/bioinformatics/btaa520" class="external-link">“Efficient toolkit implementing..”</a>.</p>
</div>
<div class="section level2">
<h2 id="polygenic-scores">Polygenic scores<a class="anchor" aria-label="anchor" href="#polygenic-scores"></a>
</h2>
<p>Polygenic scores are one of the main focus of this package. There are 3 main methods currently available:</p>
<ul>
<li><p>Penalized regressions with individual-level data (see <a href="https://doi.org/10.1534/genetics.119.302019" class="external-link">paper</a> and <a href="https://privefl.github.io/bigstatsr/articles/penalized-regressions.html" class="external-link">tutorial</a>)</p></li>
<li><p>Clumping and Thresholding (C+T) and Stacked C+T (SCT) with summary statistics and individual level data (see <a href="https://doi.org/10.1016/j.ajhg.2019.11.001" class="external-link">paper</a> and <a href="https://privefl.github.io/bigsnpr/articles/SCT.html" class="external-link">tutorial</a>).</p></li>
<li><p>LDpred2 with summary statistics (see <a href="https://doi.org/10.1093/bioinformatics/btaa1029" class="external-link">paper</a> and <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html" class="external-link">tutorial</a>), and lassosum2</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="possible-upcoming-features">Possible upcoming features<a class="anchor" aria-label="anchor" href="#possible-upcoming-features"></a>
</h2>
<ul>
<li><p>Multiple imputation for GWAS (<a href="https://doi.org/10.1371/journal.pgen.1006091" class="external-link uri">https://doi.org/10.1371/journal.pgen.1006091</a>).</p></li>
<li><p>More interactive (visual) QC.</p></li>
</ul>
<p>You can request some feature by opening an issue.</p>
</div>
<div class="section level2">
<h2 id="bug-report--support">Bug report / Support<a class="anchor" aria-label="anchor" href="#bug-report--support"></a>
</h2>
<p><a href="https://stackoverflow.com/q/5963269/6103040" class="external-link">How to make a great R reproducible example?</a></p>
<p>Please open an issue if you find a bug.</p>
<p>If you want help using {bigstatsr} (the <code>big_*()</code> functions), please open an issue on <a href="https://github.com/privefl/bigstatsr/issues" class="external-link">{bigstatsr}’s repo</a>, or post on Stack Overflow with the tag <em>bigstatsr</em>.</p>
<p>I will always redirect you to GitHub issues if you email me, so that others can benefit from our discussion.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li><p>Privé, Florian, et al. <a href="https://doi.org/10.1093/bioinformatics/bty185" class="external-link">“Efficient analysis of large-scale genome-wide data with two R packages: bigstatsr and bigsnpr.”</a> <em>Bioinformatics</em> 34.16 (2018): 2781-2787.</p></li>
<li><p>Privé, Florian, et al. <a href="https://doi.org/10.1534/genetics.119.302019" class="external-link">“Efficient implementation of penalized regression for genetic risk prediction.”</a> <em>Genetics</em> 212.1 (2019): 65-74.</p></li>
<li><p>Privé, Florian, et al. <a href="https://doi.org/10.1016/j.ajhg.2019.11.001" class="external-link">“Making the most of Clumping and Thresholding for polygenic scores.”</a> <em>The American Journal of Human Genetics</em> 105.6 (2019): 1213-1221.</p></li>
<li><p>Privé, Florian, et al. <a href="https://doi.org/10.1093/bioinformatics/btaa520" class="external-link">“Efficient toolkit implementing best practices for principal component analysis of population genetic data.”</a> <em>Bioinformatics</em> 36.16 (2020): 4449-4457.</p></li>
<li><p>Privé, Florian, et al. <a href="https://doi.org/10.1093/bioinformatics/btaa1029" class="external-link">“LDpred2: better, faster, stronger.”</a> <em>Bioinformatics</em> 36.22-23 (2020): 5424-5431.</p></li>
<li><p>Privé, Florian. <a href="https://doi.org/10.1093/bioinformatics/btab519" class="external-link">“Optimal linkage disequilibrium splitting.”</a> <em>Bioinformatics</em> 38.1 (2022): 255–256.</p></li>
<li><p>Privé, Florian. <a href="https://doi.org/10.1093/bioinformatics/btac348" class="external-link">“Using the UK Biobank as a global reference of worldwide populations: application to measuring ancestry diversity from GWAS summary statistics.”</a> <em>Bioinformatics</em> 38.13 (2022): 3477-3480.</p></li>
<li><p>Privé, Florian, et al. <a href="https://doi.org/10.1016/j.xhgg.2022.100136" class="external-link">“Identifying and correcting for misspecifications in GWAS summary statistics and polygenic scores.”</a> <em>Human Genetics and Genomics Advances</em> 3.4 (2022).</p></li>
<li><p>Privé, Florian, et al. <a href="https://doi.org/10.1101/2022.10.10.511629" class="external-link">Inferring disease architecture and predictive ability with LDpred2-auto</a>. <em>The American Journal of Human Genetics</em> 110.12 (2023): 2042-2055.</p></li>
</ul>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=bigsnpr" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/privefl/bigsnpr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/privefl/bigsnpr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing bigsnpr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Florian Privé <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/privefl/bigsnpr/actions" class="external-link"><img src="https://github.com/privefl/bigsnpr/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://app.codecov.io/gh/privefl/bigsnpr?branch=master" class="external-link"><img src="https://codecov.io/gh/privefl/bigsnpr/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://CRAN.R-project.org/package=bigsnpr" class="external-link"><img src="https://www.r-pkg.org/badges/version/bigsnpr" alt="CRAN status"></a></li>
<li><a href="http://dx.doi.org/10.1093/bioinformatics/bty185" class="external-link"><img src="https://zenodo.org/badge/doi/10.1093/bioinformatics/bty185.svg" alt="DOI"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
