<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Demonstration of packages bigsnpr and bigstatsr • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Demonstration of packages bigsnpr and bigstatsr">
<meta property="og:description" content="bigsnpr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.20</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/PLR.html">Polygenic scores using individual-level data</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
    <li>
      <a href="https://privefl.github.io/bigsnpr-extdoc/" class="external-link">Extended documentation for bigsnpr &amp; bigstatsr</a>
    </li>
    <li>
      <a href="https://privefl.github.io/statgen-course/" class="external-link">Statistical Human Genetics course using R</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Demonstration of packages bigsnpr and
bigstatsr</h1>
                        <h4 data-toc-skip class="author">Florian
Privé</h4>
            
            <h4 data-toc-skip class="date">June 28, 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/vignettes/demo.Rmd" class="external-link"><code>vignettes/demo.Rmd</code></a></small>
      <div class="hidden name"><code>demo.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="get-genotype-file">Get genotype file<a class="anchor" aria-label="anchor" href="#get-genotype-file"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the example bedfile from package bigsnpr</span></span>
<span><span class="va">bedfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example.bed"</span>, package <span class="op">=</span> <span class="st">"bigsnpr"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="read-from-the-plink-files">Read from the PLINK files<a class="anchor" aria-label="anchor" href="#read-from-the-plink-files"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages bigsnpr and bigstatsr</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: bigstatsr</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read from bed/bim/fam, it will create new files.</span></span>
<span><span class="co"># Let's put them in an temporary directory for this demo.</span></span>
<span><span class="va">tmpfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/snp_readBed.html">snp_readBed</a></span><span class="op">(</span><span class="va">bedfile</span>, backingfile <span class="op">=</span> <span class="va">tmpfile</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpcpaDGH\\file6d9c7af67e82.rds"</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attach the "bigSNP" object in R session</span></span>
<span><span class="va">obj.bigSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">tmpfile</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># See how it looks like</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">obj.bigSNP</span>, max.level <span class="op">=</span> <span class="fl">2</span>, strict.width <span class="op">=</span> <span class="st">"cut"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 3</span></span>
<span><span class="co">##  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"..</span></span>
<span><span class="co">##   ..and 26 methods, of which 12 are  possibly relevant:</span></span>
<span><span class="co">##   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,</span></span>
<span><span class="co">##   ..  check_write_permissions, copy#envRefClass, initialize,</span></span>
<span><span class="co">##   ..  initialize#FBM, save, show#envRefClass, show#FBM</span></span>
<span><span class="co">##  $ fam      :'data.frame':   517 obs. of  6 variables:</span></span>
<span><span class="co">##   ..$ family.ID  : chr [1:517] "POP1" "POP1" "POP1" "POP1" ...</span></span>
<span><span class="co">##   ..$ sample.ID  : chr [1:517] "IND0" "IND1" "IND2" "IND3" ...</span></span>
<span><span class="co">##   ..$ paternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ maternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ sex        : int [1:517] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ affection  : int [1:517] 1 1 2 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##  $ map      :'data.frame':   4542 obs. of  6 variables:</span></span>
<span><span class="co">##   ..$ chromosome  : int [1:4542] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##   ..$ marker.ID   : chr [1:4542] "SNP0" "SNP1" "SNP2" "SNP3" ...</span></span>
<span><span class="co">##   ..$ genetic.dist: int [1:4542] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ physical.pos: int [1:4542] 112 1098 2089 3107 4091 5091 6..</span></span>
<span><span class="co">##   ..$ allele1     : chr [1:4542] "A" "T" "T" "T" ...</span></span>
<span><span class="co">##   ..$ allele2     : chr [1:4542] "T" "A" "A" "A" ...</span></span>
<span><span class="co">##  - attr(*, "class")= chr "bigSNP"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get aliases for useful slots</span></span>
<span><span class="va">G</span>   <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">genotypes</span></span>
<span><span class="va">CHR</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">chromosome</span></span>
<span><span class="va">POS</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">physical.pos</span></span>
<span><span class="co"># Check some counts for the 10 first SNPs</span></span>
<span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_counts.html" class="external-link">big_counts</a></span><span class="op">(</span><span class="va">G</span>, ind.col <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span></span>
<span><span class="co">## 0     226  325  309  346  158  215  395  134  143   159</span></span>
<span><span class="co">## 1     228  171  171  151  246  236  112  269  250   235</span></span>
<span><span class="co">## 2      63   21   37   20  113   66   10  114  124   123</span></span>
<span><span class="co">## &lt;NA&gt;    0    0    0    0    0    0    0    0    0     0</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="principal-component-analysis">Principal Component Analysis<a class="anchor" aria-label="anchor" href="#principal-component-analysis"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Half of the cores you have on your computer</span></span>
<span><span class="va">NCORES</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/nb_cores.html" class="external-link">nb_cores</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Exclude Long-Range Linkage Disequilibrium Regions of the human genome</span></span>
<span><span class="co"># based on an online table. </span></span>
<span><span class="va">ind.excl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_clumping.html">snp_indLRLDR</a></span><span class="op">(</span>infos.chr <span class="op">=</span> <span class="va">CHR</span>, infos.pos <span class="op">=</span> <span class="va">POS</span><span class="op">)</span></span>
<span><span class="co"># Use clumping (on the MAF) to keep SNPs weakly correlated with each other.</span></span>
<span><span class="co"># See https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html</span></span>
<span><span class="co"># to know why I prefer using clumping than standard pruning.</span></span>
<span><span class="va">ind.keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_clumping.html">snp_clumping</a></span><span class="op">(</span><span class="va">G</span>, infos.chr <span class="op">=</span> <span class="va">CHR</span>,</span>
<span>                         exclude <span class="op">=</span> <span class="va">ind.excl</span>,</span>
<span>                         ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span>
<span><span class="co"># Get the first 10 PCs, corresponding to pruned SNPs</span></span>
<span><span class="va">obj.svd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_randomSVD.html" class="external-link">big_randomSVD</a></span><span class="op">(</span><span class="va">G</span>, fun.scaling <span class="op">=</span> <span class="fu"><a href="../reference/snp_scaleBinom.html">snp_scaleBinom</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                         ind.col <span class="op">=</span> <span class="va">ind.keep</span>,</span>
<span>                         ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span>
<span><span class="co"># As `obj.svd` has a class and a method `plot`.</span></span>
<span><span class="co"># Scree plot by default</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">obj.svd</span><span class="op">)</span></span></code></pre></div>
<p><img src="demo_files/figure-html/unnamed-chunk-3-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Score (PCs) plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">obj.svd</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span></span></code></pre></div>
<p><img src="demo_files/figure-html/unnamed-chunk-3-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># As plot returns an ggplot2 object, you can easily modify it.</span></span>
<span><span class="co"># For example, you can add colors based on the population.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: package 'ggplot2' was built under R version 4.2.3</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">obj.svd</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">pop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"POP1"</span>, <span class="st">"POP2"</span>, <span class="st">"POP3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">143</span>, <span class="fl">167</span>, <span class="fl">207</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Population"</span><span class="op">)</span></span></code></pre></div>
<p><img src="demo_files/figure-html/unnamed-chunk-3-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="genome-wide-association-study">Genome-Wide Association Study<a class="anchor" aria-label="anchor" href="#genome-wide-association-study"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Fit a logistic model between the phenotype and each SNP separately</span></span>
<span><span class="co"># while adding PCs as covariates to each model</span></span>
<span><span class="va">y01</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">fam</span><span class="op">$</span><span class="va">affection</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">obj.gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_univLogReg.html" class="external-link">big_univLogReg</a></span><span class="op">(</span><span class="va">G</span>, y01.train <span class="op">=</span> <span class="va">y01</span>,</span>
<span>                           covar.train <span class="op">=</span> <span class="va">obj.svd</span><span class="op">$</span><span class="va">u</span>,</span>
<span>                           ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span>
<span><span class="co"># Q-Q plot of the object</span></span>
<span><span class="fu"><a href="../reference/snp_qq.html">snp_qq</a></span><span class="op">(</span><span class="va">obj.gwas</span><span class="op">)</span></span></code></pre></div>
<p><img src="demo_files/figure-html/unnamed-chunk-4-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You can easily apply genomic control to this object</span></span>
<span><span class="va">obj.gwas.gc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_gc.html">snp_gc</a></span><span class="op">(</span><span class="va">obj.gwas</span><span class="op">)</span></span>
<span><span class="co"># Redo the Q-Q plot</span></span>
<span><span class="fu"><a href="../reference/snp_qq.html">snp_qq</a></span><span class="op">(</span><span class="va">obj.gwas.gc</span><span class="op">)</span></span></code></pre></div>
<p><img src="demo_files/figure-html/unnamed-chunk-4-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Manhattan plot, not quite sexy because there are only 1 chromosome here</span></span>
<span><span class="fu"><a href="../reference/snp_manhattan.html">snp_manhattan</a></span><span class="op">(</span><span class="va">obj.gwas.gc</span>, infos.chr <span class="op">=</span> <span class="va">CHR</span>, infos.pos <span class="op">=</span> <span class="va">POS</span><span class="op">)</span></span></code></pre></div>
<p><img src="demo_files/figure-html/unnamed-chunk-4-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="joint-prs">Joint PRS<a class="anchor" aria-label="anchor" href="#joint-prs"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Divide the indices in training/test sets</span></span>
<span><span class="va">ind.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span>, <span class="fl">400</span><span class="op">)</span></span>
<span><span class="va">ind.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/seq-dim.html" class="external-link">rows_along</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span>, <span class="va">ind.train</span><span class="op">)</span></span>
<span><span class="co"># Train the model</span></span>
<span><span class="va">cmsa.logit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_spLogReg.html" class="external-link">big_spLogReg</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">G</span>, y01.train <span class="op">=</span> <span class="va">y01</span><span class="op">[</span><span class="va">ind.train</span><span class="op">]</span>, </span>
<span>                           ind.train <span class="op">=</span> <span class="va">ind.train</span>, </span>
<span>                           covar.train <span class="op">=</span> <span class="va">obj.svd</span><span class="op">$</span><span class="va">u</span><span class="op">[</span><span class="va">ind.train</span>, <span class="op">]</span>,</span>
<span>                           alphas <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.5</span>, <span class="fl">0.05</span>, <span class="fl">0.001</span><span class="op">)</span>,</span>
<span>                           ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span>
<span><span class="co"># Get predictions for the test set</span></span>
<span><span class="va">preds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">cmsa.logit</span>, X <span class="op">=</span> <span class="va">G</span>, ind.row <span class="op">=</span> <span class="va">ind.test</span>, </span>
<span>                 covar.row <span class="op">=</span> <span class="va">obj.svd</span><span class="op">$</span><span class="va">u</span><span class="op">[</span><span class="va">ind.test</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Check AUC</span></span>
<span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/AUC.html" class="external-link">AUC</a></span><span class="op">(</span><span class="va">preds</span>, <span class="va">y01</span><span class="op">[</span><span class="va">ind.test</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.837013</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
