<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Exercises using packages bigsnpr and bigstatsr • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Exercises using packages bigsnpr and bigstatsr">
<meta property="og:description" content="bigsnpr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Exercises using packages bigsnpr and
bigstatsr</h1>
                        <h4 data-toc-skip class="author">Florian
Privé</h4>
            
            <h4 data-toc-skip class="date">April 18, 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/vignettes/exo.Rmd" class="external-link"><code>vignettes/exo.Rmd</code></a></small>
      <div class="hidden name"><code>exo.Rmd</code></div>

    </div>

    
    
<p>In this document, I show how to use some of the features of packages
{bigsnpr} and {bigstatsr}. Note that many functions used here come from
package {bigstatsr} and could therefore be used on other data encoded as
matrix-like (outside of the field of genotype data).</p>
<div class="section level2">
<h2 id="get-data">Get data<a class="anchor" aria-label="anchor" href="#get-data"></a>
</h2>
<p>Download <a href="https://github.com/privefl/bigsnpr/raw/master/data-raw/public-data.zip" class="external-link">data</a>
and unzip files. I store those files in a directory called
<code>"tmp-data"</code> here.</p>
<p>You can see <a href="https://github.com/privefl/bigsnpr/blob/master/data-raw/public-data.R" class="external-link">there</a>
how I generated these data from <a href="https://www.nature.com/articles/nature15393" class="external-link">the 1000 Genomes
project</a>.</p>
</div>
<div class="section level2">
<h2 id="read-from-the-plink-files">Read from the PLINK files<a class="anchor" aria-label="anchor" href="#read-from-the-plink-files"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages bigsnpr and bigstatsr</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: bigstatsr</span></span></code></pre>
<pre><code><span><span class="co">## Warning: package 'bigstatsr' was built under R version 4.2.3</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read from bed/bim/fam, it will create new files.</span></span>
<span><span class="fu"><a href="../reference/snp_readBed.html">snp_readBed</a></span><span class="op">(</span><span class="st">"tmp-data/public-data.bed"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "C:\\Users\\au639593\\Desktop\\bigsnpr\\tmp-data\\public-data.rds"</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attach the "bigSNP" object in R session</span></span>
<span><span class="va">obj.bigSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="st">"tmp-data/public-data.rds"</span><span class="op">)</span></span>
<span><span class="co"># See how it looks like</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">obj.bigSNP</span>, max.level <span class="op">=</span> <span class="fl">2</span>, strict.width <span class="op">=</span> <span class="st">"cut"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 3</span></span>
<span><span class="co">##  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 16 ..</span></span>
<span><span class="co">##   ..and 26 methods, of which 12 are  possibly relevant:</span></span>
<span><span class="co">##   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,</span></span>
<span><span class="co">##   ..  check_write_permissions, copy#envRefClass, initialize,</span></span>
<span><span class="co">##   ..  initialize#FBM, save, show#envRefClass, show#FBM</span></span>
<span><span class="co">##  $ fam      :'data.frame':   559 obs. of  6 variables:</span></span>
<span><span class="co">##   ..$ family.ID  : chr [1:559] "EUR_GBR" "EUR_GBR" "EUR_GBR" "EUR_GBR" ...</span></span>
<span><span class="co">##   ..$ sample.ID  : chr [1:559] "HG00096" "HG00097" "HG00099" "HG00100" ...</span></span>
<span><span class="co">##   ..$ paternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ maternal.ID: int [1:559] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ sex        : int [1:559] 1 2 2 2 1 2 1 2 2 1 ...</span></span>
<span><span class="co">##   ..$ affection  : int [1:559] 1 2 1 1 1 1 2 1 2 1 ...</span></span>
<span><span class="co">##  $ map      :'data.frame':   130816 obs. of  6 variables:</span></span>
<span><span class="co">##   ..$ chromosome  : int [1:130816] 2 2 2 2 2 2 2 2 2 2 ...</span></span>
<span><span class="co">##   ..$ marker.ID   : chr [1:130816] "rs13400442" "rs7594567" "rs7597758" "..</span></span>
<span><span class="co">##   ..$ genetic.dist: int [1:130816] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ physical.pos: int [1:130816] 18506 21833 22398 28228 32003 32005 36..</span></span>
<span><span class="co">##   ..$ allele1     : chr [1:130816] "C" "G" "T" "A" ...</span></span>
<span><span class="co">##   ..$ allele2     : chr [1:130816] "T" "C" "C" "G" ...</span></span>
<span><span class="co">##  - attr(*, "class")= chr "bigSNP"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get aliases for useful slots</span></span>
<span><span class="va">G</span>   <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">genotypes</span></span>
<span><span class="va">CHR</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">chromosome</span></span>
<span><span class="va">POS</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">physical.pos</span></span>
<span><span class="va">y</span>   <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">fam</span><span class="op">$</span><span class="va">affection</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">fam</span><span class="op">$</span><span class="va">sex</span></span>
<span><span class="va">pop</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">fam</span><span class="op">$</span><span class="va">family.ID</span></span>
<span><span class="va">NCORES</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/nb_cores.html" class="external-link">nb_cores</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Check some counts for the 10 first SNPs</span></span>
<span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_counts.html" class="external-link">big_counts</a></span><span class="op">(</span><span class="va">G</span>, ind.col <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span></span>
<span><span class="co">## 0     504  468  421  483  476  420  380  383  478   389</span></span>
<span><span class="co">## 1      48   69  107   66   73  124  136  135   70   133</span></span>
<span><span class="co">## 2       7   22   31   10   10   15   43   41   11    37</span></span>
<span><span class="co">## &lt;NA&gt;    0    0    0    0    0    0    0    0    0     0</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="what-you-must-do">What you must do<a class="anchor" aria-label="anchor" href="#what-you-must-do"></a>
</h2>
<p>You need to</p>
<ul>
<li><p>Explore the data</p></li>
<li><p>Assess population structure of the data</p></li>
<li><p>Find the variables associated with the disease status
(<code>y</code>)</p></li>
<li>
<p>Predict the disease status using the following indices as
training/test sets:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Divide the indices in training/test sets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ind.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span>, <span class="fl">400</span><span class="op">)</span></span>
<span><span class="va">ind.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/seq-dim.html" class="external-link">rows_along</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span>, <span class="va">ind.train</span><span class="op">)</span></span></code></pre></div>
</li>
</ul>
<p>For this, you can use whatever tools you want because the data is
quite small. In the following section, I give some (scalable) solutions
using using packages {bigstatsr} and {bigsnpr}.</p>
</div>
<div class="section level2">
<h2 id="solution-using-bigstatsr-and-bigsnpr">Solution using {bigstatsr} and {bigsnpr}<a class="anchor" aria-label="anchor" href="#solution-using-bigstatsr-and-bigsnpr"></a>
</h2>
<div class="section level3">
<h3 id="population-structure-principal-component-analysis">Population structure: Principal Component Analysis<a class="anchor" aria-label="anchor" href="#population-structure-principal-component-analysis"></a>
</h3>
<p>Let us compute first principal components of the scaled genotype
matrix:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute partial SVD (10 PCs by default) using random projections</span></span>
<span><span class="co"># big_scale() computes means and standard deviations for scaling</span></span>
<span><span class="va">svd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_randomSVD.html" class="external-link">big_randomSVD</a></span><span class="op">(</span><span class="va">G</span>, <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_scale.html" class="external-link">big_scale</a></span><span class="op">(</span><span class="op">)</span>, ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Scree plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">svd</span><span class="op">)</span></span></code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-6-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: package 'ggplot2' was built under R version 4.2.3</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Scores plot + color for population</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">svd</span>, type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">pop</span><span class="op">)</span></span></code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-6-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">svd</span>, type <span class="op">=</span> <span class="st">"scores"</span>, scores <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">pop</span><span class="op">)</span></span></code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-6-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Loadings (effects of each variable for each PC)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">svd</span>, type <span class="op">=</span> <span class="st">"loadings"</span>, loadings <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, coeff <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span></code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-7-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<p>Looking at the loadings, we can see that the PCA captures some
variation due to large correlation between variables. To learn more
about this possible pitfall, please look at <a href="https://privefl.github.io/bigsnpr/articles/how-to-PCA.html" class="external-link">this
vignette</a>.</p>
</div>
<div class="section level3">
<h3 id="association-genome-wide-association-study-gwas">Association: Genome-Wide Association Study (GWAS)<a class="anchor" aria-label="anchor" href="#association-genome-wide-association-study-gwas"></a>
</h3>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Association of each variable of `G` with `y` (adjusting for 10 PCs)</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_univLogReg.html" class="external-link">big_univLogReg</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">y</span>, covar.train <span class="op">=</span> <span class="va">svd</span><span class="op">$</span><span class="va">u</span>, ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span>
<span><span class="co"># Histogram of p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span></span></code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-8-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Q-Q plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gwas</span>, type <span class="op">=</span> <span class="st">"Q-Q"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">1</span>, <span class="cn">NA</span><span class="op">)</span>  <span class="co"># snp_qq(gwas) + xlim(1, NA)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 117734 rows containing missing values</span></span>
<span><span class="co">## (`geom_point()`).</span></span></code></pre>
<p><img src="exo_files/figure-html/unnamed-chunk-8-2.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Manhattan plot</span></span>
<span><span class="fu"><a href="../reference/snp_manhattan.html">snp_manhattan</a></span><span class="op">(</span><span class="va">gwas</span>, <span class="va">CHR</span>, <span class="va">POS</span>, npoints <span class="op">=</span> <span class="fl">20e3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">5e-8</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-8-3.png" width="692.708333333333" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="polygenic-risk-score-prs">Polygenic Risk Score (PRS)<a class="anchor" aria-label="anchor" href="#polygenic-risk-score-prs"></a>
</h3>
<div class="section level4">
<h4 id="with-clumping-and-thresholding-ct">with Clumping and Thresholding (C+T)<a class="anchor" aria-label="anchor" href="#with-clumping-and-thresholding-ct"></a>
</h4>
<p><span class="math display">\[\rm{PRS}_i = \sum_{\substack{j \in
S_\text{clumping} \\ p_j~&lt;~p_T}} \hat\beta_j \cdot
G_{i,j}~,\]</span></p>
<p>where <span class="math inline">\(\hat\beta_j\)</span> (<span class="math inline">\(p_j\)</span>) are the effect sizes (p-values)
estimated from the GWAS and <span class="math inline">\(G_{i,j}\)</span>
is the allele count (genotype) for individual <span class="math inline">\(i\)</span> and SNP <span class="math inline">\(j\)</span>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumstats</span> <span class="op">&lt;-</span> <span class="fu">bigreadr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bigreadr/man/fread2.html" class="external-link">fread2</a></span><span class="op">(</span><span class="st">"tmp-data/public-data-sumstats.txt"</span><span class="op">)</span></span>
<span><span class="va">lpval</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">sumstats</span><span class="op">$</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="va">ind.keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_clumping.html">snp_clumping</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">CHR</span>, ind.row <span class="op">=</span> <span class="va">ind.train</span>, S <span class="op">=</span> <span class="va">lpval</span>, infos.pos <span class="op">=</span> <span class="va">POS</span>, ncores <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">THR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seq_log.html">seq_log</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">8</span>, length.out <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">prs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_PRS.html">snp_PRS</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">sumstats</span><span class="op">$</span><span class="va">beta</span><span class="op">[</span><span class="va">ind.keep</span><span class="op">]</span>, ind.keep <span class="op">=</span> <span class="va">ind.keep</span>, </span>
<span>               lpS.keep <span class="op">=</span> <span class="va">lpval</span><span class="op">[</span><span class="va">ind.keep</span><span class="op">]</span>, thr.list <span class="op">=</span> <span class="va">THR</span><span class="op">)</span></span>
<span><span class="co"># Learn the optimal threshold on the training set</span></span>
<span><span class="va">aucs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">prs</span><span class="op">[</span><span class="va">ind.train</span>, <span class="op">]</span>, <span class="fl">2</span>, <span class="va">AUC</span>, target <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="va">ind.train</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">THR</span>, <span class="va">aucs</span>, xlab <span class="op">=</span> <span class="st">"-log10(p-value)"</span>, ylab <span class="op">=</span> <span class="st">"AUC"</span>, pch <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="exo_files/figure-html/unnamed-chunk-9-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Evaluate on the test set</span></span>
<span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/AUC.html" class="external-link">AUC</a></span><span class="op">(</span><span class="va">prs</span><span class="op">[</span><span class="va">ind.test</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">aucs</span><span class="op">)</span><span class="op">]</span>, <span class="va">y</span><span class="op">[</span><span class="va">ind.test</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.6689111</span></span></code></pre>
<p>(TODO: ADD SCT)</p>
</div>
<div class="section level4">
<h4 id="with-penalized-logistic-regression-plr">with Penalized Logistic Regression (PLR)<a class="anchor" aria-label="anchor" href="#with-penalized-logistic-regression-plr"></a>
</h4>
<p><span class="math display">\[\arg\!\min_{\beta_0,~\beta}(\lambda,
\alpha)\left\{ \underbrace{ -\sum_{i=1}^n \left( y_i
\log\left(p_i\right) + (1 - y_i) \log\left(1 - p_i\right) \right)
}_\text{Loss function} + \underbrace{ \lambda
\left((1-\alpha)\frac{1}{2}\|\beta\|_2^2 + \alpha \|\beta\|_1\right)
}_\text{Penalization} \right\}\]</span></p>
<p>where</p>
<ul>
<li><p><span class="math inline">\(p_i=1/\left(1+\exp\left(-(\beta_0 +
x_i^T\beta)\right)\right)\)</span></p></li>
<li><p><span class="math inline">\(x\)</span> is denoting the genotypes
and covariables (e.g. principal components),</p></li>
<li><p><span class="math inline">\(y\)</span> is the disease status we
want to predict,</p></li>
<li><p><span class="math inline">\(\lambda\)</span> is a regularization
parameter that needs to be determined and</p></li>
<li><p><span class="math inline">\(\alpha\)</span> determines relative
parts of the regularization <span class="math inline">\(0 \le \alpha \le
1\)</span>.</p></li>
</ul>
<p><br></p>
<p>If you want to learn more about our implementation of PLR, please
look at <a href="https://doi.org/10.1101/403337" class="external-link">this paper</a>.</p>
<p><br></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Penalized logistic regression for many alphas and lambdas</span></span>
<span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_spLogReg.html" class="external-link">big_spLogReg</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">y</span><span class="op">[</span><span class="va">ind.train</span><span class="op">]</span>, <span class="va">ind.train</span>, covar.train <span class="op">=</span> <span class="va">svd</span><span class="op">$</span><span class="va">u</span><span class="op">[</span><span class="va">ind.train</span>, <span class="op">]</span>,</span>
<span>                    K <span class="op">=</span> <span class="fl">5</span>, alphas <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">(</span><span class="op">-</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span>, ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot regularization paths (from high lambda to low lambda) </span></span>
<span><span class="co"># for each validation set (color) and each alpha (facet)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The `guide` argument in `scale_*()` cannot be `FALSE`. This was deprecated</span></span>
<span><span class="co">## in ggplot2 3.3.4.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use "none" instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">bigstatsr</span> package.</span></span>
<span><span class="co">##   Please report the issue at</span></span>
<span><span class="co">##   <span style="color: #0000BB; font-style: italic;">&lt;</span></span></span>
<span><span class="co"><span style="color: #0000BB; font-style: italic;">## </span><span style="color: #555555; font-style: italic;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co"><span style="font-style: italic;">## </span><span style="color: #555555; font-style: italic;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co"><span style="font-style: italic;">## </span><span style="color: #555555; font-style: italic;">generated.</span></span></span></code></pre>
<p><img src="exo_files/figure-html/unnamed-chunk-11-1.png" width="692.708333333333" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get summaries of models</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 5 × 9</span></span></span>
<span><span class="co">##    alpha power_adaptive power_scale validation_loss intercept beta   nb_var</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> 0.000<span style="text-decoration: underline;">1</span>              0           1           0.557     -<span style="color: #BB0000;">1.41</span> <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="text-decoration: underline;">67</span>876</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> 0.001               0           1           0.550     -<span style="color: #BB0000;">1.94</span> <span style="color: #949494;">&lt;dbl&gt;</span>   <span style="text-decoration: underline;">16</span>008</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> 0.01                0           1           0.547     -<span style="color: #BB0000;">2.16</span> <span style="color: #949494;">&lt;dbl&gt;</span>    <span style="text-decoration: underline;">2</span>928</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> 0.1                 0           1           0.545     -<span style="color: #BB0000;">2.30</span> <span style="color: #949494;">&lt;dbl&gt;</span>     722</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> 1                   0           1           0.543     -<span style="color: #BB0000;">2.40</span> <span style="color: #949494;">&lt;dbl&gt;</span>     398</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: message &lt;list&gt;, all_conv &lt;lgl&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the predictions for the test set</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">mod</span>, <span class="va">G</span>, <span class="va">ind.test</span>, covar.row <span class="op">=</span> <span class="va">svd</span><span class="op">$</span><span class="va">u</span><span class="op">[</span><span class="va">ind.test</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Assess the Area Under the ROC Curve</span></span>
<span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/AUC.html" class="external-link">AUC</a></span><span class="op">(</span><span class="va">pred</span>, <span class="va">y</span><span class="op">[</span><span class="va">ind.test</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.7148903</span></span></code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
