<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Polygenic scores and inference using LDpred2 • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Polygenic scores and inference using LDpred2">
<meta property="og:description" content="bigsnpr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Polygenic scores and inference using
LDpred2</h1>
                        <h4 data-toc-skip class="author">Florian
Privé</h4>
            
            <h4 data-toc-skip class="date">June 8, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/vignettes/LDpred2.Rmd" class="external-link"><code>vignettes/LDpred2.Rmd</code></a></small>
      <div class="hidden name"><code>LDpred2.Rmd</code></div>

    </div>

    
    
<p>Here I show how to compute polygenic scores using <a href="https://doi.org/10.1093/bioinformatics/btaa1029" class="external-link">LDpred2</a>, as
well as inferring genetic architecture parameters with LDpred2-auto.</p>
<p>If you have to write a standalone LDpred2 command line script, have a
look at <a href="https://github.com/comorment/containers/tree/main/scripts/pgs/LDpred2" class="external-link">this
example written by two LDpred2 users</a>.</p>
<p><strong>Please be careful about the spelling; you should write
LDpred2, not LDPred2!</strong></p>
<p>Here is a (slightly outdated) video of me going through the tutorial
and explaining the different steps:</p>
<center>
<iframe width="560" height="315" src="https://www.youtube.com/embed/aya8WsNAu6U" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
</iframe>
</center>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>In R, run</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"privefl/bigsnpr"</span><span class="op">)</span></span></code></pre></div>
<p>or for the CRAN version</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"bigsnpr"</span><span class="op">)</span></span></code></pre></div>
<ul>
<li><p>If you install {bigsnpr} &gt;= v1.10.4, LDpred2-grid and
LDpred2-auto should be much faster for large data.</p></li>
<li><p>if you install {bigsnpr} &gt;= v1.11.4, there is a new version
LDpred2-auto that was validated for inferring parameters of genetic
architectures (cf. at the end of this tutorial).</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="which-set-of-variants-to-use">Which set of variants to use?<a class="anchor" aria-label="anchor" href="#which-set-of-variants-to-use"></a>
</h2>
<p>To run LDpred2, you need</p>
<ul>
<li>GWAS summary statistics with marginal effect sizes, their standard
errors, and the corresponding sample size(s),</li>
<li>an LD (correlation) matrix computed from individuals of the same
genetic ancestry as individuals used in the GWAS,</li>
<li>individual-level data for tuning hyper-parameters (when using
LDpred2-grid or lassosum2) and for testing the final models.</li>
</ul>
<p>You need to restrict to genetic variants in common between all these
datasets.</p>
<p>We previously recommended to use a set of 1,054,330 HapMap3 variants
for LDpred2, because they provide a good coverage of the genome and are
generally well imputed and available in most studies. In <a href="https://doi.org/10.1016/j.ajhg.2023.10.010" class="external-link">this paper</a>, we
recently proposed an extension of this set, to provide an even better
coverage of the genome by adding 37% more variants, and called it
HapMap3+. This is the preferred set of variants to use with LDpred2 when
power is sufficient.</p>
<p>You should use these sets of variants only when your data is imputed
so that the overlap is good. If you only have access to a chip of
genotyped variants, you should use these. In that case, you should
compute the LD information yourself (as done for the tutorial data
below).</p>
<p>If you use HM3/HM3+ variants with European summary statistics and do
not have enough data to use as LD reference (e.g. at least 2000
individuals), we provide LD matrices to be used directly:</p>
<ul>
<li>for <a href="https://doi.org/10.6084/m9.figshare.21305061" class="external-link">HapMap3+
variants with independent LD blocks</a> (from <a href="https://doi.org/10.1016/j.ajhg.2023.10.010" class="external-link">this paper</a>)</li>
<li>for <a href="https://doi.org/10.6084/m9.figshare.19213299" class="external-link">HapMap3
variants with independent LD blocks</a> (previously recommended in <a href="https://doi.org/10.1016/j.xhgg.2022.100136" class="external-link">this paper</a>)</li>
<li>along with an <a href="https://github.com/privefl/paper-infer/blob/main/code/example-with-provided-LD.R" class="external-link">example
R script</a> on how to use them</li>
</ul>
<p>Note that forming independent LD blocks in LD matrices can be useful
for robustness and extra speed gains (see <a href="https://doi.org/10.1016/j.xhgg.2022.100136" class="external-link">this paper</a>).</p>
<p>Information about the HapMap3+ variants can be retrieved with</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># $pos is in build GRCh37 / hg19, but we provide positions in 2 other builds</span></span>
<span><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu">runonce</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/runonce/man/download_file.html" class="external-link">download_file</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://figshare.com/ndownloader/files/37802721"</span>,</span>
<span>  dir <span class="op">=</span> <span class="st">"tmp-data"</span>, fname <span class="op">=</span> <span class="st">"map_hm3_plus.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">info</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Classes 'tbl_df', 'tbl' and 'data.frame':    1444196 obs. of  10 variables:</span></span>
<span><span class="co">##  $ chr     : int  1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##  $ pos     : int  68082 77866 87409 87647 88144 89654 565935 568631 720797 741397 ...</span></span>
<span><span class="co">##  $ a0      : chr  "T" "C" "C" "T" ...</span></span>
<span><span class="co">##  $ a1      : chr  "C" "T" "T" "C" ...</span></span>
<span><span class="co">##  $ rsid    : chr  "rs367789441" "rs563593912" "rs139490478" "rs146836579" ...</span></span>
<span><span class="co">##  $ af_UKBB : num  0.0708 0.07691 0.07683 0.01353 0.00857 ...</span></span>
<span><span class="co">##  $ ld      : num  1.81 1.3 1.57 1.34 1.16 ...</span></span>
<span><span class="co">##  $ block_id: int  1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##  $ pos_hg18: int  57945 67729 77272 77510 78007 79517 555798 558494 710660 731260 ...</span></span>
<span><span class="co">##  $ pos_hg38: int  68082 77866 87409 87647 88144 89654 630555 633251 785417 806017 ...</span></span></code></pre>
<p>Information about the HapMap3 variants can be retrieved with</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># $pos is in build GRCh37 / hg19, but we provide positions in 3 other builds</span></span>
<span><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu">runonce</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/runonce/man/download_file.html" class="external-link">download_file</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://figshare.com/ndownloader/files/36360900"</span>,</span>
<span>  dir <span class="op">=</span> <span class="st">"tmp-data"</span>, fname <span class="op">=</span> <span class="st">"map_hm3.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">info</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Classes 'tbl_df', 'tbl' and 'data.frame':    1054330 obs. of  11 variables:</span></span>
<span><span class="co">##  $ chr     : int  1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##  $ pos     : int  752721 754182 760912 768448 779322 838555 846808 853954 854250 864938 ...</span></span>
<span><span class="co">##  $ a0      : chr  "A" "A" "C" "G" ...</span></span>
<span><span class="co">##  $ a1      : chr  "G" "G" "T" "A" ...</span></span>
<span><span class="co">##  $ rsid    : chr  "rs3131972" "rs3131969" "rs1048488" "rs12562034" ...</span></span>
<span><span class="co">##  $ af_UKBB : num  0.841 0.87 0.84 0.106 0.128 ...</span></span>
<span><span class="co">##  $ ld      : num  3.69 3.73 3.69 1.4 3.68 ...</span></span>
<span><span class="co">##  $ pos_hg17: int  792584 794045 800775 808311 819185 878418 886671 893817 894113 904801 ...</span></span>
<span><span class="co">##  $ pos_hg18: int  742584 744045 750775 758311 769185 828418 836671 843817 844113 854801 ...</span></span>
<span><span class="co">##  $ pos_hg38: int  817341 818802 825532 833068 843942 903175 911428 918574 918870 929558 ...</span></span>
<span><span class="co">##  $ group_id: int  1 1 1 1 1 1 1 1 1 1 ...</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="downloading-genotype-data-and-summary-statistics-for-the-tutorial">Downloading genotype data and summary statistics for the
tutorial<a class="anchor" aria-label="anchor" href="#downloading-genotype-data-and-summary-statistics-for-the-tutorial"></a>
</h2>
<p>This tutorial uses fake data for educational purposes only. Another
tutorial using another dataset can be found at <a href="https://privefl.github.io/bigsnpr-extdoc/polygenic-scores-pgs.html" class="external-link uri">https://privefl.github.io/bigsnpr-extdoc/polygenic-scores-pgs.html</a>.</p>
<p>You can download <a href="https://github.com/privefl/bigsnpr/raw/master/data-raw/public-data3.zip" class="external-link">the
tutorial data</a> and unzip files in R. We store those files in a
directory called <code>"tmp-data"</code> here.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("runonce")</span></span>
<span><span class="va">zip</span> <span class="op">&lt;-</span> <span class="fu">runonce</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/runonce/man/download_file.html" class="external-link">download_file</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://github.com/privefl/bigsnpr/raw/master/data-raw/public-data3.zip"</span>,</span>
<span>  dir <span class="op">=</span> <span class="st">"tmp-data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">zip</span><span class="op">)</span></span></code></pre></div>
<p>First, you need to read genotype data from the PLINK files (or BGEN
files) as well as the text file containing summary statistics.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages bigsnpr and bigstatsr</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: bigstatsr</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read from bed/bim/fam, it generates .bk and .rds files.</span></span>
<span><span class="fu"><a href="../reference/snp_readBed.html">snp_readBed</a></span><span class="op">(</span><span class="st">"tmp-data/public-data3.bed"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "C:\\Users\\au639593\\Desktop\\bigsnpr\\tmp-data\\public-data3.rds"</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attach the "bigSNP" object in R session</span></span>
<span><span class="va">obj.bigSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="st">"tmp-data/public-data3.rds"</span><span class="op">)</span></span>
<span><span class="co"># See how the file looks like</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">obj.bigSNP</span>, max.level <span class="op">=</span> <span class="fl">2</span>, strict.width <span class="op">=</span> <span class="st">"cut"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 3</span></span>
<span><span class="co">##  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 16 ..</span></span>
<span><span class="co">##   ..and 26 methods, of which 12 are  possibly relevant:</span></span>
<span><span class="co">##   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,</span></span>
<span><span class="co">##   ..  check_write_permissions, copy#envRefClass, initialize,</span></span>
<span><span class="co">##   ..  initialize#FBM, save, show#envRefClass, show#FBM</span></span>
<span><span class="co">##  $ fam      :'data.frame':   503 obs. of  6 variables:</span></span>
<span><span class="co">##   ..$ family.ID  : int [1:503] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ sample.ID  : chr [1:503] "HG00096" "HG00097" "HG00099" "HG00100" ...</span></span>
<span><span class="co">##   ..$ paternal.ID: int [1:503] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ maternal.ID: int [1:503] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">##   ..$ sex        : int [1:503] 1 2 2 2 1 2 1 1 2 1 ...</span></span>
<span><span class="co">##   ..$ affection  : num [1:503] -0.547 0.188 -0.407 -0.28 0.398 ...</span></span>
<span><span class="co">##  $ map      :'data.frame':   45337 obs. of  6 variables:</span></span>
<span><span class="co">##   ..$ chromosome  : int [1:45337] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##   ..$ marker.ID   : chr [1:45337] "rs3934834" "rs12726255" "rs11260549" "..</span></span>
<span><span class="co">##   ..$ genetic.dist: num [1:45337] 0.359 0.408 0.932 0.986 1.107 ...</span></span>
<span><span class="co">##   ..$ physical.pos: int [1:45337] 1005806 1049950 1121794 1162435 1314015..</span></span>
<span><span class="co">##   ..$ allele1     : chr [1:45337] "T" "G" "A" "A" ...</span></span>
<span><span class="co">##   ..$ allele2     : chr [1:45337] "C" "A" "G" "C" ...</span></span>
<span><span class="co">##  - attr(*, "class")= chr "bigSNP"</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get aliases for useful slots</span></span>
<span><span class="va">G</span>   <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">genotypes</span></span>
<span><span class="va">CHR</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">chromosome</span></span>
<span><span class="va">POS</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">physical.pos</span></span>
<span><span class="va">y</span>   <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">fam</span><span class="op">$</span><span class="va">affection</span></span>
<span><span class="op">(</span><span class="va">NCORES</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/nb_cores.html" class="external-link">nb_cores</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read external summary statistics</span></span>
<span><span class="va">sumstats</span> <span class="op">&lt;-</span> <span class="fu">bigreadr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bigreadr/man/fread2.html" class="external-link">fread2</a></span><span class="op">(</span><span class="st">"tmp-data/public-data3-sumstats.txt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">sumstats</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    50000 obs. of  9 variables:</span></span>
<span><span class="co">##  $ rsid   : chr  "rs3934834" "rs12726255" "rs11260549" "rs3766186" ...</span></span>
<span><span class="co">##  $ chr    : int  1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##  $ pos    : int  995669 1039813 1111657 1152298 1303878 1495118 1833906 2041373 2130121 2201709 ...</span></span>
<span><span class="co">##  $ a0     : chr  "G" "T" "C" "C" ...</span></span>
<span><span class="co">##  $ a1     : chr  "A" "C" "T" "A" ...</span></span>
<span><span class="co">##  $ beta   : num  0.0125 0.027 0.0171 -0.0195 -0.0057 ...</span></span>
<span><span class="co">##  $ beta_se: num  0.0157 0.0167 0.0179 0.0199 0.0213 ...</span></span>
<span><span class="co">##  $ N      : int  15155 15155 15155 15155 15155 15155 15155 15155 15155 15155 ...</span></span>
<span><span class="co">##  $ p      : num  0.426 0.107 0.342 0.328 0.789 ...</span></span></code></pre>
<p>We split the individuals from the genotype data into “validation” (to
choose best-performing hyper-parameters) and “test” (to evaluate final
polygenic scores). Here we consider that there are 350 individuals to be
used as validation set to tune hyper-parameters for LDpred2-grid and
lassosum2. The other 153 individuals are used as test set to evaluate
the final models.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ind.val</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span>, <span class="fl">350</span><span class="op">)</span></span>
<span><span class="va">ind.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/seq-dim.html" class="external-link">rows_along</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span>, <span class="va">ind.val</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="matching-variants-between-genotype-data-and-summary-statistics">Matching variants between genotype data and summary statistics<a class="anchor" aria-label="anchor" href="#matching-variants-between-genotype-data-and-summary-statistics"></a>
</h2>
<p>To match variants contained in genotype data and summary statistics,
the variables <code>"chr"</code> (chromosome number), <code>"pos"</code>
(physical genetic position in bp), <code>"a0"</code> (reference allele)
and <code>"a1"</code> (alternative allele) should be available in the
summary statistics and in the genotype data. These 4 variables are used
to match variants between the two data frames. From the summary
statistics, you need to get <code>"beta"</code>, <code>"beta_se"</code>
(standard errors), and <code>"n_eff"</code> (the effective sample sizes
per variant for a GWAS using logistic regression, and simply the sample
size for continuous traits).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sumstats$n_eff &lt;- 4 / (1 / sumstats$n_case + 1 / sumstats$n_control)</span></span>
<span><span class="co"># sumstats$n_case &lt;- sumstats$n_control &lt;- NULL</span></span>
<span><span class="va">sumstats</span><span class="op">$</span><span class="va">n_eff</span> <span class="op">&lt;-</span> <span class="va">sumstats</span><span class="op">$</span><span class="va">N</span></span>
<span><span class="va">map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">[</span><span class="op">-</span><span class="fl">3</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"rsid"</span>, <span class="st">"pos"</span>, <span class="st">"a1"</span>, <span class="st">"a0"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df_beta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_match.html">snp_match</a></span><span class="op">(</span><span class="va">sumstats</span>, <span class="va">map</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 50,000 variants to be matched.</span></span></code></pre>
<pre><code><span><span class="co">## 0 ambiguous SNPs have been removed.</span></span></code></pre>
<pre><code><span><span class="co">## 4 variants have been matched; 2 were flipped and 2 were reversed.</span></span></code></pre>
<pre><code><span><span class="co">## Error: Not enough variants have been matched.</span></span></code></pre>
<p>Here, there is problem with the matching; this is due to having
different genome builds. You can either convert between builds with
<code><a href="../reference/snp_modifyBuild.html">snp_modifyBuild()</a></code> (or directly use the converted positions
in <code>info</code>), or match by rsIDs instead.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_beta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_match.html">snp_match</a></span><span class="op">(</span><span class="va">sumstats</span>, <span class="va">map</span>, join_by_pos <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>  <span class="co"># use rsid instead of pos</span></span></code></pre></div>
<pre><code><span><span class="co">## 50,000 variants to be matched.</span></span></code></pre>
<pre><code><span><span class="co">## 0 ambiguous SNPs have been removed.</span></span></code></pre>
<pre><code><span><span class="co">## 45,337 variants have been matched; 22,758 were flipped and 15,092 were reversed.</span></span></code></pre>
<p>If no or few variants are actually flipped, you might want to disable
the strand flipping option (<code>strand_flip = FALSE</code>) and maybe
remove the few that were flipped (errors?).</p>
</div>
<div class="section level2">
<h2 id="quality-control-of-gwas-summary-statistics">Quality control of GWAS summary statistics<a class="anchor" aria-label="anchor" href="#quality-control-of-gwas-summary-statistics"></a>
</h2>
<p><strong>Some quality control on summary statistics is highly
recommended.</strong> A refined QC is described in <a href="https://doi.org/10.1016/j.xhgg.2022.100136" class="external-link">this new paper</a>.
See e.g. <a href="https://github.com/privefl/paper-misspec/tree/main/code" class="external-link">the
code</a> that was used to prepare the sumstats there.</p>
<p>Most of the QC recommended for LDpred2 consists in computing <span class="math display">\[\begin{equation}
\text{sd}(G_j) \approx \dfrac{\text{sd}(y)}{\sqrt{n_j ~
\text{se}(\hat{\gamma}_j)^2 + \hat{\gamma}_j^2}} ~,
\end{equation}\]</span> where <span class="math inline">\(\hat{\gamma}_j\)</span> is the marginal GWAS
effect size of variant <span class="math inline">\(j\)</span>, <span class="math inline">\(n_j\)</span> is the GWAS sample size associated
with variant <span class="math inline">\(j\)</span>, <span class="math inline">\(y\)</span> is the vector of phenotypes and <span class="math inline">\({G}_j\)</span> is the vector of genotypes for
variant <span class="math inline">\(j\)</span>.</p>
<p>Note that, for a binary trait for which logistic regression is used,
we have instead <span class="math display">\[\begin{equation}\label{eq:approx-sd-log}
\text{sd}(G_j) \approx \dfrac{2}{\sqrt{n_j^\text{eff} ~
\text{se}(\hat{\gamma}_j)^2 + \hat{\gamma}_j^2}} ~,
\end{equation}\]</span> where <span class="math inline">\(n_j^\text{eff}
= \frac{4}{1 / n_j^\text{cases} + 1 / n_j^\text{controls}}\)</span>.</p>
<p>Then, these standard deviations of genotypes inferred from GWAS
summary statistics can be compared to <span class="math inline">\(\sqrt{2 \cdot f_j \cdot (1 - f_j) \cdot
\text{INFO}_j}\)</span>, where <span class="math inline">\(f_j\)</span>
is the allele frequency of variant <span class="math inline">\(j\)</span> and <span class="math inline">\(\text{INFO}_j\)</span> is used to correct for the
reduced variance of imputed dosages.</p>
<p>In the equation for continuous traits, we can estimate <span class="math inline">\(\text{sd}(y)\)</span> by the first percentile of
<span class="math inline">\(\sqrt{0.5 \left(n_j ~
\text{se}(\hat{\gamma}_j)^2 + \hat{\gamma}_j^2\right)}\)</span>, where
the first percentile approximates the minimum and is robust to
outliers.</p>
<p>In practice, a consistent deviation between these two estimates of
standard deviations (from summary statistics and from allele
frequencies) can also be explained by using wrong estimates for <span class="math inline">\(n_j\)</span> or <span class="math inline">\(n_j^\text{eff}\)</span>. This is for example the
case when computing <span class="math inline">\(N^\text{eff}\)</span>
from the total number of cases and controls from a meta-analysis (cf. <a href="https://doi.org/10.1016/j.biopsych.2022.05.029" class="external-link">this paper</a>).
In that case, you can estimate the overall <span class="math inline">\(N^\text{eff}\)</span> using
e.g. <code>quantile(8 / df_beta$beta_se^2, 0.999)</code>.</p>
</div>
<div class="section level2">
<h2 id="computing-ldpred2-scores-genome-wide">Computing LDpred2 scores genome-wide<a class="anchor" aria-label="anchor" href="#computing-ldpred2-scores-genome-wide"></a>
</h2>
<p>Note that you should <strong>run LDpred2 genome-wide</strong>. Just
build the SFBM (the sparse LD matrix on disk) so that it contains
selected variants for all chromosomes at once (see the for-loop
below).</p>
<div class="section level3">
<h3 id="correlation">Correlation<a class="anchor" aria-label="anchor" href="#correlation"></a>
</h3>
<p>First, you need to compute correlations between variants. We
recommend to use a window size of 3 cM (see <a href="https://doi.org/10.1093/bioinformatics/btaa1029" class="external-link">the LDpred2
paper</a>).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># To convert physical positions (in bp) to genetic positions (in cM), use</span></span>
<span><span class="co"># POS2 &lt;- snp_asGeneticPos(CHR, POS, dir = "tmp-data", ncores = NCORES)</span></span>
<span></span>
<span><span class="co"># To avoid downloading "large" files, `POS2` has been precomputed here</span></span>
<span><span class="va">POS2</span> <span class="op">&lt;-</span> <span class="va">obj.bigSNP</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">genetic.dist</span>  </span>
<span><span class="co"># /!\ this is usually just a vector of 0s /!\</span></span>
<span><span class="co"># Use `snp_asGeneticPos()` in your analyses</span></span></code></pre></div>
<p>Before computing the LD matrices, let us filter out variants with
small minor allele frequencies (MAFs):</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ind.row</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bigparallelr/man/seq-dim.html" class="external-link">rows_along</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_MAF.html">snp_MAF</a></span><span class="op">(</span><span class="va">G</span>, ind.row <span class="op">=</span> <span class="va">ind.row</span>, ind.col <span class="op">=</span> <span class="va">df_beta</span><span class="op">$</span><span class="va">`_NUM_ID_`</span>, ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span>
<span><span class="va">maf_thr</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ind.row</span><span class="op">)</span><span class="op">)</span>  <span class="co"># threshold I like to use</span></span>
<span><span class="va">df_beta</span> <span class="op">&lt;-</span> <span class="va">df_beta</span><span class="op">[</span><span class="va">maf</span> <span class="op">&gt;</span> <span class="va">maf_thr</span>, <span class="op">]</span></span></code></pre></div>
<p>Let us create the on-disk sparse genome-wide correlation matrix
on-the-fly:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>tmpdir <span class="op">=</span> <span class="st">"tmp-data"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">chr</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="co"># print(chr)</span></span>
<span>  </span>
<span>  <span class="co">## indices in 'df_beta'</span></span>
<span>  <span class="va">ind.chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">df_beta</span><span class="op">$</span><span class="va">chr</span> <span class="op">==</span> <span class="va">chr</span><span class="op">)</span></span>
<span>  <span class="co">## indices in 'G'</span></span>
<span>  <span class="va">ind.chr2</span> <span class="op">&lt;-</span> <span class="va">df_beta</span><span class="op">$</span><span class="va">`_NUM_ID_`</span><span class="op">[</span><span class="va">ind.chr</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># here we compute LD matrices ourselves, BUT</span></span>
<span>  <span class="co"># we recall that we provide pre-computed LD matrices that are </span></span>
<span>  <span class="co"># usually much better (larger N, LD blocks for robustness, etc)</span></span>
<span>  <span class="va">corr0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_cor.html">snp_cor</a></span><span class="op">(</span><span class="va">G</span>, ind.col <span class="op">=</span> <span class="va">ind.chr2</span>, size <span class="op">=</span> <span class="fl">3</span> <span class="op">/</span> <span class="fl">1000</span>,</span>
<span>                   infos.pos <span class="op">=</span> <span class="va">POS2</span><span class="op">[</span><span class="va">ind.chr2</span><span class="op">]</span>, ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ld</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">corr0</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="va">corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bigsparser/man/SFBM-class.html" class="external-link">as_SFBM</a></span><span class="op">(</span><span class="va">corr0</span>, <span class="va">tmp</span>, compact <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">ld</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ld</span>, <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">corr0</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">corr</span><span class="op">$</span><span class="fu">add_columns</span><span class="op">(</span><span class="va">corr0</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">corr</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>To use the “compact” format for SFBMs, you need
<code>packageVersion("bigsparser") &gt;= package_version("0.5")</code>.
Make sure to reinstall {bigsnpr} after updating {bigsparser} to this new
version (to avoid crashes).</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/file.info.html" class="external-link">file.size</a></span><span class="op">(</span><span class="va">corr</span><span class="op">$</span><span class="va">sbk</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1024</span><span class="op">^</span><span class="fl">3</span>  <span class="co"># file size in GB</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.03195963</span></span></code></pre>
<p>Note that you will need at least the same memory as this file size
(to keep it cached for faster processing) + some other memory for all
the results returned. If you do not have enough memory, processing will
be very slow (because you would read the data from disk all the time).
If using the one million HapMap3 variants, having 60 GB of memory should
be enough.</p>
</div>
<div class="section level3">
<h3 id="ldpred2-inf-infinitesimal-model">LDpred2-inf: infinitesimal model<a class="anchor" aria-label="anchor" href="#ldpred2-inf-infinitesimal-model"></a>
</h3>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimate of h2 from LD Score regression</span></span>
<span><span class="op">(</span><span class="va">ldsc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">df_beta</span>, <span class="fu"><a href="../reference/snp_ldsc.html">snp_ldsc</a></span><span class="op">(</span><span class="va">ld</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ld</span><span class="op">)</span>, chi2 <span class="op">=</span> <span class="op">(</span><span class="va">beta</span> <span class="op">/</span> <span class="va">beta_se</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span>,</span>
<span>                                sample_size <span class="op">=</span> <span class="va">n_eff</span>, blocks <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       int        h2 </span></span>
<span><span class="co">## 0.9506599 0.2965952</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ldsc_h2_est</span> <span class="op">&lt;-</span> <span class="va">ldsc</span><span class="op">[[</span><span class="st">"h2"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>Note that parameter <code>ld_size</code> from <code><a href="../reference/snp_ldsc.html">snp_ldsc()</a></code>
is the initial number of variants used to compute the LD scores, and
should be equal to <code>nrow(info)</code> instead of
<code>length(ld)</code> when using the pre-computed LD scores
<code>info$ld</code>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beta_inf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LDpred2.html">snp_ldpred2_inf</a></span><span class="op">(</span><span class="va">corr</span>, <span class="va">df_beta</span>, h2 <span class="op">=</span> <span class="va">ldsc_h2_est</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_inf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_prodVec.html" class="external-link">big_prodVec</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">beta_inf</span>, ind.row <span class="op">=</span> <span class="va">ind.test</span>, ind.col <span class="op">=</span> <span class="va">df_beta</span><span class="op">[[</span><span class="st">"_NUM_ID_"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/pcor.html" class="external-link">pcor</a></span><span class="op">(</span><span class="va">pred_inf</span>, <span class="va">y</span><span class="op">[</span><span class="va">ind.test</span><span class="op">]</span>, <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.3196441 0.1695664 0.4552326</span></span></code></pre>
<p>LDpred2-inf would very likely perform worse than the other models
presented hereinafter. We actually recommend not to use it anymore.</p>
</div>
<div class="section level3">
<h3 id="ldpred2-grid-grid-of-models">LDpred2(-grid): grid of models<a class="anchor" aria-label="anchor" href="#ldpred2-grid-grid-of-models"></a>
</h3>
<p>In practice, we recommend to test multiple values for h2 and p. </p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">h2_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">ldsc_h2_est</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.7</span>, <span class="fl">1</span>, <span class="fl">1.4</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.0890 0.2076 0.2966 0.4152</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">p_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="fu"><a href="../reference/seq_log.html">seq_log</a></span><span class="op">(</span><span class="fl">1e-5</span>, <span class="fl">1</span>, length.out <span class="op">=</span> <span class="fl">21</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 1.0e-05 1.8e-05 3.2e-05 5.6e-05 1.0e-04 1.8e-04 3.2e-04 5.6e-04</span></span>
<span><span class="co">##  [9] 1.0e-03 1.8e-03 3.2e-03 5.6e-03 1.0e-02 1.8e-02 3.2e-02 5.6e-02</span></span>
<span><span class="co">## [17] 1.0e-01 1.8e-01 3.2e-01 5.6e-01 1.0e+00</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html" class="external-link">expand.grid</a></span><span class="op">(</span>p <span class="op">=</span> <span class="va">p_seq</span>, h2 <span class="op">=</span> <span class="va">h2_seq</span>, sparse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          p    h2 sparse</span></span>
<span><span class="co">## 1  1.0e-05 0.089  FALSE</span></span>
<span><span class="co">## 2  1.8e-05 0.089  FALSE</span></span>
<span><span class="co">## 3  3.2e-05 0.089  FALSE</span></span>
<span><span class="co">## 4  5.6e-05 0.089  FALSE</span></span>
<span><span class="co">## 5  1.0e-04 0.089  FALSE</span></span>
<span><span class="co">## 6  1.8e-04 0.089  FALSE</span></span>
<span><span class="co">## 7  3.2e-04 0.089  FALSE</span></span>
<span><span class="co">## 8  5.6e-04 0.089  FALSE</span></span>
<span><span class="co">## 9  1.0e-03 0.089  FALSE</span></span>
<span><span class="co">## 10 1.8e-03 0.089  FALSE</span></span>
<span><span class="co">##  [ reached 'max' / getOption("max.print") -- omitted 158 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>  <span class="co"># to get the same result every time</span></span>
<span><span class="co"># takes less than 2 min with 4 cores</span></span>
<span><span class="va">beta_grid</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LDpred2.html">snp_ldpred2_grid</a></span><span class="op">(</span><span class="va">corr</span>, <span class="va">df_beta</span>, <span class="va">params</span>, ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_grid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_prodMat.html" class="external-link">big_prodMat</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">beta_grid</span>, ind.col <span class="op">=</span> <span class="va">df_beta</span><span class="op">[[</span><span class="st">"_NUM_ID_"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">params</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">pred_grid</span><span class="op">[</span><span class="va">ind.val</span>, <span class="op">]</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="cn">NA</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">ind.val</span><span class="op">]</span> <span class="op">~</span> <span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="st">"x"</span>, <span class="fl">3</span><span class="op">]</span></span>
<span>  <span class="co"># summary(glm(y[ind.val] ~ x, family = "binomial"))$coef["x", 3]</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Note that missing values represent models that diverged
substantially.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: package 'ggplot2' was built under R version 4.2.3</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">params</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">p</span>, y <span class="op">=</span> <span class="va">score</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">h2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/theme_bigstatsr.html" class="external-link">theme_bigstatsr</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">(</span><span class="op">-</span><span class="fl">5</span><span class="op">:</span><span class="fl">0</span><span class="op">)</span>, minor_breaks <span class="op">=</span> <span class="va">params</span><span class="op">$</span><span class="va">p</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">sparse</span>, labeller <span class="op">=</span> <span class="va">label_both</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"GLM Z-Score"</span>, color <span class="op">=</span> <span class="st">"h2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span>, panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"lines"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="LDpred2_files/figure-html/unnamed-chunk-20-1.png" width="90%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">params</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>sparsity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">beta_grid</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>, id <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_at</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"score"</span>, <span class="st">"sparsity"</span><span class="op">)</span>, <span class="va">round</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       p     h2 sparse  score sparsity  id</span></span>
<span><span class="co">## 1 0.018 0.4152  FALSE 10.308    0.000  77</span></span>
<span><span class="co">## 2 0.018 0.2966   TRUE 10.237    0.745 140</span></span>
<span><span class="co">## 3 0.010 0.2966  FALSE 10.187    0.000  55</span></span>
<span><span class="co">## 4 0.018 0.2966  FALSE 10.175    0.000  56</span></span>
<span><span class="co">## 5 0.010 0.4152  FALSE 10.169    0.000  76</span></span>
<span><span class="co">##  [ reached 'max' / getOption("max.print") -- omitted 5 rows ]</span></span></code></pre>
<p>You can then choose the best model according to your preferred
criterion (e.g. max AUC or <span class="math inline">\(r^2\)</span>).
Here, we use the Z-Score from the (linear or logistic) regression of the
phenotype by the PRS since we have found it more robust than using the
correlation or the AUC. It also enables adjusting for covariates in this
step.</p>
<p>Also note that we separate both sparse and non-sparse models to show
that their predictive performance are similar (in the original LDpred2
paper). In practice, if you do not really care about sparsity, you could
choose the best LDpred2-grid model among all sparse and non-sparse
models. If you do, choose the best sparse one (if it is close enough to
the best one).</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">best_beta_grid</span> <span class="op">&lt;-</span> <span class="va">params</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>id <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># filter(sparse) %&gt;% </span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">beta_grid</span><span class="op">[</span>, <span class="va">.</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##       p     h2 sparse    score id</span></span>
<span><span class="co">## 1 0.018 0.4152  FALSE 10.30816 77</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_prodVec.html" class="external-link">big_prodVec</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">best_beta_grid</span>, ind.row <span class="op">=</span> <span class="va">ind.test</span>,</span>
<span>                    ind.col <span class="op">=</span> <span class="va">df_beta</span><span class="op">[[</span><span class="st">"_NUM_ID_"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/pcor.html" class="external-link">pcor</a></span><span class="op">(</span><span class="va">pred</span>, <span class="va">y</span><span class="op">[</span><span class="va">ind.test</span><span class="op">]</span>, <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.5109565 0.3833604 0.6194140</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="ldpred2-auto-automatic-model">LDpred2-auto: automatic model<a class="anchor" aria-label="anchor" href="#ldpred2-auto-automatic-model"></a>
</h3>
<p>Contrary to LDpred2-grid, LDpred2-auto does not need a validation set
because it can directly infer values for hyper-parameters <span class="math inline">\(h^2\)</span> and <span class="math inline">\(p\)</span>. We recommend to run many chains in
parallel with different initial values for <code>p</code>, which will be
used for QC afterwards. In <a href="https://doi.org/10.1016/j.xhgg.2022.100136" class="external-link">this paper</a>, we
have also introduced two new parameters in LDpred2-auto for improving
its robustness, <code>allow_jump_sign</code> and
<code>shrink_corr</code>, and recommend to use them.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coef_shrink</span> <span class="op">&lt;-</span> <span class="fl">0.95</span>  <span class="co"># reduce this up to 0.4 if you have some (large) mismatch with the LD ref</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>  <span class="co"># to get the same result every time</span></span>
<span><span class="co"># takes less than 2 min with 4 cores</span></span>
<span><span class="va">multi_auto</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LDpred2.html">snp_ldpred2_auto</a></span><span class="op">(</span></span>
<span>  <span class="va">corr</span>, <span class="va">df_beta</span>, h2_init <span class="op">=</span> <span class="va">ldsc_h2_est</span>,</span>
<span>  vec_p_init <span class="op">=</span> <span class="fu"><a href="../reference/seq_log.html">seq_log</a></span><span class="op">(</span><span class="fl">1e-4</span>, <span class="fl">0.2</span>, length.out <span class="op">=</span> <span class="fl">30</span><span class="op">)</span>, ncores <span class="op">=</span> <span class="va">NCORES</span>,</span>
<span>  <span class="co"># use_MLE = FALSE,  # uncomment if you have convergence issues or when power is low (need v1.11.9)</span></span>
<span>  allow_jump_sign <span class="op">=</span> <span class="cn">FALSE</span>, shrink_corr <span class="op">=</span> <span class="va">coef_shrink</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">multi_auto</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 30</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span>
<span><span class="co">##  $ :List of 12</span></span></code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">multi_auto</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 12</span></span>
<span><span class="co">##  $ beta_est      : num [1:45268] 6.96e-05 3.40e-04 1.09e-04 -1.29e-04 2.69e-05 ...</span></span>
<span><span class="co">##  $ postp_est     : num [1:45268] 0.00667 0.01525 0.00782 0.00828 0.00599 ...</span></span>
<span><span class="co">##  $ corr_est      : num [1:45268] 1.22e-05 7.14e-05 1.11e-06 -6.06e-05 3.65e-05 ...</span></span>
<span><span class="co">##  $ sample_beta   :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">##  $ path_p_est    : num [1:700] 0.000645 0.001155 0.00198 0.002246 0.002902 ...</span></span>
<span><span class="co">##  $ path_h2_est   : num [1:700] 0.0889 0.1047 0.1346 0.1391 0.1444 ...</span></span>
<span><span class="co">##  $ path_alpha_est: num [1:700] 0.5 0.5 0.302 0.159 -0.341 ...</span></span>
<span><span class="co">##  $ h2_est        : num 0.219</span></span>
<span><span class="co">##  $ p_est         : num 0.0126</span></span>
<span><span class="co">##  $ alpha_est     : num -0.497</span></span>
<span><span class="co">##  $ h2_init       : num 0.297</span></span>
<span><span class="co">##  $ p_init        : num 1e-04</span></span></code></pre>
<p>You can verify whether the chains “converged” by looking at the path
of the chains:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">auto</span> <span class="op">&lt;-</span> <span class="va">multi_auto</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>  <span class="co"># first chain</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/qplot.html" class="external-link">qplot</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">auto</span><span class="op">$</span><span class="va">path_p_est</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/theme_bigstatsr.html" class="external-link">theme_bigstatsr</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">auto</span><span class="op">$</span><span class="va">p_est</span>, col <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"p"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/qplot.html" class="external-link">qplot</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">auto</span><span class="op">$</span><span class="va">path_h2_est</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/theme_bigstatsr.html" class="external-link">theme_bigstatsr</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">auto</span><span class="op">$</span><span class="va">h2_est</span>, col <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"h2"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">1</span>, align <span class="op">=</span> <span class="st">"hv"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: `qplot()` was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<p><img src="LDpred2_files/figure-html/unnamed-chunk-25-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>In the LDpred2 paper, we proposed an automatic way of filtering bad
chains by comparing the scale of the resulting predictions. We have
tested a somewhat equivalent and simpler alternative since, which we
recommend here:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># `range` should be between 0 and 2</span></span>
<span><span class="op">(</span><span class="va">range</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">multi_auto</span>, <span class="kw">function</span><span class="op">(</span><span class="va">auto</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">auto</span><span class="op">$</span><span class="va">corr_est</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 0.1204140 0.1201149 0.1203564 0.1186828 0.1200849 0.1204433 0.1212793</span></span>
<span><span class="co">##  [8] 0.1203259 0.1206732 0.1195239 0.1193046 0.1206013 0.1199865 0.1195314</span></span>
<span><span class="co">## [15] 0.1204132 0.1203583 0.1184779 0.1189227 0.1202364 0.1199066 0.1191860</span></span>
<span><span class="co">## [22] 0.1190776 0.1199808 0.1205954 0.1207522 0.1213208 0.1211963 0.1212839</span></span>
<span><span class="co">## [29] 0.1195293 0.1209223</span></span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">range</span> <span class="op">&gt;</span> <span class="op">(</span><span class="fl">0.95</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">range</span>, <span class="fl">0.95</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23</span></span>
<span><span class="co">## [24] 24 25 26 27 28 29 30</span></span></code></pre>
<p>To get the final effects / predictions, <strong>you should only use
chains that pass this filtering</strong>:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beta_auto</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">multi_auto</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">auto</span><span class="op">)</span> <span class="va">auto</span><span class="op">$</span><span class="va">beta_est</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pred_auto</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_prodVec.html" class="external-link">big_prodVec</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">beta_auto</span>, ind.row <span class="op">=</span> <span class="va">ind.test</span>, ind.col <span class="op">=</span> <span class="va">df_beta</span><span class="op">[[</span><span class="st">"_NUM_ID_"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/pcor.html" class="external-link">pcor</a></span><span class="op">(</span><span class="va">pred_auto</span>, <span class="va">y</span><span class="op">[</span><span class="va">ind.test</span><span class="op">]</span>, <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.4979167 0.3683406 0.6085168</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="lassosum2-grid-of-models">lassosum2: grid of models<a class="anchor" aria-label="anchor" href="#lassosum2-grid-of-models"></a>
</h3>
<p>lassosum2 is a re-implementation of <a href="https://doi.org/10.1002/gepi.22050" class="external-link">the lassosum model</a> that
now uses the exact same input parameters as LDpred2 (<code>corr</code>
and <code>df_beta</code>). It should be fast to run. It can be run next
to LDpred2 and the best model can be chosen using the validation set.
Note that parameter ‘s’ from lassosum has been replaced by a new
parameter ‘delta’ in lassosum2, in order to better reflect that the
lassosum model also uses L2-regularization (therefore, elastic-net
regularization).</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beta_lassosum2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/snp_lassosum2.html">snp_lassosum2</a></span><span class="op">(</span><span class="va">corr</span>, <span class="va">df_beta</span>, ncores <span class="op">=</span> <span class="va">NCORES</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">params2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">beta_lassosum2</span>, <span class="st">"grid_param"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       lambda delta num_iter time  sparsity</span></span>
<span><span class="co">## 1 0.06111003 0.001        2 0.00 0.9998895</span></span>
<span><span class="co">## 2 0.05241382 0.001        3 0.02 0.9997128</span></span>
<span><span class="co">## 3 0.04495512 0.001        4 0.00 0.9995361</span></span>
<span><span class="co">## 4 0.03855782 0.001        7 0.00 0.9992489</span></span>
<span><span class="co">## 5 0.03307088 0.001        8 0.02 0.9984095</span></span>
<span><span class="co">## 6 0.02836476 0.001       13 0.02 0.9971061</span></span>
<span><span class="co">##  [ reached 'max' / getOption("max.print") -- omitted 114 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred_grid2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_prodMat.html" class="external-link">big_prodMat</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">beta_lassosum2</span>, ind.col <span class="op">=</span> <span class="va">df_beta</span><span class="op">[[</span><span class="st">"_NUM_ID_"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">params2</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">pred_grid2</span><span class="op">[</span><span class="va">ind.val</span>, <span class="op">]</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="cn">NA</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">ind.val</span><span class="op">]</span> <span class="op">~</span> <span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="st">"x"</span>, <span class="fl">3</span><span class="op">]</span></span>
<span>  <span class="co"># summary(glm(y[ind.val] ~ x, family = "binomial"))$coef["x", 3]</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">params2</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">lambda</span>, y <span class="op">=</span> <span class="va">score</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">delta</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/theme_bigstatsr.html" class="external-link">theme_bigstatsr</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">(</span><span class="op">-</span><span class="fl">5</span><span class="op">:</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"GLM Z-Score"</span>, color <span class="op">=</span> <span class="st">"delta"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 10 rows containing missing values or values outside the scale</span></span>
<span><span class="co">## range (`geom_point()`).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 10 rows containing missing values or values outside the scale</span></span>
<span><span class="co">## range (`geom_line()`).</span></span></code></pre>
<p><img src="LDpred2_files/figure-html/unnamed-chunk-31-1.png" width="90%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">best_grid_lassosum2</span> <span class="op">&lt;-</span> <span class="va">params2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>id <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">beta_lassosum2</span><span class="op">[</span>, <span class="va">.</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##       lambda delta num_iter time  sparsity    score id</span></span>
<span><span class="co">## 1 0.02086632 0.001       21 0.03 0.9864584 9.715472  8</span></span>
<span><span class="co">## 2 0.02086632 0.010       20 0.03 0.9864805 9.709865 38</span></span>
<span><span class="co">## 3 0.02086632 0.100       16 0.01 0.9862375 9.667334 68</span></span>
<span><span class="co">## 4 0.02432834 0.001       19 0.01 0.9942564 9.404453  7</span></span>
<span><span class="co">##  [ reached 'max' / getOption("max.print") -- omitted 116 rows ]</span></span></code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Choose the best among all LDpred2-grid and lassosum2 models</span></span>
<span><span class="va">best_grid_overall</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">`if`</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">params2</span><span class="op">$</span><span class="va">score</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">params</span><span class="op">$</span><span class="va">score</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>       <span class="va">best_grid_lassosum2</span>, <span class="va">best_beta_grid</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="inference-with-ldpred2-auto">Inference with LDpred2-auto<a class="anchor" aria-label="anchor" href="#inference-with-ldpred2-auto"></a>
</h2>
<p>Here, you need
<code>packageVersion("bigsnpr") &gt;= package_version("1.11.4")</code>.
Here is <a href="https://doi.org/10.1016/j.ajhg.2023.10.010" class="external-link">the
accompanying paper</a>.</p>
<div class="figure" style="text-align: center">
<img src="https://raw.githubusercontent.com/privefl/paper-infer/main/overview.svg" alt="Overview of what LDpred2-auto can now provide" width="80%"><p class="caption">
Overview of what LDpred2-auto can now provide
</p>
</div>
<p>LDpred2-auto has been very recently extended and improved to allow
for estimating <span class="math inline">\(h^2\)</span>, <span class="math inline">\(p\)</span>, and <span class="math inline">\(\alpha\)</span>, a third parameter that controls
how expected effect sizes relate to minor allele frequencies. The new
model assumed by LDpred2-auto is <span class="math display">\[\begin{equation}
\beta_j = S_j \gamma_j \sim \left\{
\begin{array}{ll}
\mathcal N \big( 0,~\sigma_\beta^2 \cdot (S_j^2)^{(\alpha + 1)} \big)
&amp; \mbox{with probability $p$,} \\
0 &amp; \mbox{otherwise,}
\end{array}
\right.
\end{equation}\]</span> where <span class="math inline">\(p\)</span> is
the proportion of causal variants, <span class="math inline">\(h^2\)</span> the (SNP) heritability, <span class="math inline">\(\boldsymbol{\gamma}\)</span> the effect sizes on
the allele scale, <span class="math inline">\(\boldsymbol{S}\)</span>
the standard deviations of the genotypes, and <span class="math inline">\(\boldsymbol{\beta}\)</span> the effects of the
scaled genotypes.</p>
<p>Parameters <span class="math inline">\(h^2\)</span>, <span class="math inline">\(p\)</span>, and <span class="math inline">\(\alpha\)</span> (and 95% CIs) can be estimated
using:</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># reduce this up to 0.4 if you have some (large) mismatch with the LD ref</span></span>
<span><span class="co"># /!\ but the inference of h2 and p might be biased if you do this (see paper)</span></span>
<span><span class="va">coef_shrink</span> <span class="op">&lt;-</span> <span class="fl">0.95</span>  </span>
<span></span>
<span><span class="va">multi_auto</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LDpred2.html">snp_ldpred2_auto</a></span><span class="op">(</span></span>
<span>  <span class="va">corr</span>, <span class="va">df_beta</span>, h2_init <span class="op">=</span> <span class="va">ldsc_h2_est</span>,</span>
<span>  vec_p_init <span class="op">=</span> <span class="fu"><a href="../reference/seq_log.html">seq_log</a></span><span class="op">(</span><span class="fl">1e-4</span>, <span class="fl">0.2</span>, length.out <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>, ncores <span class="op">=</span> <span class="va">NCORES</span>,</span>
<span>  burn_in <span class="op">=</span> <span class="fl">500</span>, num_iter <span class="op">=</span> <span class="fl">500</span>, report_step <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  <span class="co"># use_MLE = FALSE,  # uncomment if you have convergence issues or when power is low (need v1.11.9)</span></span>
<span>  allow_jump_sign <span class="op">=</span> <span class="cn">FALSE</span>, shrink_corr <span class="op">=</span> <span class="va">coef_shrink</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">range</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">multi_auto</span>, <span class="kw">function</span><span class="op">(</span><span class="va">auto</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">auto</span><span class="op">$</span><span class="va">corr_est</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 0.1206431 0.1196823 0.1200970 0.1206415 0.1201813 0.1201417 0.1202826</span></span>
<span><span class="co">##  [8] 0.1200190 0.1198826 0.1200731 0.1201293 0.1208081 0.1200851 0.1210606</span></span>
<span><span class="co">## [15] 0.1226234 0.1203134 0.1202046 0.1205214 0.1194405 0.1192934 0.1203470</span></span>
<span><span class="co">## [22] 0.1205446 0.1199081 0.1196137 0.1207115 0.1209532 0.1206560 0.1205325</span></span>
<span><span class="co">## [29] 0.1202039 0.1191373</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 20 entries ]</span></span></code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">range</span> <span class="op">&gt;</span> <span class="op">(</span><span class="fl">0.95</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">range</span>, <span class="fl">0.95</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23</span></span>
<span><span class="co">## [24] 24 25 26 27 28 29 30</span></span>
<span><span class="co">##  [ reached getOption("max.print") -- omitted 20 entries ]</span></span></code></pre>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_h2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">multi_auto</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">auto</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">auto</span><span class="op">$</span><span class="va">path_h2_est</span>, <span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">all_h2</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.025</span>, <span class="fl">0.975</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       50%      2.5%     97.5% </span></span>
<span><span class="co">## 0.2181211 0.1899359 0.2483795</span></span></code></pre>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">multi_auto</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">auto</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">auto</span><span class="op">$</span><span class="va">path_p_est</span>, <span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">all_p</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.025</span>, <span class="fl">0.975</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         50%        2.5%       97.5% </span></span>
<span><span class="co">## 0.012380167 0.008710519 0.017203381</span></span></code></pre>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_alpha</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">multi_auto</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">auto</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">auto</span><span class="op">$</span><span class="va">path_alpha_est</span>, <span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">all_alpha</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.025</span>, <span class="fl">0.975</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         50%        2.5%       97.5% </span></span>
<span><span class="co">## -0.49629449 -0.91317071 -0.04713767</span></span></code></pre>
<p>Predictive performance <span class="math inline">\(r^2\)</span> can
also be inferred from the Gibbs sampler:</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bsamp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">multi_auto</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">auto</span><span class="op">)</span> <span class="va">auto</span><span class="op">$</span><span class="va">sample_beta</span><span class="op">)</span></span>
<span><span class="va">all_r2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">bsamp</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ic</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">b1</span> <span class="op">&lt;-</span> <span class="va">bsamp</span><span class="op">[[</span><span class="va">ic</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">Rb1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">b1</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">coef_shrink</span> <span class="op">*</span> <span class="fu">bigsparser</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bigsparser/man/sp_prodVec.html" class="external-link">sp_prodVec</a></span><span class="op">(</span><span class="va">corr</span>, <span class="va">x</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">coef_shrink</span><span class="op">)</span> <span class="op">*</span> <span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">b2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="va">bsamp</span><span class="op">[</span><span class="op">-</span><span class="va">ic</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">b2Rb1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/matrix-products.html" class="external-link">crossprod</a></span><span class="op">(</span><span class="va">b2</span>, <span class="va">Rb1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">all_r2</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.025</span>, <span class="fl">0.975</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       50%      2.5%     97.5% </span></span>
<span><span class="co">## 0.1151870 0.1039609 0.1266073</span></span></code></pre>
<p>and compared to</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">beta_auto</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">multi_auto</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">auto</span><span class="op">)</span> <span class="va">auto</span><span class="op">$</span><span class="va">beta_est</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pred_auto</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/big_prodVec.html" class="external-link">big_prodVec</a></span><span class="op">(</span><span class="va">G</span>, <span class="va">beta_auto</span>, ind.col <span class="op">=</span> <span class="va">df_beta</span><span class="op">[[</span><span class="st">"_NUM_ID_"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/pcor.html" class="external-link">pcor</a></span><span class="op">(</span><span class="va">pred_auto</span>, <span class="va">y</span>, <span class="cn">NULL</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.2377675 0.1747289 0.3041855</span></span></code></pre>
<p>These are not exactly the same, which we attribute to the small
number of variants used in this tutorial data.</p>
<p>You can also get per-variant probabilities of being causal (for
fine-mapping purposes)</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">postp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">multi_auto</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">auto</span><span class="op">)</span> <span class="va">auto</span><span class="op">$</span><span class="va">postp_est</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/qplot.html" class="external-link">qplot</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">postp</span>, alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></span><span class="op">(</span><span class="fl">0.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/theme_bigstatsr.html" class="external-link">theme_bigstatsr</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="LDpred2_files/figure-html/unnamed-chunk-40-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>You can also infer local per-block or per-annotation <span class="math inline">\(h^2\)</span> following the code <a href="https://github.com/privefl/paper-infer/blob/3dcd2949f54d7f1fd5c346d1bbb35db4b84a719e/code/run-ldpred2-ukbb.R#L158-L165" class="external-link">here</a>
and <a href="https://github.com/privefl/paper-infer/blob/3dcd2949f54d7f1fd5c346d1bbb35db4b84a719e/code/height-meta-1600k.R#L170-L216" class="external-link">here</a>.</p>
<hr>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Some cleaning</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">corr</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/gc.html" class="external-link">gc</a></span><span class="op">(</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">tmp</span>, <span class="st">".sbk"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            used  (Mb) gc trigger  (Mb) max used  (Mb)</span></span>
<span><span class="co">## Ncells  3895939 208.1    7147978 381.8  7147978 381.8</span></span>
<span><span class="co">## Vcells 40890231 312.0   71390344 544.7 71389651 544.7</span></span></code></pre>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li><p>Privé, F., Arbel, J., &amp; Vilhjálmsson, B. J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa1029" class="external-link">LDpred2: better,
faster, stronger</a>. <em>Bioinformatics</em>, 36(22-23),
5424-5431.</p></li>
<li><p>Privé, F., Arbel, J., Aschard, H., &amp; Vilhjálmsson, B. J.
(2022). <a href="https://doi.org/10.1016/j.xhgg.2022.100136" class="external-link">Identifying
and correcting for misspecifications in GWAS summary statistics and
polygenic scores</a>. <em>Human Genetics and Genomics Advances</em>,
3(4), 100136.</p></li>
<li><p>Privé, F., Albiñana, C., Arbel, J., Pasaniuc, B., &amp;
Vilhjálmsson, B. J. (2022). <a href="https://doi.org/10.1016/j.ajhg.2023.10.010" class="external-link">Inferring disease
architecture and predictive ability with LDpred2-auto</a>. <em>The
American Journal of Human Genetics</em>, 110(12), 2042-2055.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
