<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Why clumping should be preferred over pruning • bigsnpr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Why clumping should be preferred over pruning">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Why clumping should be preferred over pruning</h1>
                        <h4 class="author">Florian Privé</h4>
            
            <h4 class="date">November 24, 2016 (updated on January 3, 2018)</h4>
      
      
      <div class="hidden name"><code>pruning-vs-clumping.Rmd</code></div>

    </div>

    
    
<p>In this vignette, I show why clumping (on MAF) should be preferred over pruning for single-nucleotide polymorphism (SNP) arrays.</p>
<p>Pruning and clumping are used to keep a subset of SNPs that are nearly uncorrelated with each other. For instance, pruning is used before Principal Component Analysis to avoid capturing too much variance of linkage disequilibrium (LD) regions. Clumping is used to keep only one representative SNP per region of LD. <!-- , for example when reporting p-values of genome-wide association studies (GWAS). --></p>
<div id="simulation" class="section level2">
<h2 class="hasAnchor">
<a href="#simulation" class="anchor"></a>Simulation</h2>
<p>I generate a (toy) SNP array with 500 individuals and 10 SNPs, where each SNP has a squared correlation &gt; 0.2 with their direct neighbors and only with them. Moreover, the SNPs have an increasing MAF.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gen &lt;-<span class="st"> </span><span class="cf">function</span>(n, m) {
  I &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>m
  p &lt;-<span class="st"> </span>I <span class="op">/</span><span class="st"> </span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>m <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)

  mat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/outer">outer</a></span>(I, I, <span class="dt">FUN =</span> <span class="cf">function</span>(i, j) {
    <span class="dv">1</span> <span class="op">/</span><span class="st"> </span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(i <span class="op">-</span><span class="st"> </span>j) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
  })

  bindata<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/bindata/topics/rmvbin">rmvbin</a></span>(n, p, <span class="dt">bincorr =</span> mat) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span>bindata<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/bindata/topics/rmvbin">rmvbin</a></span>(n, p, <span class="dt">bincorr =</span> mat)
}

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">1</span>)
X &lt;-<span class="st"> </span><span class="kw">gen</span>(<span class="dv">500</span>, <span class="dv">10</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(X, <span class="dv">20</span>))</code></pre></div>
<pre><code>##       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
##  [1,]    0    0    0    1    0    0    1    1    1     1
##  [2,]    0    0    0    0    0    0    1    1    1     1
##  [3,]    0    1    1    0    0    0    0    0    0     1
##  [4,]    0    0    0    0    0    0    0    0    1     1
##  [5,]    0    0    0    0    0    0    0    1    1     1
##  [6,]    0    0    0    0    1    1    0    0    1     1
##  [7,]    1    0    1    0    1    1    1    2    2     1
##  [8,]    1    1    1    1    1    1    0    1    1     1
##  [9,]    0    0    0    0    0    1    1    0    1     1
## [10,]    0    0    1    1    1    1    1    1    1     1
## [11,]    0    1    1    0    0    2    2    2    2     1
## [12,]    0    0    0    0    0    0    0    0    0     0
## [13,]    0    0    0    0    0    0    1    0    1     1
## [14,]    0    0    0    0    0    0    0    0    1     1
## [15,]    0    0    0    0    0    0    1    0    0     0
## [16,]    0    0    0    0    0    0    0    1    1     1
## [17,]    0    0    1    1    0    1    0    0    1     1
## [18,]    1    1    2    1    1    2    2    2    2     2
## [19,]    0    1    0    0    0    1    2    1    1     0
## [20,]    0    0    0    0    0    0    1    1    0     0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(X)<span class="op">^</span><span class="dv">2</span>, <span class="dv">2</span>)) <span class="co"># squared correlation between SNPs</span></code></pre></div>
<pre><code>##       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
##  [1,] 1.00 0.23 0.13 0.08 0.05 0.03 0.02 0.01 0.01  0.00
##  [2,] 0.23 1.00 0.21 0.09 0.05 0.05 0.01 0.01 0.01  0.01
##  [3,] 0.13 0.21 1.00 0.25 0.12 0.10 0.06 0.06 0.02  0.01
##  [4,] 0.08 0.09 0.25 1.00 0.31 0.16 0.07 0.07 0.03  0.04
##  [5,] 0.05 0.05 0.12 0.31 1.00 0.29 0.11 0.10 0.07  0.04
##  [6,] 0.03 0.05 0.10 0.16 0.29 1.00 0.25 0.17 0.08  0.06
##  [7,] 0.02 0.01 0.06 0.07 0.11 0.25 1.00 0.23 0.13  0.07
##  [8,] 0.01 0.01 0.06 0.07 0.10 0.17 0.23 1.00 0.28  0.14
##  [9,] 0.01 0.01 0.02 0.03 0.07 0.08 0.13 0.28 1.00  0.31
## [10,] 0.00 0.01 0.01 0.04 0.04 0.06 0.07 0.14 0.31  1.00</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colMeans</a></span>(X) <span class="op">/</span><span class="st"> </span><span class="dv">2</span>) <span class="co"># MAF of SNPs</span></code></pre></div>
<pre><code>##  [1] 0.043 0.091 0.154 0.177 0.234 0.283 0.338 0.372 0.436 0.450</code></pre>
<p>Let’s convert this to our format <code>bigSNP</code> in order to write corresponding PLINK files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(bigsnpr)</code></pre></div>
<pre><code>## Le chargement a nécessité le package : bigstatsr</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fake &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snp_fake.html">snp_fake</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(X), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(X))
fake<span class="op">$</span>genotypes[] &lt;-<span class="st"> </span>X
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempfile</a></span>()
<span class="kw"><a href="../reference/snp_writeBed.html">snp_writeBed</a></span>(fake, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(tmp, <span class="st">".bed"</span>))</code></pre></div>
<pre><code>## [1] "/tmp/Rtmpd9ikMt/filecc21da9b98a.bed"</code></pre>
</div>
<div id="pruning" class="section level2">
<h2 class="hasAnchor">
<a href="#pruning" class="anchor"></a>Pruning</h2>
<p>Pruning, as implemented in PLINK, sequentially scan the genomes for pairs of correlated SNPs, <strong>keeping only the one with the higher MAF</strong> (see <a href="https://github.com/chrchang/plink-ng/issues/27">this</a>). Let’s use PLINK pruning on this simulated toy dataset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(glue)
plink &lt;-<span class="st"> </span><span class="kw"><a href="../reference/download_plink.html">download_plink</a></span>()
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/glue/topics/glue">glue</a></span>(<span class="st">"{plink} --bfile {tmp} --indep-pairwise 50 1 0.2 --out {tmp}"</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/glue/topics/glue">glue</a></span>(<span class="st">"{tmp}.prune.out"</span>))</code></pre></div>
<pre><code>##      V1
## 1 snp_1
## 2 snp_2
## 3 snp_3
## 4 snp_4
## 5 snp_5
## 6 snp_6
## 7 snp_7
## 8 snp_8
## 9 snp_9</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/glue/topics/glue">glue</a></span>(<span class="st">"{tmp}.prune.in"</span>))</code></pre></div>
<pre><code>##       V1
## 1 snp_10</code></pre>
<p>The first SNP is pruned because of its correlation with the second. The second SNP is pruned because of its correlation with the third and so on. In the end, only the last SNP (10th) is kept with the LD pruning procedure of PLINK, which corresponds to less than 18% of the total variance.</p>
<p>You can also do the pruning directly in R with <code><a href="../reference/pruning-clumping.html">snp_pruning()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/pruning-clumping.html">snp_pruning</a></span>(fake<span class="op">$</span>genotypes, <span class="dt">infos.chr =</span> fake<span class="op">$</span>map<span class="op">$</span>chromosome)</code></pre></div>
<pre><code>## [1] 10</code></pre>
</div>
<div id="clumping" class="section level2">
<h2 class="hasAnchor">
<a href="#clumping" class="anchor"></a>Clumping</h2>
<p>A clumping approach would consider the SNPs sorted (in a decreasing order) by a statistic. This statistic is often a test statistic computed from a GWAS of a given phenotype. Yet, for example, for Principal Components Analysis (PCA), the thinning procedure should remain unsupervised (phenotype mustn’t be used!). So, we propose to also use the MAF as a statistic of importance, so that SNPs are sorted by decreasing MAF.</p>
<p>The benefit of the clumping procedure is that the index SNP is always the SNP that is kept (because it has the highest MAF). Let’s use PLINK again:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Compute MAFs</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">SNP =</span> fake<span class="op">$</span>map<span class="op">$</span>marker.ID,
                       <span class="dt">P =</span> <span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="kw"><a href="../reference/snp_MAF.html">snp_MAF</a></span>(fake<span class="op">$</span>genotypes)),
            <span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(tmp, <span class="st">".frq"</span>), <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>)
<span class="co"># Clumping</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/glue/topics/glue">glue</a></span>(<span class="st">"{plink} --bfile {tmp} --out {tmp}"</span>,
              <span class="st">" --clump {tmp}.frq"</span>,
              <span class="st">" --clump-p1 1 --clump-p2 1 --clump-r2 0.2"</span>))
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/glue/topics/glue">glue</a></span>(<span class="st">"{tmp}.clumped"</span>), <span class="dt">header =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>##   CHR F    SNP BP     P TOTAL NSIG S05 S01 S001 S0001      SP2
## 1   1 1 snp_10  0 0.550     1    1   0   0    0     0 snp_9(1)
## 2   1 1  snp_8  0 0.628     1    1   0   0    0     0 snp_7(1)
## 3   1 1  snp_6  0 0.717     1    1   0   0    0     0 snp_5(1)
## 4   1 1  snp_4  0 0.823     1    1   0   0    0     0 snp_3(1)
## 5   1 1  snp_2  0 0.909     1    1   0   0    0     0 snp_1(1)</code></pre>
<p>So the last SNP (10th) would be considered first and the 9th SNP would be pruned. Then the 8th SNP would be considered and the 7th SNP would be pruned and so on. So, the even SNPs would be kept and the odd SNPs would be pruned, which corresponds to 56.7% of the total variance.</p>
<p>You can also do the clumping (on MAF by default) directly in R with <code><a href="../reference/pruning-clumping.html">snp_clumping()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/pruning-clumping.html">snp_clumping</a></span>(fake<span class="op">$</span>genotypes, <span class="dt">infos.chr =</span> fake<span class="op">$</span>map<span class="op">$</span>chromosome)</code></pre></div>
<pre><code>## [1]  2  4  6  8 10</code></pre>
</div>
<div id="conclusion" class="section level2">
<h2 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h2>
<p>So, pruning removes SNPs in a way that may leave regions of the genome with no representative SNP at all.</p>
<!-- ## Acknowledgements  -->
<!-- I thank Michael G.B. Blum and Hugues Aschard, my two thesis supervisors, for thoughtful comments. -->
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#simulation">Simulation</a></li>
      <li><a href="#pruning">Pruning</a></li>
      <li><a href="#clumping">Clumping</a></li>
      <li><a href="#conclusion">Conclusion</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
