<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Projecting PCA — bed_projectPCA • bigsnpr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Projecting PCA — bed_projectPCA" />
<meta property="og:description" content="Computing and projecting PCA of reference dataset to a target dataset." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.6.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Computing polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fas fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Projecting PCA</h1>
    <small class="dont-index">Source: <a href='https://github.com/privefl/bigsnpr/blob/master/R/bed-projectPCA.R'><code>R/bed-projectPCA.R</code></a></small>
    <div class="hidden name"><code>bed_projectPCA.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Computing and projecting PCA of reference dataset to a target dataset.</p>
    </div>

    <pre class="usage"><span class='fu'>bed_projectPCA</span>(
  <span class='no'>obj.bed.ref</span>,
  <span class='no'>obj.bed.new</span>,
  <span class='kw'>k</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>ind.row.new</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>obj.bed.new</span>),
  <span class='kw'>ind.row.ref</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>obj.bed.ref</span>),
  <span class='kw'>ind.col.ref</span> <span class='kw'>=</span> <span class='fu'>cols_along</span>(<span class='no'>obj.bed.ref</span>),
  <span class='kw'>strand_flip</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>join_by_pos</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>match.min.prop</span> <span class='kw'>=</span> <span class='fl'>0.5</span>,
  <span class='kw'>build.new</span> <span class='kw'>=</span> <span class='st'>"hg19"</span>,
  <span class='kw'>build.ref</span> <span class='kw'>=</span> <span class='st'>"hg19"</span>,
  <span class='kw'>liftOver</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>obj.bed.ref</th>
      <td><p>Object of type bed, which is the mapping of the bed file of
the reference data. Use <code>obj.bed &lt;- bed(bedfile)</code> to get this object.</p></td>
    </tr>
    <tr>
      <th>obj.bed.new</th>
      <td><p>Object of type bed, which is the mapping of the bed file of
the target data. Use <code>obj.bed &lt;- bed(bedfile)</code> to get this object.</p></td>
    </tr>
    <tr>
      <th>k</th>
      <td><p>Number of principal components to compute and project.</p></td>
    </tr>
    <tr>
      <th>ind.row.new</th>
      <td><p>Rows to be used in the target data. Default uses them all.</p></td>
    </tr>
    <tr>
      <th>ind.row.ref</th>
      <td><p>Rows to be used in the reference data.
Default uses them all.</p></td>
    </tr>
    <tr>
      <th>ind.col.ref</th>
      <td><p>Columns to be potentially used in the reference data.
Default uses all the ones in common with target data.</p></td>
    </tr>
    <tr>
      <th>strand_flip</th>
      <td><p>Whether to try to flip strand? (default is <code>TRUE</code>)
If so, ambiguous alleles A/T and C/G are removed.</p></td>
    </tr>
    <tr>
      <th>join_by_pos</th>
      <td><p>Whether to join by chromosome and position (default),
or instead by rsid.</p></td>
    </tr>
    <tr>
      <th>match.min.prop</th>
      <td><p>Minimum proportion of variants in the smallest data
to be matched, otherwise stops with an error. Default is <code>50%</code>.</p></td>
    </tr>
    <tr>
      <th>build.new</th>
      <td><p>Genome build of the target data. Default is <code>hg19</code>.</p></td>
    </tr>
    <tr>
      <th>build.ref</th>
      <td><p>Genome build of the reference data. Default is <code>hg19</code>.</p></td>
    </tr>
    <tr>
      <th>liftOver</th>
      <td><p>Path to liftOver executable. Binaries can be downloaded at
<a href='https://hgdownload.cse.ucsc.edu/admin/exe/macOSX.x86_64/liftOver'>https://hgdownload.cse.ucsc.edu/admin/exe/macOSX.x86_64/liftOver</a> for Mac
and at <a href='https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/liftOver'>https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/liftOver</a>
for Linux.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Arguments passed on to <code><a href='snp_autoSVD.html'>bed_autoSVD</a></code></p><dl'>
    <dt><code>fun.scaling</code></dt><dd><p>A function that returns a named list of <code>mean</code> and <code>sd</code>
for every column, to scale each of their elements such as followed:
$$\frac{X_{i,j} - mean_j}{sd_j}.$$ Default is <code><a href='snp_scaleBinom.html'>snp_scaleBinom()</a></code>.</p></dd>
    <dt><code>roll.size</code></dt><dd><p>Radius of rolling windows to smooth log-p-values.
Default is <code>50</code>.</p></dd>
    <dt><code>int.min.size</code></dt><dd><p>Minimum number of consecutive outlier SNPs
in order to be reported as long-range LD region. Default is <code>20</code>.</p></dd>
    <dt><code>thr.r2</code></dt><dd><p>Threshold over the squared correlation between two SNPs.
Default is <code>0.2</code>. Use <code>NA</code> if you want to skip the clumping step.</p></dd>
    <dt><code>alpha.tukey</code></dt><dd><p>Default is <code>0.1</code>. The type-I error rate in outlier
detection (that is further corrected for multiple testing).</p></dd>
    <dt><code>min.mac</code></dt><dd><p>Minimum minor allele count (MAC) for variants to be included.
Default is <code>10</code>.</p></dd>
    <dt><code>max.iter</code></dt><dd><p>Maximum number of iterations of outlier detection.
Default is <code>5</code>.</p></dd>
    <dt><code>size</code></dt><dd><p>For one SNP, window size around this SNP to compute correlations.
Default is <code>100 / thr.r2</code> for clumping (0.2 -&gt; 500; 0.1 -&gt; 1000; 0.5 -&gt; 200).
If not providing <code>infos.pos</code> (<code>NULL</code>, the default), this is a window in
number of SNPs, otherwise it is a window in kb (genetic distance).
I recommend that you provide the positions if available.</p></dd>
  
</dl></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Output some information on the iterations? Default is <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>Number of cores used. Default doesn't use parallelism.
You may use nb_cores.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of 3 elements:</p><ul>
<li><p><code>$obj.svd.ref</code>: big_SVD object computed from reference data.</p></li>
<li><p><code>$simple_proj</code>: simple projection of new data into space of reference PCA.</p></li>
<li><p><code>$OADP_proj</code>: Online Augmentation, Decomposition, and Procrustes (OADP)
projection of new data into space of reference PCA.</p></li>
</ul>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


