<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ bigsnpr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.20</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/PLR.html">Polygenic scores using individual-level data</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
    <li>
      <a href="https://privefl.github.io/bigsnpr-extdoc/" class="external-link">Extended documentation for bigsnpr &amp; bigstatsr</a>
    </li>
    <li>
      <a href="https://privefl.github.io/statgen-course/" class="external-link">Statistical Human Genetics course using R</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="class-bigsnp">Class bigSNP <a href="#class-bigsnp" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="bigSNP-class.html">bigSNP-class</a></code> <code><a href="bigSNP-class.html">bigSNP</a></code> </p>
        </td>
        <td><p>Class bigSNP</p></td>
      </tr><tr><td>
          <p><code><a href="snp_subset.html">snp_subset()</a></code> <code><a href="snp_subset.html">subset(<i>&lt;bigSNP&gt;</i>)</a></code> </p>
        </td>
        <td><p>Subset a bigSNP</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="i-o">I/O <a href="#i-o" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="snp_readBed.html">snp_readBed()</a></code> <code><a href="snp_readBed.html">snp_readBed2()</a></code> </p>
        </td>
        <td><p>Read PLINK files into a "bigSNP"</p></td>
      </tr><tr><td>
          <p><code><a href="snp_writeBed.html">snp_writeBed()</a></code> </p>
        </td>
        <td><p>Write PLINK files from a "bigSNP"</p></td>
      </tr><tr><td>
          <p><code><a href="snp_readBGEN.html">snp_readBGEN()</a></code> </p>
        </td>
        <td><p>Read BGEN files into a "bigSNP"</p></td>
      </tr><tr><td>
          <p><code><a href="snp_readBGI.html">snp_readBGI()</a></code> </p>
        </td>
        <td><p>Read variant info from one BGI file</p></td>
      </tr><tr><td>
          <p><code><a href="snp_attach.html">snp_attach()</a></code> </p>
        </td>
        <td><p>Attach a "bigSNP" from backing files</p></td>
      </tr><tr><td>
          <p><code><a href="snp_attachExtdata.html">snp_attachExtdata()</a></code> </p>
        </td>
        <td><p>Attach a "bigSNP" for examples and tests</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="qc-relatedness-imputation">QC / relatedness / imputation <a href="#qc-relatedness-imputation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="download_plink.html">download_plink()</a></code> <code><a href="download_plink.html">download_plink2()</a></code> </p>
        </td>
        <td><p>Download PLINK</p></td>
      </tr><tr><td>
          <p><code><a href="download_1000G.html">download_1000G()</a></code> </p>
        </td>
        <td><p>Download 1000G</p></td>
      </tr><tr><td>
          <p><code><a href="snp_plinkQC.html">snp_plinkQC()</a></code> </p>
        </td>
        <td><p>Quality Control</p></td>
      </tr><tr><td>
          <p><code><a href="snp_plinkKINGQC.html">snp_plinkKINGQC()</a></code> </p>
        </td>
        <td><p>Relationship-based pruning</p></td>
      </tr><tr><td>
          <p><code><a href="snp_plinkIBDQC.html">snp_plinkIBDQC()</a></code> </p>
        </td>
        <td><p>Identity-by-descent</p></td>
      </tr><tr><td>
          <p><code><a href="snp_plinkRmSamples.html">snp_plinkRmSamples()</a></code> </p>
        </td>
        <td><p>Remove samples</p></td>
      </tr><tr><td>
          <p><code><a href="download_beagle.html">download_beagle()</a></code> </p>
        </td>
        <td><p>Download Beagle 4.1</p></td>
      </tr><tr><td>
          <p><code><a href="snp_beagleImpute.html">snp_beagleImpute()</a></code> </p>
        </td>
        <td><p>Imputation</p></td>
      </tr><tr><td>
          <p><code><a href="snp_fastImpute.html">snp_fastImpute()</a></code> </p>
        </td>
        <td><p>Fast imputation</p></td>
      </tr><tr><td>
          <p><code><a href="snp_fastImputeSimple.html">snp_fastImputeSimple()</a></code> </p>
        </td>
        <td><p>Fast imputation</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="pruning-correlation">Pruning / correlation <a href="#pruning-correlation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="snp_clumping.html">bed_clumping()</a></code> <code><a href="snp_clumping.html">snp_clumping()</a></code> <code><a href="snp_clumping.html">snp_pruning()</a></code> <code><a href="snp_clumping.html">snp_indLRLDR()</a></code> </p>
        </td>
        <td><p>LD clumping</p></td>
      </tr><tr><td>
          <p><code><a href="snp_autoSVD.html">snp_autoSVD()</a></code> <code><a href="snp_autoSVD.html">bed_autoSVD()</a></code> </p>
        </td>
        <td><p>Truncated SVD while limiting LD</p></td>
      </tr><tr><td>
          <p><code><a href="snp_cor.html">snp_cor()</a></code> <code><a href="snp_cor.html">bed_cor()</a></code> </p>
        </td>
        <td><p>Correlation matrix</p></td>
      </tr><tr><td>
          <p><code><a href="snp_ld_scores.html">snp_ld_scores()</a></code> <code><a href="snp_ld_scores.html">bed_ld_scores()</a></code> </p>
        </td>
        <td><p>LD scores</p></td>
      </tr><tr><td>
          <p><code><a href="snp_ldsc.html">snp_ldsc()</a></code> <code><a href="snp_ldsc.html">snp_ldsc2()</a></code> </p>
        </td>
        <td><p>LD score regression</p></td>
      </tr><tr><td>
          <p><code><a href="snp_ldsplit.html">snp_ldsplit()</a></code> </p>
        </td>
        <td><p>Independent LD blocks</p></td>
      </tr><tr><td>
          <p><code><a href="LD.wiki34.html">LD.wiki34</a></code> </p>
        </td>
        <td><p>Long-range LD regions</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="polygenic-scores">Polygenic scores <a href="#polygenic-scores" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="snp_match.html">snp_match()</a></code> </p>
        </td>
        <td><p>Match alleles</p></td>
      </tr><tr><td>
          <p><code><a href="SCT.html">snp_grid_clumping()</a></code> <code><a href="SCT.html">snp_grid_PRS()</a></code> <code><a href="SCT.html">snp_grid_stacking()</a></code> </p>
        </td>
        <td><p>Stacked C+T (SCT)</p></td>
      </tr><tr><td>
          <p><code><a href="LDpred2.html">snp_ldpred2_inf()</a></code> <code><a href="LDpred2.html">snp_ldpred2_grid()</a></code> <code><a href="LDpred2.html">snp_ldpred2_auto()</a></code> </p>
        </td>
        <td><p>LDpred2</p></td>
      </tr><tr><td>
          <p><code><a href="snp_lassosum2.html">snp_lassosum2()</a></code> </p>
        </td>
        <td><p>lassosum2</p></td>
      </tr><tr><td>
          <p><code><a href="snp_PRS.html">snp_PRS()</a></code> </p>
        </td>
        <td><p>PRS</p></td>
      </tr><tr><td>
          <p><code><a href="snp_thr_correct.html">snp_thr_correct()</a></code> </p>
        </td>
        <td><p>Thresholding and correction</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="multiple-testing-plots">Multiple testing - plots <a href="#multiple-testing-plots" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="snp_gc.html">snp_gc()</a></code> </p>
        </td>
        <td><p>Genomic Control</p></td>
      </tr><tr><td>
          <p><code><a href="snp_qq.html">snp_qq()</a></code> </p>
        </td>
        <td><p>Q-Q plot</p></td>
      </tr><tr><td>
          <p><code><a href="snp_manhattan.html">snp_manhattan()</a></code> </p>
        </td>
        <td><p>Manhattan plot</p></td>
      </tr><tr><td>
          <p><code><a href="snp_pcadapt.html">snp_pcadapt()</a></code> <code><a href="snp_pcadapt.html">bed_pcadapt()</a></code> </p>
        </td>
        <td><p>Outlier detection</p></td>
      </tr><tr><td>
          <p><code><a href="snp_MAX3.html">snp_MAX3()</a></code> </p>
        </td>
        <td><p>MAX3 statistic</p></td>
      </tr><tr><td>
          <p><code><a href="snp_fst.html">snp_fst()</a></code> </p>
        </td>
        <td><p>Fixation index (Fst)</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="utilities">Utilities <a href="#utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="snp_prodBGEN.html">snp_prodBGEN()</a></code> </p>
        </td>
        <td><p>BGEN matrix product</p></td>
      </tr><tr><td>
          <p><code><a href="snp_match.html">snp_match()</a></code> </p>
        </td>
        <td><p>Match alleles</p></td>
      </tr><tr><td>
          <p><code><a href="snp_asGeneticPos.html">snp_asGeneticPos()</a></code> </p>
        </td>
        <td><p>Interpolate to genetic positions</p></td>
      </tr><tr><td>
          <p><code><a href="snp_asGeneticPos2.html">download_genetic_map()</a></code> <code><a href="snp_asGeneticPos2.html">snp_asGeneticPos2()</a></code> </p>
        </td>
        <td><p>Download a genetic map</p></td>
      </tr><tr><td>
          <p><code><a href="snp_ancestry_summary.html">snp_ancestry_summary()</a></code> </p>
        </td>
        <td><p>Estimation of ancestry proportions</p></td>
      </tr><tr><td>
          <p><code><a href="snp_simuPheno.html">snp_simuPheno()</a></code> </p>
        </td>
        <td><p>Simulate phenotypes</p></td>
      </tr><tr><td>
          <p><code><a href="snp_modifyBuild.html">snp_modifyBuild()</a></code> </p>
        </td>
        <td><p>Modify genome build</p></td>
      </tr><tr><td>
          <p><code><a href="snp_split.html">snp_split()</a></code> </p>
        </td>
        <td><p>Split-parApply-Combine</p></td>
      </tr><tr><td>
          <p><code><a href="snp_save.html">snp_save()</a></code> </p>
        </td>
        <td><p>Save modifications</p></td>
      </tr><tr><td>
          <p><code><a href="snp_getSampleInfos.html">snp_getSampleInfos()</a></code> </p>
        </td>
        <td><p>Get sample information</p></td>
      </tr><tr><td>
          <p><code><a href="snp_MAF.html">snp_MAF()</a></code> </p>
        </td>
        <td><p>MAF</p></td>
      </tr><tr><td>
          <p><code><a href="snp_scaleBinom.html">snp_scaleAlpha()</a></code> <code><a href="snp_scaleBinom.html">snp_scaleBinom()</a></code> </p>
        </td>
        <td><p>Binomial(n, p) scaling</p></td>
      </tr><tr><td>
          <p><code><a href="same_ref.html">same_ref()</a></code> </p>
        </td>
        <td><p>Determine reference divergence</p></td>
      </tr><tr><td>
          <p><code><a href="sub_bed.html">sub_bed()</a></code> </p>
        </td>
        <td><p>Replace extension '.bed'</p></td>
      </tr><tr><td>
          <p><code><a href="seq_log.html">seq_log()</a></code> </p>
        </td>
        <td><p>Sequence, evenly spaced on a logarithmic scale</p></td>
      </tr><tr><td>
          <p><code><a href="coef_to_liab.html">coef_to_liab()</a></code> </p>
        </td>
        <td><p>Liability scale</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="plink-bed-files">PLINK bed files <a href="#plink-bed-files" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="bed-class.html">bed()</a></code> </p>
        </td>
        <td><p>Class bed</p></td>
      </tr><tr><td>
          <p><code><a href="bed_MAF.html">bed_MAF()</a></code> </p>
        </td>
        <td><p>Allele frequencies</p></td>
      </tr><tr><td>
          <p><code><a href="snp_autoSVD.html">snp_autoSVD()</a></code> <code><a href="snp_autoSVD.html">bed_autoSVD()</a></code> </p>
        </td>
        <td><p>Truncated SVD while limiting LD</p></td>
      </tr><tr><td>
          <p><code><a href="snp_clumping.html">bed_clumping()</a></code> <code><a href="snp_clumping.html">snp_clumping()</a></code> <code><a href="snp_clumping.html">snp_pruning()</a></code> <code><a href="snp_clumping.html">snp_indLRLDR()</a></code> </p>
        </td>
        <td><p>LD clumping</p></td>
      </tr><tr><td>
          <p><code><a href="bed_counts.html">bed_counts()</a></code> </p>
        </td>
        <td><p>Counts</p></td>
      </tr><tr><td>
          <p><code><a href="bed_cprodVec.html">bed_cprodVec()</a></code> </p>
        </td>
        <td><p>Cross-product with a vector</p></td>
      </tr><tr><td>
          <p><code><a href="snp_pcadapt.html">snp_pcadapt()</a></code> <code><a href="snp_pcadapt.html">bed_pcadapt()</a></code> </p>
        </td>
        <td><p>Outlier detection</p></td>
      </tr><tr><td>
          <p><code><a href="bed_prodVec.html">bed_prodVec()</a></code> </p>
        </td>
        <td><p>Product with a vector</p></td>
      </tr><tr><td>
          <p><code><a href="bed_projectPCA.html">bed_projectPCA()</a></code> </p>
        </td>
        <td><p>Projecting PCA</p></td>
      </tr><tr><td>
          <p><code><a href="bed_projectSelfPCA.html">bed_projectSelfPCA()</a></code> <code><a href="bed_projectSelfPCA.html">snp_projectSelfPCA()</a></code> </p>
        </td>
        <td><p>Projecting PCA</p></td>
      </tr><tr><td>
          <p><code><a href="bed_randomSVD.html">bed_randomSVD()</a></code> </p>
        </td>
        <td><p>Randomized partial SVD</p></td>
      </tr><tr><td>
          <p><code><a href="bed_scaleBinom.html">bed_scaleBinom()</a></code> </p>
        </td>
        <td><p>Binomial(2, p) scaling</p></td>
      </tr><tr><td>
          <p><code><a href="bed_tcrossprodSelf.html">bed_tcrossprodSelf()</a></code> </p>
        </td>
        <td><p>tcrossprod / GRM</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Florian PrivÃ©.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

