<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>LDpred2 — snp_ldpred2_inf • bigsnpr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="LDpred2 — snp_ldpred2_inf"><meta property="og:description" content="LDpred2. Tutorial at https://privefl.github.io/bigsnpr/articles/LDpred2.html."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>LDpred2</h1>
    <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/R/LDpred2.R" class="external-link"><code>R/LDpred2.R</code></a></small>
    <div class="hidden name"><code>LDpred2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>LDpred2. Tutorial at <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html" class="external-link">https://privefl.github.io/bigsnpr/articles/LDpred2.html</a>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">snp_ldpred2_inf</span><span class="op">(</span><span class="va">corr</span>, <span class="va">df_beta</span>, <span class="va">h2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">snp_ldpred2_grid</span><span class="op">(</span></span>
<span>  <span class="va">corr</span>,</span>
<span>  <span class="va">df_beta</span>,</span>
<span>  <span class="va">grid_param</span>,</span>
<span>  burn_in <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  num_iter <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  return_sampling_betas <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ind.corr <span class="op">=</span> <span class="fu">cols_along</span><span class="op">(</span><span class="va">corr</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">snp_ldpred2_auto</span><span class="op">(</span></span>
<span>  <span class="va">corr</span>,</span>
<span>  <span class="va">df_beta</span>,</span>
<span>  <span class="va">h2_init</span>,</span>
<span>  vec_p_init <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  burn_in <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  num_iter <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  sparse <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  report_step <span class="op">=</span> <span class="va">num_iter</span> <span class="op">+</span> <span class="fl">1L</span>,</span>
<span>  allow_jump_sign <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  shrink_corr <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  use_MLE <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  p_bounds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e-05</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  alpha_bounds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>  ind.corr <span class="op">=</span> <span class="fu">cols_along</span><span class="op">(</span><span class="va">corr</span><span class="op">)</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>corr</dt>
<dd><p>Sparse correlation matrix as an SFBM.
If <code>corr</code> is a dsCMatrix or a dgCMatrix, you can use <code>as_SFBM(corr)</code>.</p></dd>


<dt>df_beta</dt>
<dd><p>A data frame with 3 columns:</p><ul><li><p><code>$beta</code>: effect size estimates</p></li>
<li><p><code>$beta_se</code>: standard errors of effect size estimates</p></li>
<li><p><code>$n_eff</code>: either tha GWAS sample size(s) when estimating <code>beta</code> for a
continuous trait, or in the case of a binary trait, this is
<code>4 / (1 / n_control + 1 / n_case)</code>; in the case of a meta-analysis, you
should sum the effective sample sizes of each study instead of using the
total numbers of cases and controls, see <a href="https://doi.org/10.1016/j.biopsych.2022.05.029" class="external-link">doi:10.1016/j.biopsych.2022.05.029</a>
;
when using a mixed model, the effective sample size needs to be adjusted
as well, see <a href="https://doi.org/10.1016/j.xhgg.2022.100136" class="external-link">doi:10.1016/j.xhgg.2022.100136</a>
.</p></li>
</ul></dd>


<dt>h2</dt>
<dd><p>Heritability estimate.</p></dd>


<dt>grid_param</dt>
<dd><p>A data frame with 3 columns as a grid of hyper-parameters:</p><ul><li><p><code>$p</code>: proportion of causal variants</p></li>
<li><p><code>$h2</code>: heritability (captured by the variants used)</p></li>
<li><p><code>$sparse</code>: boolean, whether a sparse model is sought
They can be run in parallel by changing <code>ncores</code>.</p></li>
</ul></dd>


<dt>burn_in</dt>
<dd><p>Number of burn-in iterations.</p></dd>


<dt>num_iter</dt>
<dd><p>Number of iterations after burn-in.</p></dd>


<dt>ncores</dt>
<dd><p>Number of cores used. Default doesn't use parallelism.
You may use <code>nb_cores()</code>.</p></dd>


<dt>return_sampling_betas</dt>
<dd><p>Whether to return all sampling betas (after
burn-in)? This is useful for assessing the uncertainty of the PRS at the
individual level (see <a href="https://doi.org/10.1101/2020.11.30.403188" class="external-link">doi:10.1101/2020.11.30.403188</a>
).
Default is <code>FALSE</code> (only returns the averaged final vectors of betas).
If <code>TRUE</code>, only one set of parameters is allowed.</p></dd>


<dt>ind.corr</dt>
<dd><p>Indices to "subset" <code>corr</code>, as if this was run with
<code>corr[ind.corr, ind.corr]</code> instead. No subsetting by default.</p></dd>


<dt>h2_init</dt>
<dd><p>Heritability estimate for initialization.</p></dd>


<dt>vec_p_init</dt>
<dd><p>Vector of initial values for p. Default is <code>0.1</code>.</p></dd>


<dt>sparse</dt>
<dd><p>In LDpred2-auto, whether to also report a sparse solution by
running LDpred2-grid with the estimates of p and h2 from LDpred2-auto, and
sparsity enabled. Default is <code>FALSE</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Whether to print "p // h2" estimates at each iteration.
Disabled when parallelism is used.</p></dd>


<dt>report_step</dt>
<dd><p>Step to report sampling betas (after burn-in and before
unscaling). Nothing is reported by default. If using <code>num_iter = 200</code> and
<code>report_step = 20</code>, then 10 vectors of sampling betas are reported
(as a sparse matrix with 10 columns).</p></dd>


<dt>allow_jump_sign</dt>
<dd><p>Whether to allow for effects sizes to change sign in
consecutive iterations? Default is <code>TRUE</code> (normal sampling). You can use
<code>FALSE</code> to force effects to go through 0 first before changing sign. Setting
this parameter to <code>FALSE</code> could be useful to prevent instability (oscillation
and ultimately divergence) of the Gibbs sampler. This would also be useful
for accelerating convergence of chains with a large initial value for p.</p></dd>


<dt>shrink_corr</dt>
<dd><p>Shrinkage multiplicative coefficient to apply to off-diagonal
elements of the correlation matrix. Default is <code>1</code> (unchanged).
You can use e.g. <code>0.95</code> to add a bit of regularization.</p></dd>


<dt>use_MLE</dt>
<dd><p>Whether to use maximum likelihood estimation (MLE) to estimate
alpha and the variance component (since v1.11.4), or assume that alpha is
-1 and estimate the variance of (scaled) effects as h2/(m*p), as it was
done in earlier versions of LDpred2-auto (e.g. in v1.10.8). Default is <code>TRUE</code>,
which should provide a better model fit, but might also be less robust.</p></dd>


<dt>p_bounds</dt>
<dd><p>Boundaries for the estimates of p (the polygenicity).
Default is <code>c(1e-5, 1)</code>. You can use the same value twice to fix p.</p></dd>


<dt>alpha_bounds</dt>
<dd><p>Boundaries for the estimates of \(\alpha\).
Default is <code>c(-1.5, 0.5)</code>. You can use the same value twice to fix \(\alpha\).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>snp_ldpred2_inf</code>: A vector of effects, assuming an infinitesimal model.</p>


<p><code>snp_ldpred2_grid</code>: A matrix of effect sizes, one vector (column)
for each row of <code>grid_param</code>. Missing values are returned when strong
divergence is detected. If using <code>return_sampling_betas</code>, each column
corresponds to one iteration instead (after burn-in).</p>


<p><code>snp_ldpred2_auto</code>: A list (over <code>vec_p_init</code>) of lists with</p><ul><li><p><code>$beta_est</code>: vector of effect sizes (on the allele scale); note that
missing values are returned when strong divergence is detected</p></li>
<li><p><code>$beta_est_sparse</code> (only when <code>sparse = TRUE</code>): sparse vector of effect sizes</p></li>
<li><p><code>$postp_est</code>: vector of posterior probabilities of being causal</p></li>
<li><p><code>$corr_est</code>, the "imputed" correlations between variants and phenotypes,
which can be used for post-QCing variants by comparing those to
<code>with(df_beta, beta / sqrt(n_eff * beta_se^2 + beta^2))</code></p></li>
<li><p><code>$sample_beta</code>: sparse matrix of sampling betas (see parameter <code>report_step</code>),
<em>not</em> on the allele scale, for which you need to multiply by
<code>with(df_beta, sqrt(n_eff * beta_se^2 + beta^2))</code></p></li>
<li><p><code>$path_p_est</code>: full path of p estimates (including burn-in);
useful to check convergence of the iterative algorithm</p></li>
<li><p><code>$path_h2_est</code>: full path of h2 estimates (including burn-in);
useful to check convergence of the iterative algorithm</p></li>
<li><p><code>$path_alpha_est</code>: full path of alpha estimates (including burn-in);
useful to check convergence of the iterative algorithm</p></li>
<li><p><code>$h2_est</code>: estimate of the (SNP) heritability (also see <a href="coef_to_liab.html">coef_to_liab</a>)</p></li>
<li><p><code>$p_est</code>: estimate of p, the proportion of causal variants</p></li>
<li><p><code>$alpha_est</code>: estimate of alpha, the parameter controlling the
relationship between allele frequencies and expected effect sizes</p></li>
<li><p><code>$h2_init</code> and <code>$p_init</code>: input parameters, for convenience</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>For reproducibility, <code><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed()</a></code> can be used to ensure that two runs of
LDpred2 give the exact same results (since v1.10).</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

