<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Stacked C+T (SCT) — SCT • bigsnpr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Stacked C+T (SCT) — SCT"><meta property="og:description" content="Polygenic Risk Scores for a grid of clumping and thresholding parameters.
Stacking over many Polygenic Risk Scores, corresponding to a grid of many
different parameters for clumping and thresholding."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.20</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/PLR.html">Polygenic scores using individual-level data</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
    <li>
      <a href="https://privefl.github.io/bigsnpr-extdoc/" class="external-link">Extended documentation for bigsnpr &amp; bigstatsr</a>
    </li>
    <li>
      <a href="https://privefl.github.io/statgen-course/" class="external-link">Statistical Human Genetics course using R</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Stacked C+T (SCT)</h1>
    <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/R/SCT.R" class="external-link"><code>R/SCT.R</code></a></small>
    <div class="hidden name"><code>SCT.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Polygenic Risk Scores for a grid of clumping and thresholding parameters.</p>
<p>Stacking over many Polygenic Risk Scores, corresponding to a grid of many
different parameters for clumping and thresholding.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">snp_grid_clumping</span><span class="op">(</span></span>
<span>  <span class="va">G</span>,</span>
<span>  <span class="va">infos.chr</span>,</span>
<span>  <span class="va">infos.pos</span>,</span>
<span>  <span class="va">lpS</span>,</span>
<span>  ind.row <span class="op">=</span> <span class="fu">rows_along</span><span class="op">(</span><span class="va">G</span><span class="op">)</span>,</span>
<span>  grid.thr.r2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="fl">0.95</span><span class="op">)</span>,</span>
<span>  grid.base.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">500</span><span class="op">)</span>,</span>
<span>  infos.imp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  grid.thr.imp <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu">cols_along</span><span class="op">(</span><span class="va">G</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">snp_grid_PRS</span><span class="op">(</span></span>
<span>  <span class="va">G</span>,</span>
<span>  <span class="va">all_keep</span>,</span>
<span>  <span class="va">betas</span>,</span>
<span>  <span class="va">lpS</span>,</span>
<span>  n_thr_lpS <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  grid.lpS.thr <span class="op">=</span> <span class="fl">0.9999</span> <span class="op">*</span> <span class="fu"><a href="seq_log.html">seq_log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">lpS</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">lpS</span>, na.rm <span class="op">=</span></span>
<span>    <span class="cn">TRUE</span><span class="op">)</span>, <span class="va">n_thr_lpS</span><span class="op">)</span>,</span>
<span>  ind.row <span class="op">=</span> <span class="fu">rows_along</span><span class="op">(</span><span class="va">G</span><span class="op">)</span>,</span>
<span>  backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"float"</span>, <span class="st">"double"</span><span class="op">)</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">snp_grid_stacking</span><span class="op">(</span></span>
<span>  <span class="va">multi_PRS</span>,</span>
<span>  <span class="va">y.train</span>,</span>
<span>  alphas <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.01</span>, <span class="fl">1e-04</span><span class="op">)</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>G</dt>
<dd><p>A <a href="https://privefl.github.io/bigstatsr/reference/FBM.code256-class.html" class="external-link">FBM.code256</a>
(typically <code>&lt;bigSNP&gt;$genotypes</code>).<br><strong>You shouldn't have missing values.</strong> Also, remember to do quality control,
e.g. some algorithms in this package won't work if you use SNPs with 0 MAF.</p></dd>


<dt>infos.chr</dt>
<dd><p>Vector of integers specifying each SNP's chromosome.<br>
Typically <code>&lt;bigSNP&gt;$map$chromosome</code>.</p></dd>


<dt>infos.pos</dt>
<dd><p>Vector of integers specifying the physical position
on a chromosome (in base pairs) of each SNP.<br>
Typically <code>&lt;bigSNP&gt;$map$physical.pos</code>.</p></dd>


<dt>lpS</dt>
<dd><p>Numeric vector of <code>-log10(p-value)</code> associated with <code>betas</code>.</p></dd>


<dt>ind.row</dt>
<dd><p>An optional vector of the row indices (individuals) that
are used. If not specified, all rows are used.<br><strong>Don't use negative indices.</strong></p></dd>


<dt>grid.thr.r2</dt>
<dd><p>Grid of thresholds over the squared correlation between
two SNPs for clumping. Default is <code>c(0.01, 0.05, 0.1, 0.2, 0.5, 0.8, 0.95)</code>.</p></dd>


<dt>grid.base.size</dt>
<dd><p>Grid for base window sizes. Sizes are then computed as
<code>base.size / thr.r2</code> (in kb). Default is <code>c(50, 100, 200, 500)</code>.</p></dd>


<dt>infos.imp</dt>
<dd><p>Vector of imputation scores. Default is all <code>1</code> if you do
not provide it.</p></dd>


<dt>grid.thr.imp</dt>
<dd><p>Grid of thresholds over <code>infos.imp</code> (default is <code>1</code>), but
you should change it (e.g. <code>c(0.3, 0.6, 0.9, 0.95)</code>) if providing <code>infos.imp</code>.</p></dd>


<dt>groups</dt>
<dd><p>List of vectors of indices to define your own categories.
This could be used e.g. to derive C+T scores using two different GWAS
summary statistics, or to include other information such as functional
annotations. Default just makes one group with all variants.</p></dd>


<dt>exclude</dt>
<dd><p>Vector of SNP indices to exclude anyway.</p></dd>


<dt>ncores</dt>
<dd><p>Number of cores used. Default doesn't use parallelism.
You may use <code><a href="https://privefl.github.io/bigstatsr/reference/reexports.html" class="external-link">bigstatsr::nb_cores()</a></code>.</p></dd>


<dt>all_keep</dt>
<dd><p>Output of <code>snp_grid_clumping()</code> (indices passing clumping).</p></dd>


<dt>betas</dt>
<dd><p>Numeric vector of weights (effect sizes from GWAS) associated
with each variant (column of <code>G</code>). If alleles are reversed, make sure to
multiply corresponding effects by <code>-1</code>.</p></dd>


<dt>n_thr_lpS</dt>
<dd><p>Length for default <code>grid.lpS.thr</code>. Default is <code>50</code>.</p></dd>


<dt>grid.lpS.thr</dt>
<dd><p>Sequence of thresholds to apply on <code>lpS</code>.
Default is a grid (of length <code>n_thr_lpS</code>) evenly spaced on a logarithmic
scale, i.e. on a log-log scale for p-values.</p></dd>


<dt>backingfile</dt>
<dd><p>Prefix for backingfiles where to store scores of C+T.
As we typically use a large grid, this can result in a large matrix so that
we store it on disk. Default uses a temporary file.</p></dd>


<dt>type</dt>
<dd><p>Type of backingfile values. Either <code>"float"</code> (the default) or
<code>"double"</code>. Using <code>"float"</code> requires half disk space.</p></dd>


<dt>multi_PRS</dt>
<dd><p>Output of <code>snp_grid_PRS()</code>.</p></dd>


<dt>y.train</dt>
<dd><p>Vector of phenotypes. If there are two levels (binary 0/1),
it uses <code><a href="https://privefl.github.io/bigstatsr/reference/big_spLogReg.html" class="external-link">bigstatsr::big_spLogReg()</a></code> for stacking, otherwise <code><a href="https://privefl.github.io/bigstatsr/reference/big_spLinReg.html" class="external-link">bigstatsr::big_spLinReg()</a></code>.</p></dd>


<dt>alphas</dt>
<dd><p>Vector of values for grid-search. See <code><a href="https://privefl.github.io/bigstatsr/reference/big_spLogReg.html" class="external-link">bigstatsr::big_spLogReg()</a></code>.
Default for this function is <code>c(1, 0.01, 0.0001)</code>.</p></dd>


<dt>...</dt>
<dd><p>Other parameters to be passed to <code><a href="https://privefl.github.io/bigstatsr/reference/big_spLogReg.html" class="external-link">bigstatsr::big_spLogReg()</a></code>. For example,
using <code>covar.train</code>, you can add covariates in the model with all C+T scores.
You can also use <code>pf.covar</code> if you do not want to penalize these covariates.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>snp_grid_PRS()</code>: An <code>FBM</code> (matrix on disk) that stores the C+T scores
for all parameters of the grid (and for each chromosome separately).
It also stores as attributes the input parameters <code>all_keep</code>, <code>betas</code>,
<code>lpS</code> and <code>grid.lpS.thr</code> that are also needed in <code>snp_grid_stacking()</code>.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

