<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>LD clumping — bed_clumping • bigsnpr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="LD clumping — bed_clumping"><meta property="og:description" content='For a bigSNP:
snp_pruning(): LD pruning. Similar to "--indep-pairwise (size+1) 1 thr.r2"
in PLINK.
This function is deprecated (see
this article).
snp_clumping() (and bed_clumping()): LD clumping. If you do not provide
any statistic to rank SNPs, it would use minor allele frequencies (MAFs),
making clumping similar to pruning.
snp_indLRLDR(): Get SNP indices of long-range LD regions for the
human genome.

'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>LD clumping</h1>
    <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/R/bed-clumping.R" class="external-link"><code>R/bed-clumping.R</code></a>, <a href="https://github.com/privefl/bigsnpr/blob/HEAD/R/clumping.R" class="external-link"><code>R/clumping.R</code></a></small>
    <div class="hidden name"><code>snp_clumping.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For a <code>bigSNP</code>:</p><ul><li><p><code>snp_pruning()</code>: LD pruning. Similar to "<code>--indep-pairwise (size+1) 1 thr.r2</code>"
in <a href="https://www.cog-genomics.org/plink/1.9/ld" class="external-link">PLINK</a>.
<strong>This function is deprecated (see
<a href="https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html" class="external-link">this article</a>).</strong></p></li>
<li><p><code>snp_clumping()</code> (and <code>bed_clumping()</code>): LD clumping. If you do not provide
any statistic to rank SNPs, it would use minor allele frequencies (MAFs),
making clumping similar to pruning.</p></li>
<li><p><code>snp_indLRLDR()</code>: Get SNP indices of long-range LD regions for the
human genome.</p></li>
</ul></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bed_clumping</span><span class="op">(</span></span>
<span>  <span class="va">obj.bed</span>,</span>
<span>  ind.row <span class="op">=</span> <span class="fu">rows_along</span><span class="op">(</span><span class="va">obj.bed</span><span class="op">)</span>,</span>
<span>  S <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thr.r2 <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">100</span><span class="op">/</span><span class="va">thr.r2</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">snp_clumping</span><span class="op">(</span></span>
<span>  <span class="va">G</span>,</span>
<span>  <span class="va">infos.chr</span>,</span>
<span>  ind.row <span class="op">=</span> <span class="fu">rows_along</span><span class="op">(</span><span class="va">G</span><span class="op">)</span>,</span>
<span>  S <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thr.r2 <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">100</span><span class="op">/</span><span class="va">thr.r2</span>,</span>
<span>  infos.pos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  is.size.in.bp <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">snp_pruning</span><span class="op">(</span></span>
<span>  <span class="va">G</span>,</span>
<span>  <span class="va">infos.chr</span>,</span>
<span>  ind.row <span class="op">=</span> <span class="fu">rows_along</span><span class="op">(</span><span class="va">G</span><span class="op">)</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">49</span>,</span>
<span>  is.size.in.bp <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  infos.pos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thr.r2 <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nploidy <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">snp_indLRLDR</span><span class="op">(</span><span class="va">infos.chr</span>, <span class="va">infos.pos</span>, LD.regions <span class="op">=</span> <span class="va">LD.wiki34</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>obj.bed</dt>
<dd><p>Object of type <a href="bed-class.html">bed</a>, which is the mapping of some bed file.
Use <code>obj.bed &lt;- bed(bedfile)</code> to get this object.</p></dd>


<dt>ind.row</dt>
<dd><p>An optional vector of the row indices (individuals) that
are used. If not specified, all rows are used.<br><strong>Don't use negative indices.</strong></p></dd>


<dt>S</dt>
<dd><p>A vector of column statistics which express the importance
of each SNP (the more important is the SNP, the greater should be
the corresponding statistic).<br>
For example, if <code>S</code> follows the standard normal distribution, and "important"
means significantly different from 0, you must use <code>abs(S)</code> instead.<br><strong>If not specified, MAFs are computed and used.</strong></p></dd>


<dt>thr.r2</dt>
<dd><p>Threshold over the squared correlation between two SNPs.
Default is <code>0.2</code>.</p></dd>


<dt>size</dt>
<dd><p>For one SNP, window size around this SNP to compute correlations.
Default is <code>100 / thr.r2</code> for clumping (0.2 -&gt; 500; 0.1 -&gt; 1000; 0.5 -&gt; 200).
If not providing <code>infos.pos</code> (<code>NULL</code>, the default), this is a window in
number of SNPs, otherwise it is a window in kb (genetic distance).
I recommend that you provide the positions if available.</p></dd>


<dt>exclude</dt>
<dd><p>Vector of SNP indices to exclude anyway. For example,
can be used to exclude long-range LD regions (see Price2008). Another use
can be for thresholding with respect to p-values associated with <code>S</code>.</p></dd>


<dt>ncores</dt>
<dd><p>Number of cores used. Default doesn't use parallelism.
You may use <code><a href="https://privefl.github.io/bigstatsr/reference/reexports.html" class="external-link">bigstatsr::nb_cores()</a></code>.</p></dd>


<dt>G</dt>
<dd><p>A <a href="https://privefl.github.io/bigstatsr/reference/FBM.code256-class.html" class="external-link">FBM.code256</a>
(typically <code>&lt;bigSNP&gt;$genotypes</code>).<br><strong>You shouldn't have missing values.</strong> Also, remember to do quality control,
e.g. some algorithms in this package won't work if you use SNPs with 0 MAF.</p></dd>


<dt>infos.chr</dt>
<dd><p>Vector of integers specifying each SNP's chromosome.<br>
Typically <code>&lt;bigSNP&gt;$map$chromosome</code>.</p></dd>


<dt>infos.pos</dt>
<dd><p>Vector of integers specifying the physical position
on a chromosome (in base pairs) of each SNP.<br>
Typically <code>&lt;bigSNP&gt;$map$physical.pos</code>.</p></dd>


<dt>is.size.in.bp</dt>
<dd><p>Deprecated.</p></dd>


<dt>nploidy</dt>
<dd><p>Number of trials, parameter of the binomial distribution.
Default is <code>2</code>, which corresponds to diploidy, such as for the human genome.</p></dd>


<dt>LD.regions</dt>
<dd><p>A <code>data.frame</code> with columns "Chr", "Start" and "Stop".
Default use <a href="LD.wiki34.html">LD.wiki34</a>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<ul><li><p><code>snp_clumping()</code> (and <code>bed_clumping()</code>): SNP indices that are <strong>kept</strong>.</p></li>
<li><p><code>snp_indLRLDR()</code>: SNP indices to be used as (part of) the '<strong><code>exclude</code></strong>'
parameter of <code>snp_clumping()</code>.</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Price AL, Weale ME, Patterson N, et al.
Long-Range LD Can Confound Genome Scans in Admixed Populations.
Am J Hum Genet. 2008;83(1):132-135. <a href="https://doi.org/10.1016/j.ajhg.2008.06.005" class="external-link">doi:10.1016/j.ajhg.2008.06.005</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="snp_attachExtdata.html">snp_attachExtdata</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">G</span> <span class="op">&lt;-</span> <span class="va">test</span><span class="op">$</span><span class="va">genotypes</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># clumping (prioritizing higher MAF)</span></span></span>
<span class="r-in"><span><span class="va">ind.keep</span> <span class="op">&lt;-</span> <span class="fu">snp_clumping</span><span class="op">(</span><span class="va">G</span>, infos.chr <span class="op">=</span> <span class="va">test</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">chromosome</span>,</span></span>
<span class="r-in"><span>                         infos.pos <span class="op">=</span> <span class="va">test</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">physical.pos</span>,</span></span>
<span class="r-in"><span>                         thr.r2 <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># keep most of them -&gt; not much LD in this simulated dataset</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ind.keep</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.7919419</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

