<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Subset a bigSNP — snp_subset • bigsnpr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Subset a bigSNP — snp_subset"><meta property="og:description" content="Subset (copy) of a bigSNP, also stored on disk."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Subset a bigSNP</h1>
    <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/R/subset-QC.R" class="external-link"><code>R/subset-QC.R</code></a></small>
    <div class="hidden name"><code>snp_subset.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Subset (copy) of a <code>bigSNP</code>, also stored on disk.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">snp_subset</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  ind.row <span class="op">=</span> <span class="fu">rows_along</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">genotypes</span><span class="op">)</span>,</span>
<span>  ind.col <span class="op">=</span> <span class="fu">cols_along</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">genotypes</span><span class="op">)</span>,</span>
<span>  backingfile <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for bigSNP</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  ind.row <span class="op">=</span> <span class="fu">rows_along</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">fam</span><span class="op">)</span>,</span>
<span>  ind.col <span class="op">=</span> <span class="fu">rows_along</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">map</span><span class="op">)</span>,</span>
<span>  backingfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>A <a href="bigSNP-class.html">bigSNP</a>.</p></dd>


<dt>ind.row</dt>
<dd><p>Indices of the rows (individuals) to keep.
Negative indices <strong>can</strong> be used to exclude row indices.
Default: keep them all.</p></dd>


<dt>ind.col</dt>
<dd><p>Indices of the columns (SNPs) to keep.
Negative indices <strong>can</strong> be used to exclude column indices.
Default: keep them all.</p></dd>


<dt>backingfile</dt>
<dd><p>Prefix of the two new files created (".bk" and ".rds").
By default, it is automatically determined by appending "_sub" and a number
to the prefix of the input bigSNP backing files.</p></dd>


<dt>...</dt>
<dd><p>Not used.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>The path to the RDS file that stores the <code>bigSNP</code> object.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="bigSNP-class.html">bigSNP</a></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="snp_attachExtdata.html">snp_attachExtdata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 16 fields</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr      :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr_rw   :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ nrow        : int 517</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ ncol        : int 4542</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type        : Named int 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..- attr(*, "names")= chr "unsigned char"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ backingfile : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpITJXCm\\file5f8c66ca6c85.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_read_only: logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address     :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address_rw  :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ bk          : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpITJXCm\\file5f8c66ca6c85.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ rds         : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpITJXCm\\file5f8c66ca6c85.rds"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_saved    : logi TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_chr    : chr "unsigned char"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_size   : int 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ file_size   : num 2348214</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ code256     : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..and 26 methods, of which 12 are  possibly relevant:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  check_write_permissions, copy#envRefClass, initialize, initialize#FBM,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  save, show#FBM, show#envRefClass</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ fam      :'data.frame':	517 obs. of  6 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ family.ID  : chr [1:517] "POP1" "POP1" "POP1" "POP1" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ sample.ID  : chr [1:517] "IND0" "IND1" "IND2" "IND3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ paternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ maternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ sex        : int [1:517] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ affection  : int [1:517] 1 1 2 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ map      :'data.frame':	4542 obs. of  6 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ chromosome  : int [1:4542] 1 1 1 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ marker.ID   : chr [1:4542] "SNP0" "SNP1" "SNP2" "SNP3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ genetic.dist: int [1:4542] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ physical.pos: int [1:4542] 112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ allele1     : chr [1:4542] "A" "T" "T" "T" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ allele2     : chr [1:4542] "T" "A" "A" "A" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "class")= chr "bigSNP"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># keep only first 50 samples and SNPs</span></span></span>
<span class="r-in"><span><span class="va">rdsfile</span> <span class="op">&lt;-</span> <span class="fu">snp_subset</span><span class="op">(</span><span class="va">test</span>, ind.row <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span>, ind.col <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="va">rdsfile</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 16 fields</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr      :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr_rw   :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ nrow        : int 50</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ ncol        : int 50</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type        : Named int 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..- attr(*, "names")= chr "unsigned char"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ backingfile : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpITJXCm\\file5f8c66ca6c85_sub1.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_read_only: logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address     :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address_rw  :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ bk          : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpITJXCm\\file5f8c66ca6c85_sub1.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ rds         : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpITJXCm\\file5f8c66ca6c85_sub1.rds"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_saved    : logi TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_chr    : chr "unsigned char"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_size   : int 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ file_size   : num 2500</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ code256     : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..and 26 methods, of which 12 are  possibly relevant:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  check_write_permissions, copy#envRefClass, initialize, initialize#FBM,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  save, show#FBM, show#envRefClass</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ fam      :'data.frame':	50 obs. of  6 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ family.ID  : chr [1:50] "POP1" "POP1" "POP1" "POP1" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ sample.ID  : chr [1:50] "IND0" "IND1" "IND2" "IND3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ paternal.ID: int [1:50] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ maternal.ID: int [1:50] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ sex        : int [1:50] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ affection  : int [1:50] 1 1 2 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ map      :'data.frame':	50 obs. of  6 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ chromosome  : int [1:50] 1 1 1 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ marker.ID   : chr [1:50] "SNP0" "SNP1" "SNP2" "SNP3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ genetic.dist: int [1:50] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ physical.pos: int [1:50] 112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ allele1     : chr [1:50] "A" "T" "T" "T" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ allele2     : chr [1:50] "T" "A" "A" "A" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "class")= chr "bigSNP"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># remove only first 50 samples and SNPs</span></span></span>
<span class="r-in"><span><span class="va">rdsfile2</span> <span class="op">&lt;-</span> <span class="fu">snp_subset</span><span class="op">(</span><span class="va">test</span>, ind.row <span class="op">=</span> <span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span>, ind.col <span class="op">=</span> <span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="va">rdsfile2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 16 fields</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr      :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr_rw   :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ nrow        : int 467</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ ncol        : int 4492</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type        : Named int 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..- attr(*, "names")= chr "unsigned char"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ backingfile : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpITJXCm\\file5f8c66ca6c85_sub2.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_read_only: logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address     :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address_rw  :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ bk          : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpITJXCm\\file5f8c66ca6c85_sub2.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ rds         : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\RtmpITJXCm\\file5f8c66ca6c85_sub2.rds"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_saved    : logi TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_chr    : chr "unsigned char"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_size   : int 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ file_size   : num 2097764</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ code256     : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..and 26 methods, of which 12 are  possibly relevant:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  check_write_permissions, copy#envRefClass, initialize, initialize#FBM,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  save, show#FBM, show#envRefClass</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ fam      :'data.frame':	467 obs. of  6 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ family.ID  : chr [1:467] "POP2" "POP2" "POP2" "POP2" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ sample.ID  : chr [1:467] "IND50" "IND51" "IND52" "IND53" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ paternal.ID: int [1:467] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ maternal.ID: int [1:467] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ sex        : int [1:467] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ affection  : int [1:467] 1 1 2 1 1 1 1 1 1 2 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ map      :'data.frame':	4492 obs. of  6 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ chromosome  : int [1:4492] 1 1 1 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ marker.ID   : chr [1:4492] "SNP50" "SNP51" "SNP52" "SNP53" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ genetic.dist: int [1:4492] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ physical.pos: int [1:4492] 50100 51109 52101 53106 54101 55097 56103 57110 58101 59088 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ allele1     : chr [1:4492] "A" "A" "A" "T" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ allele2     : chr [1:4492] "T" "T" "T" "A" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "class")= chr "bigSNP"</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

