<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Correlation — snp_cor • bigsnpr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Correlation — snp_cor" />
<meta property="og:description" content="Get significant correlations between nearby SNPs of the same chromosome
(p-values are computed using a two-sided t-test)." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Computing polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Correlation</h1>
    <small class="dont-index">Source: <a href='https://github.com/privefl/bigsnpr/blob/master/R/corr.R'><code>R/corr.R</code></a></small>
    <div class="hidden name"><code>snp_cor.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Get significant correlations between nearby SNPs of the same chromosome
(p-values are computed using a two-sided t-test).</p>
    </div>

    <pre class="usage"><span class='fu'>snp_cor</span>(
  <span class='no'>Gna</span>,
  <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>Gna</span>),
  <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fu'>cols_along</span>(<span class='no'>Gna</span>),
  <span class='kw'>size</span> <span class='kw'>=</span> <span class='fl'>500</span>,
  <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>fill.diag</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>infos.pos</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>Gna</th>
      <td><p>A FBM.code256
(typically <code>&lt;bigSNP&gt;$genotypes</code>).<br />
You can have missing values in these data.</p></td>
    </tr>
    <tr>
      <th>ind.row</th>
      <td><p>An optional vector of the row indices (individuals) that
are used. If not specified, all rows are used.<br />
<strong>Don't use negative indices.</strong></p></td>
    </tr>
    <tr>
      <th>ind.col</th>
      <td><p>An optional vector of the column indices (SNPs) that are used.
If not specified, all columns are used.<br />
<strong>Don't use negative indices.</strong></p></td>
    </tr>
    <tr>
      <th>size</th>
      <td><p>For one SNP, window size around this SNP to compute correlations.
Default is <code>500</code>. If not providing <code>infos.pos</code> (<code>NULL</code>, the default), this is
a window in number of SNPs, otherwise it is a window in kb (genetic distance).</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>Type-I error for testing correlations.
Default is <code>1</code> (no threshold is applied).</p></td>
    </tr>
    <tr>
      <th>fill.diag</th>
      <td><p>Whether to fill the diagonal with 1s (the default)
or to keep it as 0s.</p></td>
    </tr>
    <tr>
      <th>infos.pos</th>
      <td><p>Vector of integers specifying the physical position
on a chromosome (in base pairs) of each SNP.<br />
Typically <code>&lt;bigSNP&gt;$map$physical.pos</code>.</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>Number of cores used. Default doesn't use parallelism.
You may use nb_cores.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The correlation matrix. This is a sparse symmetric matrix.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>test</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>()

<span class='no'>corr</span> <span class='kw'>&lt;-</span> <span class='fu'>snp_cor</span>(<span class='no'>test</span>$<span class='no'>genotypes</span>, <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fl'>1000</span>)
<span class='no'>corr</span>[<span class='fl'>1</span>:<span class='fl'>10</span>, <span class='fl'>1</span>:<span class='fl'>10</span>]</div><div class='output co'>#&gt; 10 x 10 sparse Matrix of class "dsCMatrix"
#&gt;                                                                      
#&gt;  [1,]  1.000000000 -0.059420393  0.005664354 -0.024489153  0.07467733
#&gt;  [2,] -0.059420393  1.000000000  0.011193114  0.069041892  0.01201638
#&gt;  [3,]  0.005664354  0.011193114  1.000000000 -0.102606763  0.03146612
#&gt;  [4,] -0.024489153  0.069041892 -0.102606763  1.000000000  0.06581927
#&gt;  [5,]  0.074677326  0.012016376  0.031466118  0.065819274  1.00000000
#&gt;  [6,] -0.016720892  0.006950400  0.075573785 -0.009996247  0.01995194
#&gt;  [7,]  0.003680272 -0.073816948  0.015793311  0.067027181  0.03088233
#&gt;  [8,] -0.017764606 -0.033245425 -0.087165386  0.052163271 -0.04179906
#&gt;  [9,]  0.111349765  0.008661176 -0.060211948  0.062986807 -0.03244854
#&gt; [10,] -0.051804449 -0.065532091  0.017048961 -0.093032007 -0.01146874
#&gt;                                                                      
#&gt;  [1,] -0.016720892  0.003680272 -0.01776461  0.111349765 -0.051804449
#&gt;  [2,]  0.006950400 -0.073816948 -0.03324542  0.008661176 -0.065532091
#&gt;  [3,]  0.075573785  0.015793311 -0.08716539 -0.060211948  0.017048961
#&gt;  [4,] -0.009996247  0.067027181  0.05216327  0.062986807 -0.093032007
#&gt;  [5,]  0.019951940  0.030882334 -0.04179906 -0.032448541 -0.011468736
#&gt;  [6,]  1.000000000  0.053884169 -0.04024761 -0.045578013  0.052820407
#&gt;  [7,]  0.053884169  1.000000000 -0.06369954 -0.012107321  0.006552636
#&gt;  [8,] -0.040247606 -0.063699540  1.00000000 -0.022351147 -0.058557214
#&gt;  [9,] -0.045578013 -0.012107321 -0.02235115  1.000000000 -0.041503959
#&gt; [10,]  0.052820407  0.006552636 -0.05855721 -0.041503959  1.000000000</div><div class='input'>
<span class='co'># Sparsity</span>
<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>corr</span>@<span class='kw'>x</span>) / <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>corr</span>)</div><div class='output co'>#&gt; [1] 0.375676</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


