<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Save modifications — snp_save • bigsnpr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Save modifications — snp_save"><meta property="og:description" content='Save a bigSNP after having made some modifications to it.
As bigSNP is an S3 class, you can add any slot you want
to an object of this class, then use snp_save to
save these modifications in the corresponding ".rds" backing file.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.12.20</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/PLR.html">Polygenic scores using individual-level data</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Polygenic scores and inference using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
    <li>
      <a href="https://privefl.github.io/bigsnpr-extdoc/" class="external-link">Extended documentation for bigsnpr &amp; bigstatsr</a>
    </li>
    <li>
      <a href="https://privefl.github.io/statgen-course/" class="external-link">Statistical Human Genetics course using R</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Save modifications</h1>
    <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/R/get-save-infos.R" class="external-link"><code>R/get-save-infos.R</code></a></small>
    <div class="hidden name"><code>snp_save.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Save a <code>bigSNP</code> after having made some modifications to it.
As <code>bigSNP</code> is an S3 class, you can add any slot you want
to an object of this class, then use <code>snp_save</code> to
save these modifications in the corresponding ".rds" backing file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">snp_save</span><span class="op">(</span><span class="va">x</span>, version <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>A <a href="bigSNP-class.html">bigSNP</a>.</p></dd>


<dt>version</dt>
<dd><p>the workspace format version to use.  <code>NULL</code>
    specifies the current default version (3). The only other supported
    value is 2, the default from <span style="R">R</span> 1.4.0 to <span style="R">R</span> 3.5.0.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>The (saved) <code>bigSNP</code>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Reading example</span></span></span>
<span class="r-in"><span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="snp_attachExtdata.html">snp_attachExtdata</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># I can add whatever I want to an S3 class</span></span></span>
<span class="r-in"><span><span class="va">test</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">`p-values`</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">test</span><span class="op">$</span><span class="va">map</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">test</span><span class="op">$</span><span class="va">map</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 'data.frame':	4542 obs. of  7 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ chromosome  : int  1 1 1 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ marker.ID   : chr  "SNP0" "SNP1" "SNP2" "SNP3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ genetic.dist: int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ physical.pos: int  112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ allele1     : chr  "A" "T" "T" "T" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ allele2     : chr  "T" "A" "A" "A" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ p-values    : num  0.266 0.372 0.573 0.908 0.202 ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Reading again</span></span></span>
<span class="r-in"><span><span class="va">rds</span> <span class="op">&lt;-</span> <span class="va">test</span><span class="op">$</span><span class="va">genotypes</span><span class="op">$</span><span class="va">rds</span></span></span>
<span class="r-in"><span><span class="va">test2</span> <span class="op">&lt;-</span> <span class="fu"><a href="snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="va">rds</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">test2</span><span class="op">$</span><span class="va">map</span><span class="op">)</span> <span class="co"># new slot wasn't saved</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 'data.frame':	4542 obs. of  6 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ chromosome  : int  1 1 1 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ marker.ID   : chr  "SNP0" "SNP1" "SNP2" "SNP3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ genetic.dist: int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ physical.pos: int  112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ allele1     : chr  "A" "T" "T" "T" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ allele2     : chr  "T" "A" "A" "A" ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Save it</span></span></span>
<span class="r-in"><span><span class="fu">snp_save</span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Reading again</span></span></span>
<span class="r-in"><span><span class="va">test3</span> <span class="op">&lt;-</span> <span class="fu"><a href="snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="va">rds</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">test3</span><span class="op">$</span><span class="va">map</span><span class="op">)</span> <span class="co"># it is saved now</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 'data.frame':	4542 obs. of  7 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ chromosome  : int  1 1 1 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ marker.ID   : chr  "SNP0" "SNP1" "SNP2" "SNP3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ genetic.dist: int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ physical.pos: int  112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ allele1     : chr  "A" "T" "T" "T" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ allele2     : chr  "T" "A" "A" "A" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ p-values    : num  0.266 0.372 0.573 0.908 0.202 ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># The complicated code of this function</span></span></span>
<span class="r-in"><span><span class="va">snp_save</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> function (x, version = NULL) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> {</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     saveRDS(x, file = x$genotypes$rds, version = version)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     invisible(x)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> }</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;bytecode: 0x000002020baf7f70&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> &lt;environment: namespace:bigsnpr&gt;</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

