<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Read PLINK files into a "bigSNP" — snp_readBed • bigsnpr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Read PLINK files into a " bigsnp snp_readbed><meta property="og:description" content="Functions to read bed/bim/fam files into a bigSNP."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.11.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/ancestry.html">Ancestry proportions and ancestry grouping</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Computing polygenic scores using LDpred2</a>
    </li>
    <li>
      <a href="../articles/prs_uncertainty.html">Estimating uncertainty in polygenic scores using LDpred2</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://privefl.github.io/about.html" class="external-link">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr" class="external-link">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o fa"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read PLINK files into a "bigSNP"</h1>
    <small class="dont-index">Source: <a href="https://github.com/privefl/bigsnpr/blob/HEAD/R/read-plink.R" class="external-link"><code>R/read-plink.R</code></a></small>
    <div class="hidden name"><code>snp_readBed.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Functions to read bed/bim/fam files into a <a href="bigSNP-class.html">bigSNP</a>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">snp_readBed</span><span class="op">(</span><span class="va">bedfile</span>, backingfile <span class="op">=</span> <span class="fu"><a href="sub_bed.html">sub_bed</a></span><span class="op">(</span><span class="va">bedfile</span><span class="op">)</span><span class="op">)</span>

<span class="fu">snp_readBed2</span><span class="op">(</span>
  <span class="va">bedfile</span>,
  backingfile <span class="op">=</span> <span class="fu"><a href="sub_bed.html">sub_bed</a></span><span class="op">(</span><span class="va">bedfile</span><span class="op">)</span>,
  ind.row <span class="op">=</span> <span class="fu">rows_along</span><span class="op">(</span><span class="va">obj.bed</span><span class="op">)</span>,
  ind.col <span class="op">=</span> <span class="fu">cols_along</span><span class="op">(</span><span class="va">obj.bed</span><span class="op">)</span>,
  ncores <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>bedfile</dt>
<dd><p>Path to file with extension ".bed" to read.
You need the corresponding ".bim" and ".fam" in the same directory.</p></dd>


<dt>backingfile</dt>
<dd><p>The path (without extension) for the backing files
for the cache of the <a href="bigSNP-class.html">bigSNP</a> object. Default takes the bedfile
without the ".bed" extension.</p></dd>


<dt>ind.row</dt>
<dd><p>An optional vector of the row indices (individuals) that
are used. If not specified, all rows are used.<br><strong>Don't use negative indices.</strong></p></dd>


<dt>ind.col</dt>
<dd><p>An optional vector of the column indices (SNPs) that are used.
If not specified, all columns are used.<br><strong>Don't use negative indices.</strong></p></dd>


<dt>ncores</dt>
<dd><p>Number of cores used. Default doesn't use parallelism.
You may use nb_cores.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>The path to the RDS file that stores the <code>bigSNP</code> object.
Note that this function creates one other file which stores the values of
the Filebacked Big Matrix.<br></p>


<p><strong>You shouldn't read from PLINK files more than once.</strong> Instead, use
<a href="snp_attach.html">snp_attach</a> to load the "bigSNP" object in any R session from backing files.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>For more information on these formats, please visit
<a href="https://www.cog-genomics.org/plink/1.9/formats#bed" class="external-link">PLINK webpage</a>.
For other formats, please use PLINK to convert them in bedfiles,
which require minimal space to store and are faster to read. For example,
to convert from a VCF file, use the <code>--vcf</code> option. See <a href="snp_plinkQC.html">snp_plinkQC</a>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="op">(</span><span class="va">bedfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example.bed"</span>, package <span class="op">=</span> <span class="st">"bigsnpr"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "C:/Users/au639593/Documents/R/win-library/3.6/bigsnpr/extdata/example.bed"</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Reading the bedfile and storing the data in temporary directory</span></span>
<span class="r-in"><span class="va">rds</span> <span class="op">&lt;-</span> <span class="fu">snp_readBed</span><span class="op">(</span><span class="va">bedfile</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Loading the data from backing files</span></span>
<span class="r-in"><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="va">rds</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 16 fields</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr      :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr_rw   :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ nrow        : int 517</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ ncol        : int 4542</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type        : Named int 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..- attr(*, "names")= chr "unsigned char"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ backingfile : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\Rtmp4qNGV2\\file66478da1742.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_read_only: logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address     :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address_rw  :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ bk          : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\Rtmp4qNGV2\\file66478da1742.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ rds         : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\Rtmp4qNGV2\\file66478da1742.rds"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_saved    : logi TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_chr    : chr "unsigned char"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_size   : int 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ file_size   : num 2348214</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ code256     : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..and 26 methods, of which 12 are  possibly relevant:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  check_write_permissions, copy#envRefClass, initialize, initialize#FBM,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  save, show#FBM, show#envRefClass</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ fam      :'data.frame':	517 obs. of  6 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ family.ID  : chr [1:517] "POP1" "POP1" "POP1" "POP1" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ sample.ID  : chr [1:517] "IND0" "IND1" "IND2" "IND3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ paternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ maternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ sex        : int [1:517] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ affection  : int [1:517] 1 1 2 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ map      :'data.frame':	4542 obs. of  6 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ chromosome  : int [1:4542] 1 1 1 1 1 1 1 1 1 1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ marker.ID   : chr [1:4542] "SNP0" "SNP1" "SNP2" "SNP3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ genetic.dist: int [1:4542] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ physical.pos: int [1:4542] 112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ allele1     : chr [1:4542] "A" "T" "T" "T" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ allele2     : chr [1:4542] "T" "A" "A" "A" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  - attr(*, "class")= chr "bigSNP"</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">G</span> <span class="op">&lt;-</span> <span class="va">test</span><span class="op">$</span><span class="va">genotypes</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]  517 4542</span>
<span class="r-in"><span class="va">G</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]    0    0    2    0    1    1    0    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]    1    0    1    0    0    1    0    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,]    0    1    1    0    2    1    0    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,]    0    0    2    0    2    1    0    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,]    1    0    0    0    2    2    1    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6,]    0    1    0    0    2    1    0    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [7,]    0    1    1    0    2    2    1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [8,]    1    0    1    1    1    1    0    1</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

